<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>SCRIPTS: orthoXML &mdash; A Python Environment for (phylogenetic) Tree Exploration</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '2.3.6',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="top" title="A Python Environment for (phylogenetic) Tree Exploration" href="../index.html" />
    <link rel="up" title="The ETE tutorial" href="index.html" />
    <link rel="next" title="ETE’s Reference Guide" href="../reference/index.html" />
    <link rel="prev" title="Dealing with the NCBI Taxonomy database" href="tutorial_ncbitaxonomy.html" />
 

<!-- <div class="warning" align="center"> -->
<!--   This site is still under construction and refers to -->
<!--   the new <b>ETE 2.1</b> version (temporary module -->
<!--   name: <b>ete2a1</b>). Many sections may be incomplete or contain -->
<!--   errors. -->
<!-- </div> -->


<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-783841-12']);
  _gaq.push(['_setCookiePath', '/ete2/']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>






  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../reference/index.html" title="ETE’s Reference Guide"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial_ncbitaxonomy.html" title="Dealing with the NCBI Taxonomy database"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">ETE documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">The ETE tutorial</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <span class="target" id="module-ete2"></span><div class="section" id="scripts-orthoxml">
<span id="etree2orthoxml"></span><h1><a class="toc-backref" href="#id1">SCRIPTS: orthoXML</a><a class="headerlink" href="#scripts-orthoxml" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#scripts-orthoxml" id="id1">SCRIPTS: orthoXML</a><ul>
<li><a class="reference internal" href="#orthoxml-parser" id="id2">OrthoXML parser</a></li>
<li><a class="reference internal" href="#the-etree2orthoxml-script" id="id3">The etree2orthoxml script</a><ul>
<li><a class="reference internal" href="#usage" id="id4">Usage</a></li>
<li><a class="reference internal" href="#example-using-custom-evolutionary-annotation" id="id5">Example: Using custom evolutionary annotation</a></li>
<li><a class="reference internal" href="#example-automatic-detection-of-species-names" id="id6">Example: Automatic detection of species names</a></li>
<li><a class="reference internal" href="#example-tree-rooting" id="id7">Example: Tree rooting</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="orthoxml-parser">
<h2><a class="toc-backref" href="#id2">OrthoXML parser</a><a class="headerlink" href="#orthoxml-parser" title="Permalink to this headline">¶</a></h2>
<p><tt class="xref py py-attr docutils literal"><span class="pre">etree2orthoxml</span></tt> is a python script distributed as a part of the
ETE toolkit package. It uses an automatic python parser generated on
the basis of the OrthoXML schema to convert the evolutionary events in
phylogenetic tree topologies into the orthoXML format.</p>
<p>ETE OrthoXML parser is a low level python module that allows to
operate with the orthoXML structure using python objects. Every
element defined in the orthoXML schema has its akin in the parser
module, so a complete orthoXML structure can be generated from scratch
within a python script. In other words, low level access to the
orthoXML parser allows to create orthoxml documents in a programmatic
way.</p>
<p>The following example will create a basic orthoXML document</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete2</span> <span class="kn">import</span> <span class="n">orthoxml</span>
<span class="c"># Creates an empty orthoXML object</span>
<span class="n">oxml</span> <span class="o">=</span> <span class="n">orthoxml</span><span class="o">.</span><span class="n">orthoXML</span><span class="p">()</span>

<span class="c"># Add an ortho group container to the orthoXML document</span>
<span class="n">ortho_groups</span> <span class="o">=</span> <span class="n">orthoxml</span><span class="o">.</span><span class="n">groups</span><span class="p">()</span>
<span class="n">oxml</span><span class="o">.</span><span class="n">set_groups</span><span class="p">(</span><span class="n">ortho_groups</span><span class="p">)</span>

<span class="c"># Add an orthology group including two sequences</span>
<span class="n">orthologs</span> <span class="o">=</span> <span class="n">orthoxml</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
<span class="n">orthologs</span><span class="o">.</span><span class="n">add_geneRef</span><span class="p">(</span><span class="n">orthoxml</span><span class="o">.</span><span class="n">geneRef</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
<span class="n">orthologs</span><span class="o">.</span><span class="n">add_geneRef</span><span class="p">(</span><span class="n">orthoxml</span><span class="o">.</span><span class="n">geneRef</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span>
<span class="n">ortho_groups</span><span class="o">.</span><span class="n">add_orthologGroup</span><span class="p">(</span><span class="n">orthologs</span><span class="p">)</span>

<span class="n">oxml_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&quot;test_orthoxml.xml&quot;</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span>
<span class="n">oxml</span><span class="o">.</span><span class="n">export</span><span class="p">(</span><span class="n">oxml_file</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">oxml_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<span class="c"># producing the following output</span>
<span class="c">#&lt;ortho:orthoXML&gt;</span>
<span class="c">#   &lt;ortho:groups&gt;</span>
<span class="c">#       &lt;ortho:orthologGroup&gt;</span>
<span class="c">#           &lt;ortho:geneRef id=&quot;1&quot;/&gt;</span>
<span class="c">#           &lt;ortho:geneRef id=&quot;2&quot;/&gt;</span>
<span class="c">#       &lt;/ortho:orthologGroup&gt;</span>
<span class="c">#   &lt;/ortho:groups&gt;</span>
<span class="c">#&lt;/ortho:orthoXML&gt;</span>
</pre></div>
</div>
</div>
<div class="section" id="the-etree2orthoxml-script">
<h2><a class="toc-backref" href="#id3">The etree2orthoxml script</a><a class="headerlink" href="#the-etree2orthoxml-script" title="Permalink to this headline">¶</a></h2>
<p><tt class="xref py py-attr docutils literal"><span class="pre">etree2orthoxml</span></tt> is a standalone python script that allows to
read a phylogenetic tree in newick format and export their
evolutionary events (duplication and speciation events) as an orthoXML
document. The program is installed along with ETE, so it should be
found in your path. Alternatively you can found it in the script
folder of the latest ETE package release
(<a class="reference external" href="http://etetoolkit.org/releases/ete2/">http://etetoolkit.org/releases/ete2/</a>).</p>
<p>To work, <tt class="xref py py-attr docutils literal"><span class="pre">etree2orthoxml</span></tt> requires only one argument containing
the newick representation of a tree or the name of the file that
contains it. By default, automatic detection of speciation and
duplication events will be carried out using the built-in
<a class="reference internal" href="tutorial_phylogeny.html#spoverlap"><em>species overlap algorithm</em></a>, although this behavior
can be easily disabled when event information is provided along with
the newick tree. In the following sections you will find some use case
examples.</p>
<p>Also, consider reading the source code of the script. It is documented
and it can be used as a template for more specific applications. Note
that <tt class="xref py py-attr docutils literal"><span class="pre">etree2orthoxml</span></tt> is a work in progress, so feel free to use
the <a class="reference external" href="https://groups.google.com/forum/#!forum/etetoolkit">etetoolkit mailing list</a> to report any
feedback or improvement to the code.</p>
<div class="section" id="usage">
<h3><a class="toc-backref" href="#id4">Usage</a><a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><pre>usage: etree2orthoxml [-h] [--sp_delimiter SPECIES_DELIMITER]
                      [--sp_field SPECIES_FIELD] [--root [ROOT [ROOT ...]]]
                      [--skip_ortholog_detection]
                      [--evoltype_attr EVOLTYPE_ATTR] [--database DATABASE]
                      [--show] [--ascii] [--newick]
                      tree_file

etree2orthoxml is a python script that extracts evolutionary events
(speciation and duplication) from a newick tree and exports them as a
OrthoXML file.

positional arguments:
  tree_file             A tree file (or text string) in newick format.

optional arguments:
  -h, --help            show this help message and exit
  --sp_delimiter SPECIES_DELIMITER
                        When species names are guessed from node names, this
                        argument specifies how to split node name to guess the
                        species code
  --sp_field SPECIES_FIELD
                        When species names are guessed from node names, this
                        argument specifies the position of the species name
                        code relative to the name splitting delimiter
  --root [ROOT [ROOT ...]]
                        Roots the tree to the node grouping the list of node
                        names provided (space separated). In example:'--root
                        human rat mouse'
  --skip_ortholog_detection
                        Skip automatic detection of speciation and duplication
                        events, thus relying in the correct annotation of the
                        provided tree using the extended newick format (i.e.
                        '((A, A)[&amp;&amp;NHX:evoltype=D], B)[&amp;&amp;NHX:evoltype=S];')
  --evoltype_attr EVOLTYPE_ATTR
                        When orthology detection is disabled, the attribute
                        name provided here will be expected to exist in all
                        internal nodes and read from the extended newick
                        format
  --database DATABASE   Database name
  --show                Show the tree and its evolutionary events before
                        orthoXML export
  --ascii               Show the tree using ASCII representation and all its
                        evolutionary events before orthoXML export
  --newick              print the extended newick format for provided tree
                        using ASCII representation and all its evolutionary
                        events before orthoXML export</pre>
</div>
</div>
<div class="section" id="example-using-custom-evolutionary-annotation">
<h3><a class="toc-backref" href="#id5">Example: Using custom evolutionary annotation</a><a class="headerlink" href="#example-using-custom-evolutionary-annotation" title="Permalink to this headline">¶</a></h3>
<p>If all internal nodes in the provided tree are correctly label as
duplication or speciation nodes, automatic detection of events can be
disabled using the <tt class="xref py py-attr docutils literal"><span class="pre">--skip_ortholog_detection</span></tt> flag.</p>
<p>Node labeling should be provided using the extended newick
format. Duplication nodes should contain the label <tt class="xref py py-attr docutils literal"><span class="pre">evoltype</span></tt>
set to <tt class="xref py py-attr docutils literal"><span class="pre">D</span></tt>, while speciation nodes should be set to
<tt class="xref py py-attr docutils literal"><span class="pre">evoltype=S</span></tt>. If tag names is different, the option
<tt class="xref py py-attr docutils literal"><span class="pre">evoltype_attr</span></tt> can be used as convenient.</p>
<p>In the following example, we force the HUMAN clade to be considered a
speciation node.</p>
<div class="highlight-python"><pre># etree2orthoxml --skip_ortholog_detection '((HUMAN_A, HUMAN_B)[&amp;&amp;NHX:evoltype=S], MOUSE_B)[&amp;&amp;NHX:evoltype=S];'

 &lt;orthoXML&gt;
     &lt;species name="A"&gt;
         &lt;database name=""&gt;
            &lt;genes&gt;
                 &lt;gene protId="HUMAN_A" id="0"/&gt;
             &lt;/genes&gt;
         &lt;/database&gt;
     &lt;/species&gt;
     &lt;species name="B"&gt;
         &lt;database name=""&gt;
             &lt;genes&gt;
                 &lt;gene protId="HUMAN_B" id="1"/&gt;
                 &lt;gene protId="MOUSE_B" id="2"/&gt;
             &lt;/genes&gt;
         &lt;/database&gt;
     &lt;/species&gt;
     &lt;groups&gt;
         &lt;orthologGroup&gt;
             &lt;geneRef id="2"/&gt;
             &lt;orthologGroup&gt;
                 &lt;geneRef id="0"/&gt;
                 &lt;geneRef id="1"/&gt;
             &lt;/orthologGroup&gt;
         &lt;/orthologGroup&gt;
     &lt;/groups&gt;
 &lt;/orthoXML&gt;</pre>
</div>
<p>You can avoid tree reformatting when node labels are slightly
different by using the <tt class="xref py py-attr docutils literal"><span class="pre">evoltype_attr</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># etree2orthoxml --evoltype_attr E --skip_ortholog_detection &#39;((HUMAN_A, HUMAN_B)[&amp;&amp;NHX:E=S], MOUSE_B)[&amp;&amp;NHX:E=S];&#39;</span>
</pre></div>
</div>
<p>However, more complex modifications on raw trees can be easily
performed using the core methods of the ETE library, so they match the
requirements of the <tt class="xref py py-attr docutils literal"><span class="pre">etree2orthoxml</span></tt> script.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">ete2</span> <span class="kn">import</span> <span class="n">Tree</span>
<span class="c"># Having the followin tree</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">Tree</span><span class="p">(</span><span class="s">&#39;((HUMAN_A, HUMAN_B)[&amp;&amp;NHX:speciation=N], MOUSE_B)[&amp;&amp;NHX:speciation=Y];&#39;</span><span class="p">)</span>

<span class="c"># We read the speciation tag from nodes and convert it into a vaild evoltree label</span>
<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
   <span class="k">if</span> <span class="ow">not</span> <span class="n">node</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">():</span>
      <span class="n">etype</span> <span class="o">=</span> <span class="s">&quot;D&quot;</span> <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">speciation</span> <span class="o">==</span> <span class="s">&quot;N&quot;</span> <span class="k">else</span> <span class="s">&quot;S&quot;</span>
      <span class="n">node</span><span class="o">.</span><span class="n">add_features</span><span class="p">(</span><span class="n">evoltype</span><span class="o">=</span><span class="n">etype</span><span class="p">)</span>

<span class="c"># We the export a newick string that is compatible with etree2orthoxml script</span>
<span class="n">t</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s">&quot;evoltype&quot;</span><span class="p">],</span> <span class="n">format_root_node</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

<span class="c"># converted newick:</span>
<span class="c"># &#39;((HUMAN_A:1,HUMAN_B:1)1:1[&amp;&amp;NHX:evoltype=D],MOUSE_B:1)1:1[&amp;&amp;NHX:evoltype=S];&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="example-automatic-detection-of-species-names">
<h3><a class="toc-backref" href="#id6">Example: Automatic detection of species names</a><a class="headerlink" href="#example-automatic-detection-of-species-names" title="Permalink to this headline">¶</a></h3>
<p>As different databases and software may produce slightly different
newick tree formats, the script provides several customization
options.</p>
<p>In gene family trees, species names are usually encoded as a part of
leaf names (i.e. P53_HUMAN). If such codification follows a simple
rule, <tt class="xref py py-attr docutils literal"><span class="pre">etree2orthoxml</span></tt> can automatically detect species name and
used to populate the relevant sections within the orthoXML document.</p>
<p>For this, the <tt class="xref py py-attr docutils literal"><span class="pre">sp_delimiter</span></tt> and <tt class="xref py py-attr docutils literal"><span class="pre">sp_field</span></tt> arguments can
be used. Note how species are correctly detected in the following example:</p>
<div class="highlight-python"><pre># etree2orthoxml --database TestDB --evoltype_attr E --skip_ortholog_detection --sp_delimiter '_' --sp_field 0  '((HUMAN_A, HUMAN_B)[&amp;&amp;NHX:E=S], MOUSE_B)[&amp;&amp;NHX:E=S];'
&lt;orthoXML&gt;
 &lt;species name="HUMAN"&gt;
     &lt;database name="TestDB"&gt;
         &lt;genes&gt;
             &lt;gene protId="HUMAN_A" id="0"/&gt;
             &lt;gene protId="HUMAN_B" id="1"/&gt;
         &lt;/genes&gt;
     &lt;/database&gt;
 &lt;/species&gt;
 &lt;species name="MOUSE"&gt;
     &lt;database name="TestDB"&gt;
         &lt;genes&gt;
             &lt;gene protId="MOUSE_B" id="2"/&gt;
         &lt;/genes&gt;
     &lt;/database&gt;
 &lt;/species&gt;
 &lt;groups&gt;
     &lt;orthologGroup&gt;
         &lt;geneRef id="2"/&gt;
         &lt;orthologGroup&gt;
             &lt;geneRef id="0"/&gt;
             &lt;geneRef id="1"/&gt;
         &lt;/orthologGroup&gt;
     &lt;/orthologGroup&gt;
 &lt;/groups&gt;
&lt;/orthoXML&gt;</pre>
</div>
</div>
<div class="section" id="example-tree-rooting">
<h3><a class="toc-backref" href="#id7">Example: Tree rooting</a><a class="headerlink" href="#example-tree-rooting" title="Permalink to this headline">¶</a></h3>
<p>When evolutionary events are expected to be automatically inferred
from tree topology, outgroup information can be passed to the program to
root the tree before performing the detection.</p>
<div class="highlight-python"><pre># etree2orthoxml --ascii --root FLY_1 FLY_2 --sp_delimiter '_' --sp_field 0  '((HUMAN_A, HUMAN_B), MOUSE_B, (FLY_1, FLY_2));'



                      /-FLY_1
             /D, NoName
            |         \-FLY_2
    -S, NoName
            |                  /-HUMAN_A
            |         /D, NoName
             \S, NoName        \-HUMAN_B
                     |
                      \-MOUSE_B


    &lt;orthoXML&gt;
        &lt;species name="FLY"&gt;
            &lt;database name=""&gt;
                &lt;genes&gt;
                    &lt;gene protId="FLY_1" id="0"/&gt;
                    &lt;gene protId="FLY_2" id="1"/&gt;
                &lt;/genes&gt;
            &lt;/database&gt;
        &lt;/species&gt;
        &lt;species name="HUMAN"&gt;
            &lt;database name=""&gt;
                &lt;genes&gt;
                    &lt;gene protId="HUMAN_A" id="2"/&gt;
                    &lt;gene protId="HUMAN_B" id="3"/&gt;
                &lt;/genes&gt;
            &lt;/database&gt;
        &lt;/species&gt;
        &lt;species name="MOUSE"&gt;
            &lt;database name=""&gt;
                &lt;genes&gt;
                    &lt;gene protId="MOUSE_B" id="4"/&gt;
                &lt;/genes&gt;
            &lt;/database&gt;
        &lt;/species&gt;
        &lt;groups&gt;
            &lt;orthologGroup&gt;
                &lt;paralogGroup&gt;
                    &lt;geneRef id="0"/&gt;
                    &lt;geneRef id="1"/&gt;
                &lt;/paralogGroup&gt;
                &lt;orthologGroup&gt;
                    &lt;geneRef id="4"/&gt;
                    &lt;paralogGroup&gt;
                        &lt;geneRef id="2"/&gt;
                        &lt;geneRef id="3"/&gt;
                    &lt;/paralogGroup&gt;
                &lt;/orthologGroup&gt;
            &lt;/orthologGroup&gt;
        &lt;/groups&gt;
    &lt;/orthoXML&gt;</pre>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/ete_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">SCRIPTS: orthoXML</a><ul>
<li><a class="reference internal" href="#orthoxml-parser">OrthoXML parser</a></li>
<li><a class="reference internal" href="#the-etree2orthoxml-script">The etree2orthoxml script</a><ul>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#example-using-custom-evolutionary-annotation">Example: Using custom evolutionary annotation</a></li>
<li><a class="reference internal" href="#example-automatic-detection-of-species-names">Example: Automatic detection of species names</a></li>
<li><a class="reference internal" href="#example-tree-rooting">Example: Tree rooting</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial_ncbitaxonomy.html"
                        title="previous chapter">Dealing with the NCBI Taxonomy database</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="../reference/index.html"
                        title="next chapter">ETE&#8217;s Reference Guide</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/tutorial/tutorial_etree2orthoxml.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="../reference/index.html" title="ETE’s Reference Guide"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial_ncbitaxonomy.html" title="Dealing with the NCBI Taxonomy database"
             >previous</a> |</li>
        <li><a href="../index.html">ETE documentation</a> &raquo;</li>
          <li><a href="index.html" >The ETE tutorial</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2015, Jaime Huerta-Cepas.
      Last updated on Jul 24, 2015.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b1.
    </div>
  </body>
</html>