<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN"
            "http://www.w3.org/TR/REC-html40/loose.dtd">
<HTML>
<HEAD>
<TITLE>Biopython Tutorial and Cookbook
</TITLE>

<META http-equiv="Content-Type" content="text/html; charset=US-ASCII">
<META name="GENERATOR" content="hevea 1.10">
<LINK rel="stylesheet" type="text/css" href="Tutorial.css">
</HEAD>
<BODY >
<!--HEVEA command line is: hevea -fix Tutorial.tex -->
<!--HACHA command line is: hacha Tutorial.html -->
<P>
<P ALIGN="center">
<IMG ALIGN="center" SRC="images/biopython.jpg" TITLE="Biopython Logo" ALT="[Biopython Logo]" width="1024" height="288" />
</p>
</P><TABLE CLASS="title"><TR><TD><H1 CLASS="titlemain">Biopython Tutorial and Cookbook</H1><H3 CLASS="titlerest">Jeff Chang, Brad Chapman, Iddo Friedberg, Thomas Hamelryck,<BR>
Michiel de Hoon, Peter Cock, Tiago Ant&#XE3;o</H3><H3 CLASS="titlerest">Last Update &#X2013; 3 April 2009 (Biopython 1.50 <I>beta</I>)</H3></TD></TR>
</TABLE><UL>
<LI><A HREF="Tutorial001.html">Contents</A>
</LI><LI><A HREF="Tutorial002.html">Introduction</A>
<UL>
<LI><A HREF="Tutorial002.html#toc1">What is Biopython?</A>
</LI><LI><A HREF="Tutorial002.html#toc2">Installing Biopython</A>
</LI><LI><A HREF="Tutorial002.html#toc3">FAQ</A>
</LI></UL>
</LI><LI><A HREF="Tutorial003.html">Quick Start &#X2013; What can you do with Biopython?</A>
<UL>
<LI><A HREF="Tutorial003.html#toc4">General overview of what Biopython provides</A>
</LI><LI><A HREF="Tutorial003.html#toc5">Working with sequences</A>
</LI><LI><A HREF="Tutorial003.html#toc6">A usage example</A>
</LI><LI><A HREF="Tutorial003.html#toc7">Parsing sequence file formats</A>
</LI><LI><A HREF="Tutorial003.html#toc8">Connecting with biological databases</A>
</LI><LI><A HREF="Tutorial003.html#toc9">What to do next</A>
</LI></UL>
</LI><LI><A HREF="Tutorial004.html">Sequence objects</A>
<UL>
<LI><A HREF="Tutorial004.html#toc10">Sequences and Alphabets</A>
</LI><LI><A HREF="Tutorial004.html#toc11">Sequences act like strings</A>
</LI><LI><A HREF="Tutorial004.html#toc12">Slicing a sequence</A>
</LI><LI><A HREF="Tutorial004.html#toc13">Turning Seq objects into strings</A>
</LI><LI><A HREF="Tutorial004.html#toc14">Concatenating or adding sequences</A>
</LI><LI><A HREF="Tutorial004.html#toc15">Nucleotide sequences and (reverse) complements</A>
</LI><LI><A HREF="Tutorial004.html#toc16">Transcription</A>
</LI><LI><A HREF="Tutorial004.html#toc17">Translation</A>
</LI><LI><A HREF="Tutorial004.html#toc18">Translation Tables</A>
</LI><LI><A HREF="Tutorial004.html#toc19">MutableSeq objects</A>
</LI><LI><A HREF="Tutorial004.html#toc20">Working with directly strings</A>
</LI></UL>
</LI><LI><A HREF="Tutorial005.html">Sequence Input/Output</A>
<UL>
<LI><A HREF="Tutorial005.html#toc21">Parsing or Reading Sequences</A>
</LI><LI><A HREF="Tutorial005.html#toc22">Parsing sequences from the net</A>
</LI><LI><A HREF="Tutorial005.html#toc23">Sequence files as Dictionaries</A>
</LI><LI><A HREF="Tutorial005.html#toc24">Writing Sequence Files</A>
</LI></UL>
</LI><LI><A HREF="Tutorial006.html">Sequence Alignment Input/Output</A>
<UL>
<LI><A HREF="Tutorial006.html#toc25">Parsing or Reading Sequence Alignments</A>
</LI><LI><A HREF="Tutorial006.html#toc26">Writing Alignments</A>
</LI></UL>
</LI><LI><A HREF="Tutorial007.html">BLAST</A>
<UL>
<LI><A HREF="Tutorial007.html#toc27">Running BLAST locally</A>
</LI><LI><A HREF="Tutorial007.html#toc28">Running BLAST over the Internet</A>
</LI><LI><A HREF="Tutorial007.html#toc29">Saving BLAST output</A>
</LI><LI><A HREF="Tutorial007.html#toc30">Parsing BLAST output</A>
</LI><LI><A HREF="Tutorial007.html#toc31">The BLAST record class</A>
</LI><LI><A HREF="Tutorial007.html#toc32">Deprecated BLAST parsers</A>
</LI><LI><A HREF="Tutorial007.html#toc33">Dealing with PSI-BLAST</A>
</LI><LI><A HREF="Tutorial007.html#toc34">Dealing with RPS-BLAST</A>
</LI></UL>
</LI><LI><A HREF="Tutorial008.html">Accessing NCBI&#X2019;s Entrez databases</A>
<UL>
<LI><A HREF="Tutorial008.html#toc35">Entrez Guidelines</A>
</LI><LI><A HREF="Tutorial008.html#toc36">EInfo: Obtaining information about the Entrez databases</A>
</LI><LI><A HREF="Tutorial008.html#toc37">ESearch: Searching the Entrez databases</A>
</LI><LI><A HREF="Tutorial008.html#toc38">EPost: Uploading a list of identifiers</A>
</LI><LI><A HREF="Tutorial008.html#toc39">ESummary: Retrieving summaries from primary IDs</A>
</LI><LI><A HREF="Tutorial008.html#toc40">EFetch: Downloading full records from Entrez</A>
</LI><LI><A HREF="Tutorial008.html#toc41">ELink: Searching for related items in NCBI Entrez</A>
</LI><LI><A HREF="Tutorial008.html#toc42">EGQuery: Obtaining counts for search terms</A>
</LI><LI><A HREF="Tutorial008.html#toc43">ESpell: Obtaining spelling suggestions</A>
</LI><LI><A HREF="Tutorial008.html#toc44">Specialized parsers</A>
</LI><LI><A HREF="Tutorial008.html#toc45">Using a proxy</A>
</LI><LI><A HREF="Tutorial008.html#toc46">Examples</A>
</LI><LI><A HREF="Tutorial008.html#toc47">Using the history and WebEnv</A>
</LI></UL>
</LI><LI><A HREF="Tutorial009.html">Swiss-Prot and ExPASy</A>
<UL>
<LI><A HREF="Tutorial009.html#toc48">Parsing Swiss-Prot files</A>
</LI><LI><A HREF="Tutorial009.html#toc49">Parsing Prosite records</A>
</LI><LI><A HREF="Tutorial009.html#toc50">Parsing Prosite documentation records</A>
</LI><LI><A HREF="Tutorial009.html#toc51">Parsing Enzyme records</A>
</LI><LI><A HREF="Tutorial009.html#toc52">Accessing the ExPASy server</A>
</LI><LI><A HREF="Tutorial009.html#toc53">Scanning the Prosite database</A>
</LI></UL>
</LI><LI><A HREF="Tutorial010.html">Going 3D: The PDB module</A>
<UL>
<LI><A HREF="Tutorial010.html#toc54">Structure representation</A>
</LI><LI><A HREF="Tutorial010.html#toc55">Disorder</A>
</LI><LI><A HREF="Tutorial010.html#toc56">Hetero residues</A>
</LI><LI><A HREF="Tutorial010.html#toc57">Some random usage examples</A>
</LI><LI><A HREF="Tutorial010.html#toc58">Common problems in PDB files</A>
</LI><LI><A HREF="Tutorial010.html#toc59">Other features</A>
</LI></UL>
</LI><LI><A HREF="Tutorial011.html">Bio.PopGen: Population genetics</A>
<UL>
<LI><A HREF="Tutorial011.html#toc60">GenePop</A>
</LI><LI><A HREF="Tutorial011.html#toc61">Coalescent simulation</A>
</LI><LI><A HREF="Tutorial011.html#toc62">Other applications</A>
</LI><LI><A HREF="Tutorial011.html#toc63">Future Developments</A>
</LI></UL>
</LI><LI><A HREF="Tutorial012.html">Supervised learning methods</A>
<UL>
<LI><A HREF="Tutorial012.html#toc64">The Logistic Regression Model</A>
</LI><LI><A HREF="Tutorial012.html#toc65"><I>k</I>-Nearest Neighbors</A>
</LI><LI><A HREF="Tutorial012.html#toc66">Na&#XEF;ve Bayes</A>
</LI><LI><A HREF="Tutorial012.html#toc67">Maximum Entropy</A>
</LI><LI><A HREF="Tutorial012.html#toc68">Markov Models</A>
</LI></UL>
</LI><LI><A HREF="Tutorial013.html">Graphics including GenomeDiagram</A>
<UL>
<LI><A HREF="Tutorial013.html#toc69">GenomeDiagram</A>
</LI><LI><A HREF="Tutorial013.html#toc70">Chromosomes</A>
</LI></UL>
</LI><LI><A HREF="Tutorial014.html">Cookbook &#X2013; Cool things to do with it</A>
<UL>
<LI><A HREF="Tutorial014.html#toc71">Sequence parsing plus simple plots</A>
</LI><LI><A HREF="Tutorial014.html#toc72">Dealing with alignments</A>
</LI><LI><A HREF="Tutorial014.html#toc73">Substitution Matrices</A>
</LI><LI><A HREF="Tutorial014.html#toc74">BioSQL &#X2013; storing sequences in a relational database</A>
</LI><LI><A HREF="Tutorial014.html#toc75">InterPro</A>
</LI></UL>
</LI><LI><A HREF="Tutorial015.html">The Biopython testing framework</A>
<UL>
<LI><A HREF="Tutorial015.html#toc76">Running the tests</A>
</LI><LI><A HREF="Tutorial015.html#toc77">Writing tests</A>
</LI><LI><A HREF="Tutorial015.html#toc78">Writing doctests</A>
</LI></UL>
</LI><LI><A HREF="Tutorial016.html">Advanced</A>
<UL>
<LI><A HREF="Tutorial016.html#toc79">The SeqRecord and SeqFeature classes</A>
</LI><LI><A HREF="Tutorial016.html#toc80">Parser Design</A>
</LI><LI><A HREF="Tutorial016.html#toc81">Substitution Matrices</A>
</LI></UL>
</LI><LI><A HREF="Tutorial017.html">Where to go from here &#X2013; contributing to Biopython</A>
<UL>
<LI><A HREF="Tutorial017.html#toc82">Bug Reports + Feature Requests</A>
</LI><LI><A HREF="Tutorial017.html#toc83">Mailing lists and helping newcomers</A>
</LI><LI><A HREF="Tutorial017.html#toc84">Contributing Documentation</A>
</LI><LI><A HREF="Tutorial017.html#toc85">Maintaining a distribution for a platform</A>
</LI><LI><A HREF="Tutorial017.html#toc86">Contributing Unit Tests</A>
</LI><LI><A HREF="Tutorial017.html#toc87">Contributing Code</A>
</LI></UL>
</LI><LI><A HREF="Tutorial018.html">Appendix: Useful stuff about Python</A>
<UL>
<LI><A HREF="Tutorial018.html#toc88">What the heck is a handle?</A>
</LI></UL>
</LI></UL>
<!--FOOTER-->
<HR SIZE=2><BLOCKQUOTE CLASS="quote"><EM>This document was translated from L<sup>A</sup>T<sub>E</sub>X by
</EM><A HREF="http://hevea.inria.fr/index.html"><EM>H</EM><EM><FONT SIZE=2><sup>E</sup></FONT></EM><EM>V</EM><EM><FONT SIZE=2><sup>E</sup></FONT></EM><EM>A</EM></A><EM>.</EM></BLOCKQUOTE></BODY>
</HTML>
