<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN"
            "http://www.w3.org/TR/REC-html40/loose.dtd">
<HTML>
<HEAD>

<META http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
<META name="GENERATOR" content="hevea 1.07">
<TITLE>
Bio.Entrez: Accessing NCBI's Entrez databases
</TITLE>
</HEAD>
<BODY >
<A HREF="Tutorial006.html"><IMG SRC ="previous_motif.gif" ALT="Previous"></A>
<A HREF="index.html"><IMG SRC ="contents_motif.gif" ALT="Up"></A>
<A HREF="Tutorial008.html"><IMG SRC ="next_motif.gif" ALT="Next"></A>
<HR>

<H1><A NAME="htoc52">Chapter&nbsp;6</A>&nbsp;&nbsp;Bio.Entrez: Accessing NCBI's Entrez databases</H1>
<A NAME="chapter:entrez"></A>
Entrez (<A HREF="http://www.ncbi.nlm.nih.gov/Entrez"><TT>http://www.ncbi.nlm.nih.gov/Entrez</TT></A>) is a data retrieval system that provides users access to NCBI's databases such as PubMed, GenBank, GEO, and many others. You can access Entrez from a web browser to manually enter queries, or you can use Biopython's <CODE>Bio.Entrez</CODE> module for programmatic access to Entrez. The latter allows you for example to search PubMed or download GenBank records from within a Python script.<BR>
<BR>
The <CODE>Bio.Entrez</CODE> module makes use of the Entrez Programming Utilities, consisting of eight tools that are described in detail on NCBI's page at <A HREF="http://www.ncbi.nlm.nih.gov/entrez/utils/"><TT>http://www.ncbi.nlm.nih.gov/entrez/utils/</TT></A>. Each of these tools corresponds to one Python function in the <CODE>Bio.Entrez</CODE> module, as described in the sections below. This module makes sure that the correct URL is used for the queries, and that not more than one request is made every three seconds, as required by NCBI.<BR>
<BR>
The output returned by the Entrez Programming Utilities is typically in XML format. Currently, Biopython does not contain parsers for the XML output generated by the Entrez Programming Utilities. However, if you know what you're looking for, it is fairly easy to pull out the information you need from the XML output. For sequence databases, the Entrez Programming Utilities can also generate output in other formats (such as the Fasta and GenBank file format). This can then be parsed into a SeqRecord using <CODE>Bio.SeqIO</CODE> (see Chapter&nbsp;<A HREF="Tutorial005.html#chapter:Bio.SeqIO">4</A>, and the example below).<BR>
<BR>
<A NAME="toc30"></A>
<H2><A NAME="htoc53">6.1</A>&nbsp;&nbsp;EInfo: Obtaining information about the Entrez databases</H2>
EInfo provides field index term counts, last update, and available links for each of NCBI's databases. In addition, you can use EInfo to obtain a list of all database names accessible through the Entrez utilities:
<PRE>
&gt;&gt;&gt; from Bio import Entrez
&gt;&gt;&gt; handle = Entrez.einfo()
&gt;&gt;&gt; result = handle.read()
</PRE>The variable <CODE>result</CODE> now contains a list of databases in XML format:
<PRE>
&gt;&gt;&gt; print result
&lt;?xml version="1.0"?&gt;
&lt;!DOCTYPE eInfoResult PUBLIC "-//NLM//DTD eInfoResult, 11 May 2002//EN" "http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd"&gt;
&lt;eInfoResult&gt;
&lt;DbList&gt;
        &lt;DbName&gt;pubmed&lt;/DbName&gt;
        &lt;DbName&gt;protein&lt;/DbName&gt;
        &lt;DbName&gt;nucleotide&lt;/DbName&gt;
        &lt;DbName&gt;nuccore&lt;/DbName&gt;
        &lt;DbName&gt;nucgss&lt;/DbName&gt;
        &lt;DbName&gt;nucest&lt;/DbName&gt;
        &lt;DbName&gt;structure&lt;/DbName&gt;
        &lt;DbName&gt;genome&lt;/DbName&gt;
        &lt;DbName&gt;books&lt;/DbName&gt;
        &lt;DbName&gt;cancerchromosomes&lt;/DbName&gt;
        &lt;DbName&gt;cdd&lt;/DbName&gt;
        &lt;DbName&gt;gap&lt;/DbName&gt;
        &lt;DbName&gt;domains&lt;/DbName&gt;
        &lt;DbName&gt;gene&lt;/DbName&gt;
        &lt;DbName&gt;genomeprj&lt;/DbName&gt;
        &lt;DbName&gt;gensat&lt;/DbName&gt;
        &lt;DbName&gt;geo&lt;/DbName&gt;
        &lt;DbName&gt;gds&lt;/DbName&gt;
        &lt;DbName&gt;homologene&lt;/DbName&gt;
        &lt;DbName&gt;journals&lt;/DbName&gt;
        &lt;DbName&gt;mesh&lt;/DbName&gt;
        &lt;DbName&gt;ncbisearch&lt;/DbName&gt;
        &lt;DbName&gt;nlmcatalog&lt;/DbName&gt;
        &lt;DbName&gt;omia&lt;/DbName&gt;
        &lt;DbName&gt;omim&lt;/DbName&gt;
        &lt;DbName&gt;pmc&lt;/DbName&gt;
        &lt;DbName&gt;popset&lt;/DbName&gt;
        &lt;DbName&gt;probe&lt;/DbName&gt;
        &lt;DbName&gt;proteinclusters&lt;/DbName&gt;
        &lt;DbName&gt;pcassay&lt;/DbName&gt;
        &lt;DbName&gt;pccompound&lt;/DbName&gt;
        &lt;DbName&gt;pcsubstance&lt;/DbName&gt;
        &lt;DbName&gt;snp&lt;/DbName&gt;
        &lt;DbName&gt;taxonomy&lt;/DbName&gt;
        &lt;DbName&gt;toolkit&lt;/DbName&gt;
        &lt;DbName&gt;unigene&lt;/DbName&gt;
        &lt;DbName&gt;unists&lt;/DbName&gt;
&lt;/DbList&gt;
&lt;/eInfoResult&gt;
</PRE>
For each of these databases, we can use EInfo again to obtain more information:
<PRE>
&gt;&gt;&gt; handle = Entrez.einfo(db="pubmed")
&gt;&gt;&gt; print handle.read()
&lt;?xml version="1.0"?&gt;
&lt;!DOCTYPE eInfoResult PUBLIC "-//NLM//DTD eInfoResult, 11 May 2002//EN" "http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd"&gt;
&lt;eInfoResult&gt;
&lt;DbInfo&gt;
        &lt;DbName&gt;pubmed&lt;/DbName&gt;
        &lt;MenuName&gt;PubMed&lt;/MenuName&gt;
        &lt;Description&gt;PubMed bibliographic record&lt;/Description&gt;
        &lt;Count&gt;17781992&lt;/Count&gt;
        &lt;LastUpdate&gt;2008/02/18 01:22&lt;/LastUpdate&gt;
        &lt;FieldList&gt;
                &lt;Field&gt;
                        &lt;Name&gt;ALL&lt;/Name&gt;
...
</PRE>
<A NAME="toc31"></A>
<H2><A NAME="htoc54">6.2</A>&nbsp;&nbsp;ESearch: Searching the Entrez databases</H2>
To search any of these databases, we use <CODE>Bio.Entrez.esearch()</CODE>. For example, let's search in PubMed for publications related to Biopython:
<PRE>
&gt;&gt;&gt; from Bio import Entrez
&gt;&gt;&gt; handle = Entrez.esearch(db="pubmed", term="biopython")
&gt;&gt;&gt; print handle.read()
&lt;?xml version="1.0"?&gt;
&lt;!DOCTYPE eSearchResult PUBLIC "-//NLM//DTD eSearchResult, 11 May 2002//EN" "http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eSearch_020511.dtd"&gt;
&lt;eSearchResult&gt;
        &lt;Count&gt;5&lt;/Count&gt;
        &lt;RetMax&gt;5&lt;/RetMax&gt;
        &lt;RetStart&gt;0&lt;/RetStart&gt;
        &lt;IdList&gt;
                &lt;Id&gt;16403221&lt;/Id&gt;
                &lt;Id&gt;16377612&lt;/Id&gt;
                &lt;Id&gt;14871861&lt;/Id&gt;
                &lt;Id&gt;14630660&lt;/Id&gt;
                &lt;Id&gt;12230038&lt;/Id&gt;
        &lt;/IdList&gt;
        &lt;TranslationSet&gt;
        &lt;/TranslationSet&gt;
        &lt;TranslationStack&gt;
                &lt;TermSet&gt;
                        &lt;Term&gt;biopython[All Fields]&lt;/Term&gt;
                        &lt;Field&gt;All Fields&lt;/Field&gt;
                        &lt;Count&gt;5&lt;/Count&gt;
                        &lt;Explode&gt;Y&lt;/Explode&gt;
                &lt;/TermSet&gt;
                &lt;OP&gt;GROUP&lt;/OP&gt;
        &lt;/TranslationStack&gt;
        &lt;QueryTranslation&gt;biopython[All Fields]&lt;/QueryTranslation&gt;
&lt;/eSearchResult&gt;
</PRE>In this output, you see five PubMed IDs (16403221, 16377612, 14871861, 14630660, 12230038), which can be retrieved by EFetch (see section <A HREF="#sec:efetch">6.5</A>).<BR>
<BR>
You can also use ESearch to search GenBank. Here we'll do a quick search for the <EM>rpl16</EM> gene in <EM>Opuntia</EM>:
<PRE>
&gt;&gt;&gt; handle = Entrez.esearch(db="nucleotide",term="Opuntia and rpl16")
&gt;&gt;&gt; print handle.read()
&lt;?xml version="1.0"?&gt;
&lt;!DOCTYPE eSearchResult PUBLIC "-//NLM//DTD eSearchResult, 11 May 2002//EN" "http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eSearch_020511.dtd"&gt;
&lt;eSearchResult&gt;
        &lt;Count&gt;9&lt;/Count&gt;
        &lt;RetMax&gt;9&lt;/RetMax&gt;
        &lt;RetStart&gt;0&lt;/RetStart&gt;
        &lt;IdList&gt;
                &lt;Id&gt;57240072&lt;/Id&gt;
                &lt;Id&gt;57240071&lt;/Id&gt;
                &lt;Id&gt;6273287&lt;/Id&gt;
                &lt;Id&gt;6273291&lt;/Id&gt;
                &lt;Id&gt;6273290&lt;/Id&gt;
                &lt;Id&gt;6273289&lt;/Id&gt;
                &lt;Id&gt;6273286&lt;/Id&gt;
                &lt;Id&gt;6273285&lt;/Id&gt;
                &lt;Id&gt;6273284&lt;/Id&gt;
        &lt;/IdList&gt;
        &lt;TranslationSet&gt;
        &lt;/TranslationSet&gt;
        &lt;QueryTranslation&gt;&lt;/QueryTranslation&gt;
&lt;/eSearchResult&gt;
</PRE>
Each of the IDs (<CODE>&lt;Id&gt;57240072&lt;/Id&gt;</CODE>, ...) is a GenBank identifier. See section&nbsp;<A HREF="#sec:efetch">6.5</A> for information on how to actually download these GenBank records.<BR>
<BR>
As a final example, let's get a list of computational journal titles:
<PRE>
&gt;&gt;&gt; handle = Entrez.esearch(db="journals", term="computational")
&gt;&gt;&gt; print handle.read()
&lt;?xml version="1.0"?&gt;
&lt;!DOCTYPE eSearchResult PUBLIC "-//NLM//DTD eSearchResult, 11 May 2002//EN" "http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eSearch_020511.dtd"&gt;
&lt;eSearchResult&gt;
        &lt;Count&gt;15&lt;/Count&gt;
        &lt;RetMax&gt;15&lt;/RetMax&gt;
        &lt;RetStart&gt;0&lt;/RetStart&gt;
        &lt;IdList&gt;
                &lt;Id&gt;30367&lt;/Id&gt;
                &lt;Id&gt;33843&lt;/Id&gt;
                &lt;Id&gt;33823&lt;/Id&gt;
                &lt;Id&gt;32989&lt;/Id&gt;
                &lt;Id&gt;33190&lt;/Id&gt;
                &lt;Id&gt;33009&lt;/Id&gt;
                &lt;Id&gt;31986&lt;/Id&gt;
                &lt;Id&gt;8799&lt;/Id&gt;
                &lt;Id&gt;22857&lt;/Id&gt;
                &lt;Id&gt;32675&lt;/Id&gt;
                &lt;Id&gt;20258&lt;/Id&gt;
                &lt;Id&gt;33859&lt;/Id&gt;
                &lt;Id&gt;32534&lt;/Id&gt;
                &lt;Id&gt;32357&lt;/Id&gt;
                &lt;Id&gt;32249&lt;/Id&gt;
        &lt;/IdList&gt;
        &lt;TranslationSet&gt;
        &lt;/TranslationSet&gt;
        &lt;TranslationStack&gt;
                &lt;TermSet&gt;
                        &lt;Term&gt;computational[All Fields]&lt;/Term&gt;
                        &lt;Field&gt;All Fields&lt;/Field&gt;
                        &lt;Count&gt;15&lt;/Count&gt;
                        &lt;Explode&gt;Y&lt;/Explode&gt;
                &lt;/TermSet&gt;
                &lt;OP&gt;GROUP&lt;/OP&gt;
        &lt;/TranslationStack&gt;
        &lt;QueryTranslation&gt;computational[All Fields]&lt;/QueryTranslation&gt;
&lt;/eSearchResult&gt;
</PRE>Again, we could use EFetch to obtain more information for each of these journal IDs.<BR>
<BR>
ESearch has many useful options --- see the <A HREF="http://www.ncbi.nlm.nih.gov/entrez/query/static/esearch_help.html">ESearch help page</A> for more information.<BR>
<BR>
<A NAME="toc32"></A>
<H2><A NAME="htoc55">6.3</A>&nbsp;&nbsp;EPost</H2>
EPost posts a list of UIs for use in subsequent search strategies; see the <A HREF="http://www.ncbi.nlm.nih.gov/entrez/query/static/epost_help.html">EPost help page</A> for more information. It is available from Biopython through <CODE>Bio.Entrez.epost()</CODE>.<BR>
<BR>
<A NAME="toc33"></A>
<H2><A NAME="htoc56">6.4</A>&nbsp;&nbsp;ESummary: Retrieving summaries from primary IDs</H2>
ESummary retrieves document summaries from a list of primary IDs (see the <A HREF="http://www.ncbi.nlm.nih.gov/entrez/query/static/esummary_help.html">ESummary help page</A> for more information). In Biopython, ESummary is available as <CODE>Bio.Entrez.esummary()</CODE>. Using the search result above, we can for example find out more about the journal with ID 30367:
<PRE>
&gt;&gt;&gt; from Bio import Entrez
&gt;&gt;&gt; handle = Entrez.esummary(db="journals", id="30367")
&gt;&gt;&gt; print handle.read()
&lt;?xml version="1.0"?&gt;
&lt;!DOCTYPE eSummaryResult PUBLIC "-//NLM//DTD eSummaryResult, 29 October 2004//EN" "http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eSummary_041029.dtd"&gt;
&lt;eSummaryResult&gt;
&lt;DocSum&gt;
        &lt;Id&gt;30367&lt;/Id&gt;
        &lt;Item Name="Title" Type="String"&gt;Computational biology and chemistry&lt;/Item&gt;
        &lt;Item Name="MedAbbr" Type="String"&gt;Comput Biol Chem&lt;/Item&gt;
        &lt;Item Name="IsoAbbr" Type="String"&gt;&lt;/Item&gt;
        &lt;Item Name="NlmId" Type="String"&gt;101157394&lt;/Item&gt;
        &lt;Item Name="pISSN" Type="String"&gt;1476-9271&lt;/Item&gt;
        &lt;Item Name="eISSN" Type="String"&gt;&lt;/Item&gt;
        &lt;Item Name="PublicationStartYear" Type="String"&gt;2003&lt;/Item&gt;
        &lt;Item Name="PublicationEndYear" Type="String"&gt;&lt;/Item&gt;
        &lt;Item Name="Publisher" Type="String"&gt;Pergamon,&lt;/Item&gt;
        &lt;Item Name="Language" Type="String"&gt;eng&lt;/Item&gt;
        &lt;Item Name="Country" Type="String"&gt;England&lt;/Item&gt;
        &lt;Item Name="BroadHeading" Type="List"&gt;
                &lt;Item Name="string" Type="String"&gt;Biology&lt;/Item&gt;
                &lt;Item Name="string" Type="String"&gt;Chemistry&lt;/Item&gt;
                &lt;Item Name="string" Type="String"&gt;Medical Informatics&lt;/Item&gt;
        &lt;/Item&gt;
        &lt;Item Name="ContinuationNotes" Type="String"&gt;Continues: Computers &amp;amp; chemistry. &lt;/Item&gt;
&lt;/DocSum&gt;
&lt;/eSummaryResult&gt;
</PRE>
<A NAME="toc34"></A>
<H2><A NAME="htoc57">6.5</A>&nbsp;&nbsp;EFetch: Downloading full records from Entrez</H2>
<A NAME="sec:efetch"></A>
EFetch is what you use when you want to retrieve a full record from Entrez.
For the <EM>Opuntia</EM> example above, we can download GenBank record 57240072 using <CODE>Bio.Entrez.efetch</CODE>:
<PRE>
&gt;&gt;&gt; handle = Entrez.efetch(db="nucleotide", id="57240072",rettype="genbank")
&gt;&gt;&gt; print handle.read()
LOCUS       AY851612                 892 bp    DNA     linear   PLN 10-APR-2007
DEFINITION  Opuntia subulata rpl16 gene, intron; chloroplast.
ACCESSION   AY851612
VERSION     AY851612.1  GI:57240072
KEYWORDS    .
SOURCE      chloroplast Austrocylindropuntia subulata
  ORGANISM  Austrocylindropuntia subulata
            Eukaryota; Viridiplantae; Streptophyta; Embryophyta; Tracheophyta;
            Spermatophyta; Magnoliophyta; eudicotyledons; core eudicotyledons;
            Caryophyllales; Cactaceae; Opuntioideae; Austrocylindropuntia.
REFERENCE   1  (bases 1 to 892)
  AUTHORS   Butterworth,C.A. and Wallace,R.S.
  TITLE     Molecular Phylogenetics of the Leafy Cactus Genus Pereskia
            (Cactaceae)
  JOURNAL   Syst. Bot. 30 (4), 800-808 (2005)
REFERENCE   2  (bases 1 to 892)
  AUTHORS   Butterworth,C.A. and Wallace,R.S.
  TITLE     Direct Submission
  JOURNAL   Submitted (10-DEC-2004) Desert Botanical Garden, 1201 North Galvin
            Parkway, Phoenix, AZ 85008, USA
FEATURES             Location/Qualifiers
     source          1..892
                     /organism="Austrocylindropuntia subulata"
                     /organelle="plastid:chloroplast"
                     /mol_type="genomic DNA"
                     /db_xref="taxon:106982"
     gene            &lt;1..&gt;892
                     /gene="rpl16"
     intron          &lt;1..&gt;892
                     /gene="rpl16"
ORIGIN      
        1 cattaaagaa gggggatgcg gataaatgga aaggcgaaag aaagaaaaaa atgaatctaa
       61 atgatatacg attccactat gtaaggtctt tgaatcatat cataaaagac aatgtaataa
      121 agcatgaata cagattcaca cataattatc tgatatgaat ctattcatag aaaaaagaaa
      181 aaagtaagag cctccggcca ataaagacta agagggttgg ctcaagaaca aagttcatta
      241 agagctccat tgtagaattc agacctaatc attaatcaag aagcgatggg aacgatgtaa
      301 tccatgaata cagaagattc aattgaaaaa gatcctaatg atcattggga aggatggcgg
      361 aacgaaccag agaccaattc atctattctg aaaagtgata aactaatcct ataaaactaa
      421 aatagatatt gaaagagtaa atattcgccc gcgaaaattc cttttttatt aaattgctca
      481 tattttattt tagcaatgca atctaataaa atatatctat acaaaaaaat atagacaaac
      541 tatatatata taatatattt caaatttcct tatataccca aatataaaaa tatctaataa
      601 attagatgaa tatcaaagaa tctattgatt tagtgtatta ttaaatgtat atcttaattc
      661 aatattatta ttctattcat ttttattcat tttcaaattt ataatatatt aatctatata
      721 ttaatttata attctattct aattcgaatt caatttttaa atattcatat tcaattaaaa
      781 ttgaaatttt ttcattcgcg aggagccgga tgagaagaaa ctctcatgtc cggttctgta
      841 gtagagatgg aattaagaaa aaaccatcaa ctataacccc aagagaacca ga
//
</PRE>
The argument <CODE>rettype="genbank"</CODE> lets us download this record in the GenBank format. Alternatively, you could for example use <CODE>rettype="fasta"</CODE> to get the Fasta-format; see the <A HREF="http://www.ncbi.nlm.nih.gov/entrez/query/static/efetchseq_help.html">EFetch Help page</A> for other options. The available formats depend on which database you are downloading from.<BR>
<BR>
If you fetch the record in one of the formats accepted by <CODE>Bio.SeqIO</CODE> (see Chapter&nbsp;<A HREF="Tutorial005.html#chapter:Bio.SeqIO">4</A>), you can directly parse it into a <CODE>SeqRecord</CODE>:
<PRE>
&gt;&gt;&gt; from Bio import Entrez, SeqIO
&gt;&gt;&gt; handle = Entrez.efetch(db="nucleotide", id="57240072",rettype="genbank")
&gt;&gt;&gt; record = SeqIO.read(handle, "genbank")
&gt;&gt;&gt; print record
ID: AY851612.1
Name: AY851612
Desription: Opuntia subulata rpl16 gene, intron; chloroplast.
/sequence_version=1
/source=chloroplast Austrocylindropuntia subulata
/taxonomy=['Eukaryota', 'Viridiplantae', 'Streptophyta', 'Embryophyta', 'Tracheophyta', 'Spermatophyta', 'Magnoliophyta', 'eudicotyledons', 'core eudicotyledons', 'Caryophyllales', 'Cactaceae', 'Opuntioideae', 'Austrocylindropuntia']
/keywords=['']
/references=[&lt;Bio.SeqFeature.Reference instance at 0x141d3a0&gt;, &lt;Bio.SeqFeature.Reference instance at 0x14173a0&gt;]
/accessions=['AY851612']
/data_file_division=PLN
/date=10-APR-2007
/organism=Austrocylindropuntia subulata
/gi=57240072
Seq('CATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGA...AGA', IUPACAmbiguousDNA())
</PRE>
<A NAME="toc35"></A>
<H2><A NAME="htoc58">6.6</A>&nbsp;&nbsp;ELink</H2>
For help on ELink, see the <A HREF="http://www.ncbi.nlm.nih.gov/entrez/query/static/elink_help.html">ELink help page</A>. ELink is available from Biopython through <CODE>Bio.Entrez.elink()</CODE>.<BR>
<BR>
<A NAME="toc36"></A>
<H2><A NAME="htoc59">6.7</A>&nbsp;&nbsp;EGQuery: Obtaining counts for search terms</H2>
EGQuery provides counts for a search term in each of the Entrez databases. In this example, we use <CODE>Bio.Entrez.egquery()</CODE> to obtain the counts for ``Biopython'':
<PRE>
&gt;&gt;&gt; handle = Entrez.egquery(term="biopython") 
&gt;&gt;&gt; print handle.read()
&lt;?xml version="1.0"?&gt;
&lt;!DOCTYPE Result PUBLIC "-//NLM//DTD eSearchResult, January 2004//EN" "http://www.ncbi.nlm.nih.gov/entrez/query/DTD/egquery.dtd"&gt;
&lt;Result&gt;

        &lt;Term&gt;biopython&lt;/Term&gt;

        &lt;eGQueryResult&gt;

             &lt;ResultItem&gt;
                  &lt;DbName&gt;pubmed&lt;/DbName&gt;
                  &lt;MenuName&gt;PubMed&lt;/MenuName&gt;
                  &lt;Count&gt;706&lt;/Count&gt;
                  &lt;Status&gt;Ok&lt;/Status&gt;
             &lt;/ResultItem&gt;

             &lt;ResultItem&gt;
                  &lt;DbName&gt;pmc&lt;/DbName&gt;
                  &lt;MenuName&gt;PMC&lt;/MenuName&gt;
                  &lt;Count&gt;359&lt;/Count&gt;
                  &lt;Status&gt;Ok&lt;/Status&gt;
             &lt;/ResultItem&gt;
...
</PRE>See the <A HREF="http://www.ncbi.nlm.nih.gov/entrez/query/static/egquery_help.html">EGQuery help page</A> for more information.<BR>
<BR>
<A NAME="toc37"></A>
<H2><A NAME="htoc60">6.8</A>&nbsp;&nbsp;ESpell: Obtaining spelling suggestions</H2>
ESpell retrieves spelling suggestions. In this example, we use <CODE>Bio.Entrez.espell()</CODE> to obtain the correct spelling of Biopython:
<PRE>
&gt;&gt;&gt; from Bio import Entrez
&gt;&gt;&gt; handle = Entrez.espell(term="biopythooon")
&gt;&gt;&gt; print handle.read()
&lt;eSpellResult&gt;
        &lt;Database&gt;pubmed&lt;/Database&gt;
        &lt;Query&gt;biopythooon&lt;/Query&gt;
        &lt;CorrectedQuery&gt;biopython&lt;/CorrectedQuery&gt;
        &lt;SpelledQuery&gt;&lt;Replaced&gt;biopython&lt;/Replaced&gt;&lt;/SpelledQuery&gt;
        &lt;ERROR/&gt;
&lt;/eSpellResult&gt;
</PRE>See the <A HREF="http://www.ncbi.nlm.nih.gov/entrez/query/static/espell_help.html">ESpell help page</A> for more information.<BR>
<BR>
<A NAME="toc38"></A>
<H2><A NAME="htoc61">6.9</A>&nbsp;&nbsp;Creating web links to the Entrez databases</H2>
In addition to the eight Entrez Programming Utilities, you can also create URLs to information of the Entrez databases in HTML format. This is primarily intended to create links or bookmarks to the Entrez databases. To do so, you can use the function <CODE>Bio.Entrez.query</CODE>. Detailed information of this service is available from <A HREF="http://www.ncbi.nlm.nih.gov/books/bv.fcgi?rid=helplinks.chapter.linkshelp"><TT>http://www.ncbi.nlm.nih.gov/books/bv.fcgi?rid=helplinks.chapter.linkshelp</TT></A>NCBI. For heavy usage of the NCBI databases, please use the Entrez Programming Utilities instead of <CODE>Bio.Entrez.query</CODE>.<BR>
<BR>
<HR>
<A HREF="Tutorial006.html"><IMG SRC ="previous_motif.gif" ALT="Previous"></A>
<A HREF="index.html"><IMG SRC ="contents_motif.gif" ALT="Up"></A>
<A HREF="Tutorial008.html"><IMG SRC ="next_motif.gif" ALT="Next"></A>
</BODY>
</HTML>
