<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN"
            "http://www.w3.org/TR/REC-html40/loose.dtd">
<HTML>
<HEAD>

<META http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
<META name="GENERATOR" content="hevea 1.07">
<TITLE>
Introduction
</TITLE>
</HEAD>
<BODY >
<A HREF="Tutorial001.html"><IMG SRC ="previous_motif.gif" ALT="Previous"></A>
<A HREF="index.html"><IMG SRC ="contents_motif.gif" ALT="Up"></A>
<A HREF="Tutorial003.html"><IMG SRC ="next_motif.gif" ALT="Next"></A>
<HR>

<H1><A NAME="htoc1">Chapter&nbsp;1</A>&nbsp;&nbsp;Introduction</H1>
<A NAME="chapter:introduction"></A>
<A NAME="toc1"></A>
<H2><A NAME="htoc2">1.1</A>&nbsp;&nbsp;What is Biopython?</H2>
The Biopython Project is an international association of developers of freely available Python (<A HREF="http://www.python.org"><TT>http://www.python.org</TT></A>) tools for computational molecular biology. The web site <A HREF="http://www.biopython.org"><TT>http://www.biopython.org</TT></A> provides an online resource for modules, scripts, and web links for developers of Python-based software for life science research.<BR>
<BR>
Basically, we just like to program in python and want to make it as easy as possible to use python for bioinformatics by creating high-quality, reusable modules and scripts.<BR>
<BR>

<H3><A NAME="htoc3">1.1.1</A>&nbsp;&nbsp;What can I find in the Biopython package</H3>
The main Biopython releases have lots of functionality, including:
<UL><LI>
 The ability to parse bioinformatics files into python utilizable data structures, including support for the following formats:
<UL><LI>
 Blast output -- both from standalone and WWW Blast
 <LI>Clustalw
 <LI>FASTA
 <LI>GenBank
 <LI>PubMed and Medline
 <LI>Expasy files, like Enzyme, Prodoc and Prosite
 <LI>SCOP, including `dom' and `lin' files
 <LI>Rebase
 <LI>UniGene
 <LI>SwissProt
 </UL><BR>
<BR>
<LI>Files in the supported formats can be iterated over record by record or indexed and accessed via a Dictionary interface.<BR>
<BR>
<LI>Code to deal with popular on-line bioinformatics destinations such as:
<UL><LI>
 NCBI -- Blast, Entrez and PubMed services
 <LI>Expasy -- Prodoc and Prosite entries
 </UL><BR>
<BR>
<LI>Interfaces to common bioinformatics programs such as:
<UL><LI>
 Standalone Blast from NCBI
 <LI>Clustalw alignment program.
 </UL><BR>
<BR>
<LI>A standard sequence class that deals with sequences, ids on sequences, and sequence features.<BR>
<BR>
<LI>Tools for performing common operations on sequences, such as translation, transcription and weight calculations.<BR>
<BR>
<LI>Code to perform classification of data using k Nearest Neighbors, Naive Bayes or Support Vector Machines.<BR>
<BR>
<LI>Code for dealing with alignments, including a standard way to create and deal with substitution matrices.<BR>
<BR>
<LI>Code making it easy to split up parallelizable tasks into separate processes.<BR>
<BR>
<LI>GUI-based programs to do basic sequence manipulations, translations, BLASTing, etc.<BR>
<BR>
<LI>Extensive documentation and help with using the modules, including this file, on-line wiki documentation, the web site, and the mailing list.<BR>
<BR>
<LI>Integration with other languages, including the Bioperl and Biojava projects, using the BioCorba interface standard (available with the biopython-corba module).</UL>
We hope this gives you plenty of reasons to download and start using Biopython!<BR>
<BR>
<A NAME="toc2"></A>
<H2><A NAME="htoc4">1.2</A>&nbsp;&nbsp;Installing Biopython</H2>
All of the installation information for Biopython was separated from
this document to make it easier to keep updated. The instructions cover
installation of python, Biopython dependencies and Biopython itself.
It is available in pdf
(<A HREF="http://biopython.org/DIST/docs/install/Installation.pdf"><TT>http://biopython.org/DIST/docs/install/Installation.pdf</TT></A>)
and html formats
(<A HREF="http://biopython.org/DIST/docs/install/Installation.html"><TT>http://biopython.org/DIST/docs/install/Installation.html</TT></A>).<BR>
<BR>
<A NAME="toc3"></A>
<H2><A NAME="htoc5">1.3</A>&nbsp;&nbsp;FAQ</H2>
<OL type=1><LI><EM>Why doesn't</EM> <CODE>Bio.SeqIO</CODE> <EM>work? It imports fine but there is no parse function etc.</EM><BR>
You need Biopython 1.43 or later. Older versions did contain some related code under the <CODE>Bio.SeqIO</CODE> name which has since been deprecated - and this is why the import ``works''.<BR>
<BR>
<LI><EM>Why doesn't</EM> <CODE>Bio.SeqIO.read()</CODE> <EM>work? The module imports fine but there is no read function!</EM><BR>
You need Biopython 1.45 or later.<BR>
<BR>
<LI><EM>Why doesn't</EM> <CODE>Bio.Blast</CODE> <EM>work with the latest plain text NCBI blast output?</EM><BR>
The NCBI keep tweaking the plain text output from the BLAST tools, and keeping our parser up to date is an ongoing struggle. We recommend you use the XML output instead, which is designed to be read by a computer program.<BR>
<BR>
<LI><EM>I looked in a directory for code, but I couldn't seem to find the code that does something. Where's it hidden?</EM><BR>
One thing to know is that we put code in <CODE>__init__.py</CODE> files. If you are not used to looking for code in this file this can be confusing. The reason we do this is to make the imports easier for users. For instance, instead of having to do a ``repetitive'' import like <CODE>from Bio.GenBank import GenBank</CODE>, you can just import like <CODE>from Bio import GenBank</CODE>.</OL>
<HR>
<A HREF="Tutorial001.html"><IMG SRC ="previous_motif.gif" ALT="Previous"></A>
<A HREF="index.html"><IMG SRC ="contents_motif.gif" ALT="Up"></A>
<A HREF="Tutorial003.html"><IMG SRC ="next_motif.gif" ALT="Next"></A>
</BODY>
</HTML>
