<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN"
            "http://www.w3.org/TR/REC-html40/loose.dtd">
<HTML>
<HEAD>

<META http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
<META name="GENERATOR" content="hevea 1.07">
<TITLE>
Table of Contents
</TITLE>
</HEAD>
<BODY >
<A HREF="index.html"><IMG SRC ="contents_motif.gif" ALT="Up"></A>
<A HREF="Tutorial002.html"><IMG SRC ="next_motif.gif" ALT="Next"></A>
<HR>

<H1>Table of Contents</H1>
<UL><LI>
<A HREF="Tutorial002.html#htoc1">Chapter&nbsp;1&nbsp;&nbsp;Introduction</A>
<UL><LI>
<A HREF="Tutorial002.html#htoc2">1.1&nbsp;&nbsp;What is Biopython?</A>
<UL><LI>
<A HREF="Tutorial002.html#htoc3">1.1.1&nbsp;&nbsp;What can I find in the Biopython package</A>
</UL>
<LI><A HREF="Tutorial002.html#htoc4">1.2&nbsp;&nbsp;Installing Biopython</A>
<LI><A HREF="Tutorial002.html#htoc5">1.3&nbsp;&nbsp;FAQ</A>
</UL>
<LI><A HREF="Tutorial003.html#htoc6">Chapter&nbsp;2&nbsp;&nbsp;Quick Start -- What can you do with Biopython?</A>
<UL><LI>
<A HREF="Tutorial003.html#htoc7">2.1&nbsp;&nbsp;General overview of what Biopython provides</A>
<LI><A HREF="Tutorial003.html#htoc8">2.2&nbsp;&nbsp;Working with sequences</A>
<LI><A HREF="Tutorial003.html#htoc9">2.3&nbsp;&nbsp;A usage example</A>
<LI><A HREF="Tutorial003.html#htoc10">2.4&nbsp;&nbsp;Parsing sequence file formats</A>
<UL><LI>
<A HREF="Tutorial003.html#htoc11">2.4.1&nbsp;&nbsp;Simple FASTA parsing example</A>
<LI><A HREF="Tutorial003.html#htoc12">2.4.2&nbsp;&nbsp;Simple GenBank parsing example</A>
<LI><A HREF="Tutorial003.html#htoc13">2.4.3&nbsp;&nbsp;I love parsing -- please don't stop talking about it!</A>
</UL>
<LI><A HREF="Tutorial003.html#htoc14">2.5&nbsp;&nbsp;Connecting with biological databases</A>
<LI><A HREF="Tutorial003.html#htoc15">2.6&nbsp;&nbsp;What to do next</A>
</UL>
<LI><A HREF="Tutorial004.html#htoc16">Chapter&nbsp;3&nbsp;&nbsp;Sequence objects</A>
<UL><LI>
<A HREF="Tutorial004.html#htoc17">3.1&nbsp;&nbsp;Sequences and Alphabets</A>
<LI><A HREF="Tutorial004.html#htoc18">3.2&nbsp;&nbsp;Sequences act like strings</A>
<LI><A HREF="Tutorial004.html#htoc19">3.3&nbsp;&nbsp;Slicing a sequence</A>
<LI><A HREF="Tutorial004.html#htoc20">3.4&nbsp;&nbsp;Turning Seq objects into strings</A>
<LI><A HREF="Tutorial004.html#htoc21">3.5&nbsp;&nbsp;Nucleotide sequences and (reverse) complements</A>
<LI><A HREF="Tutorial004.html#htoc22">3.6&nbsp;&nbsp;Concatenating or adding sequences</A>
<LI><A HREF="Tutorial004.html#htoc23">3.7&nbsp;&nbsp;MutableSeq objects</A>
<LI><A HREF="Tutorial004.html#htoc24">3.8&nbsp;&nbsp;Transcribing and Translation</A>
<LI><A HREF="Tutorial004.html#htoc25">3.9&nbsp;&nbsp;Working with directly strings</A>
</UL>
<LI><A HREF="Tutorial005.html#htoc26">Chapter&nbsp;4&nbsp;&nbsp;Sequence Input/Output</A>
<UL><LI>
<A HREF="Tutorial005.html#htoc27">4.1&nbsp;&nbsp;Parsing or Reading Sequences</A>
<UL><LI>
<A HREF="Tutorial005.html#htoc28">4.1.1&nbsp;&nbsp;Reading Sequence Files</A>
<LI><A HREF="Tutorial005.html#htoc29">4.1.2&nbsp;&nbsp;Iterating over the records in a sequence file</A>
<LI><A HREF="Tutorial005.html#htoc30">4.1.3&nbsp;&nbsp;Getting a list of the records in a sequence file</A>
<LI><A HREF="Tutorial005.html#htoc31">4.1.4&nbsp;&nbsp;Extracting data</A>
</UL>
<LI><A HREF="Tutorial005.html#htoc32">4.2&nbsp;&nbsp;Parsing sequences from the net</A>
<UL><LI>
<A HREF="Tutorial005.html#htoc33">4.2.1&nbsp;&nbsp;Parsing GenBank records from the net</A>
<LI><A HREF="Tutorial005.html#htoc34">4.2.2&nbsp;&nbsp;Parsing SwissProt sequences from the net</A>
</UL>
<LI><A HREF="Tutorial005.html#htoc35">4.3&nbsp;&nbsp;Sequence files as Dictionaries</A>
<UL><LI>
<A HREF="Tutorial005.html#htoc36">4.3.1&nbsp;&nbsp;Specifying the dictionary keys</A>
<LI><A HREF="Tutorial005.html#htoc37">4.3.2&nbsp;&nbsp;Indexing a dictionary using the SEGUID checksum</A>
</UL>
<LI><A HREF="Tutorial005.html#htoc38">4.4&nbsp;&nbsp;Writing Sequence Files</A>
<UL><LI>
<A HREF="Tutorial005.html#htoc39">4.4.1&nbsp;&nbsp;Converting between sequence file formats</A>
<LI><A HREF="Tutorial005.html#htoc40">4.4.2&nbsp;&nbsp;Converting a file of sequences to their reverse complements</A>
</UL>
</UL>
<LI><A HREF="Tutorial006.html#htoc41">Chapter&nbsp;5&nbsp;&nbsp;BLAST</A>
<UL><LI>
<A HREF="Tutorial006.html#htoc42">5.1&nbsp;&nbsp;Running BLAST locally</A>
<LI><A HREF="Tutorial006.html#htoc43">5.2&nbsp;&nbsp;Running BLAST over the Internet</A>
<LI><A HREF="Tutorial006.html#htoc44">5.3&nbsp;&nbsp;Saving BLAST output</A>
<LI><A HREF="Tutorial006.html#htoc45">5.4&nbsp;&nbsp;Parsing BLAST output</A>
<LI><A HREF="Tutorial006.html#htoc46">5.5&nbsp;&nbsp;The BLAST record class</A>
<LI><A HREF="Tutorial006.html#htoc47">5.6&nbsp;&nbsp;Deprecated BLAST parsers</A>
<UL><LI>
<A HREF="Tutorial006.html#htoc48">5.6.1&nbsp;&nbsp;Parsing plain-text BLAST output</A>
<LI><A HREF="Tutorial006.html#htoc49">5.6.2&nbsp;&nbsp;Parsing a file full of BLAST runs</A>
<LI><A HREF="Tutorial006.html#htoc50">5.6.3&nbsp;&nbsp;Finding a bad record somewhere in a huge file</A>
</UL>
<LI><A HREF="Tutorial006.html#htoc51">5.7&nbsp;&nbsp;Dealing with PSIBlast</A>
</UL>
<LI><A HREF="Tutorial007.html#htoc52">Chapter&nbsp;6&nbsp;&nbsp;Bio.Entrez: Accessing NCBI's Entrez databases</A>
<UL><LI>
<A HREF="Tutorial007.html#htoc53">6.1&nbsp;&nbsp;EInfo: Obtaining information about the Entrez databases</A>
<LI><A HREF="Tutorial007.html#htoc54">6.2&nbsp;&nbsp;ESearch: Searching the Entrez databases</A>
<LI><A HREF="Tutorial007.html#htoc55">6.3&nbsp;&nbsp;EPost</A>
<LI><A HREF="Tutorial007.html#htoc56">6.4&nbsp;&nbsp;ESummary: Retrieving summaries from primary IDs</A>
<LI><A HREF="Tutorial007.html#htoc57">6.5&nbsp;&nbsp;EFetch: Downloading full records from Entrez</A>
<LI><A HREF="Tutorial007.html#htoc58">6.6&nbsp;&nbsp;ELink</A>
<LI><A HREF="Tutorial007.html#htoc59">6.7&nbsp;&nbsp;EGQuery: Obtaining counts for search terms</A>
<LI><A HREF="Tutorial007.html#htoc60">6.8&nbsp;&nbsp;ESpell: Obtaining spelling suggestions</A>
<LI><A HREF="Tutorial007.html#htoc61">6.9&nbsp;&nbsp;Creating web links to the Entrez databases</A>
</UL>
<LI><A HREF="Tutorial008.html#htoc62">Chapter&nbsp;7&nbsp;&nbsp;Swiss-Prot, Prosite, Prodoc, and ExPASy</A>
<UL><LI>
<A HREF="Tutorial008.html#htoc63">7.1&nbsp;&nbsp;Bio.SwissProt: Parsing Swiss-Prot records</A>
<LI><A HREF="Tutorial008.html#htoc64">7.2&nbsp;&nbsp;Bio.Prosite: Parsing Prosite records</A>
<LI><A HREF="Tutorial008.html#htoc65">7.3&nbsp;&nbsp;Bio.Prosite.Prodoc: Parsing Prodoc records</A>
<LI><A HREF="Tutorial008.html#htoc66">7.4&nbsp;&nbsp;Bio.ExPASy: Accessing the ExPASy server</A>
<UL><LI>
<A HREF="Tutorial008.html#htoc67">7.4.1&nbsp;&nbsp;Retrieving a Swiss-Prot record</A>
<LI><A HREF="Tutorial008.html#htoc68">7.4.2&nbsp;&nbsp;Searching Swiss-Prot</A>
<LI><A HREF="Tutorial008.html#htoc69">7.4.3&nbsp;&nbsp;Retrieving Prosite and Prodoc records</A>
</UL>
</UL>
<LI><A HREF="Tutorial009.html#htoc70">Chapter&nbsp;8&nbsp;&nbsp;Cookbook -- Cool things to do with it</A>
<UL><LI>
<A HREF="Tutorial009.html#htoc71">8.1&nbsp;&nbsp;PubMed</A>
<UL><LI>
<A HREF="Tutorial009.html#htoc72">8.1.1&nbsp;&nbsp;Sending a query to PubMed</A>
<LI><A HREF="Tutorial009.html#htoc73">8.1.2&nbsp;&nbsp;Retrieving a PubMed record</A>
</UL>
<LI><A HREF="Tutorial009.html#htoc74">8.2&nbsp;&nbsp;GenBank</A>
<UL><LI>
<A HREF="Tutorial009.html#htoc75">8.2.1&nbsp;&nbsp;Retrieving GenBank entries from NCBI</A>
<LI><A HREF="Tutorial009.html#htoc76">8.2.2&nbsp;&nbsp;Parsing GenBank records</A>
<LI><A HREF="Tutorial009.html#htoc77">8.2.3&nbsp;&nbsp;Iterating over GenBank records</A>
<LI><A HREF="Tutorial009.html#htoc78">8.2.4&nbsp;&nbsp;Making your very own GenBank database</A>
</UL>
<LI><A HREF="Tutorial009.html#htoc79">8.3&nbsp;&nbsp;Dealing with alignments</A>
<UL><LI>
<A HREF="Tutorial009.html#htoc80">8.3.1&nbsp;&nbsp;Clustalw</A>
<LI><A HREF="Tutorial009.html#htoc81">8.3.2&nbsp;&nbsp;Calculating summary information</A>
<LI><A HREF="Tutorial009.html#htoc82">8.3.3&nbsp;&nbsp;Calculating a quick consensus sequence</A>
<LI><A HREF="Tutorial009.html#htoc83">8.3.4&nbsp;&nbsp;Position Specific Score Matrices</A>
<LI><A HREF="Tutorial009.html#htoc84">8.3.5&nbsp;&nbsp;Information Content</A>
<LI><A HREF="Tutorial009.html#htoc85">8.3.6&nbsp;&nbsp;Translating between Alignment formats</A>
</UL>
<LI><A HREF="Tutorial009.html#htoc86">8.4&nbsp;&nbsp;Substitution Matrices</A>
<UL><LI>
<A HREF="Tutorial009.html#htoc87">8.4.1&nbsp;&nbsp;Using common substitution matrices</A>
<LI><A HREF="Tutorial009.html#htoc88">8.4.2&nbsp;&nbsp;Creating your own substitution matrix from an alignment</A>
</UL>
<LI><A HREF="Tutorial009.html#htoc89">8.5&nbsp;&nbsp;BioRegistry -- automatically finding sequence sources</A>
<UL><LI>
<A HREF="Tutorial009.html#htoc90">8.5.1&nbsp;&nbsp;Finding resources using a configuration file</A>
<LI><A HREF="Tutorial009.html#htoc91">8.5.2&nbsp;&nbsp;Finding resources through a biopython specific interface</A>
</UL>
<LI><A HREF="Tutorial009.html#htoc92">8.6&nbsp;&nbsp;BioSQL -- storing sequences in a relational database</A>
<LI><A HREF="Tutorial009.html#htoc93">8.7&nbsp;&nbsp;BioCorba</A>
<LI><A HREF="Tutorial009.html#htoc94">8.8&nbsp;&nbsp;Going 3D: The PDB module</A>
<UL><LI>
<A HREF="Tutorial009.html#htoc95">8.8.1&nbsp;&nbsp;Structure representation</A>
<LI><A HREF="Tutorial009.html#htoc96">8.8.2&nbsp;&nbsp;Disorder</A>
<LI><A HREF="Tutorial009.html#htoc97">8.8.3&nbsp;&nbsp;Hetero residues</A>
<LI><A HREF="Tutorial009.html#htoc98">8.8.4&nbsp;&nbsp;Some random usage examples</A>
<LI><A HREF="Tutorial009.html#htoc99">8.8.5&nbsp;&nbsp;Common problems in PDB files</A>
<LI><A HREF="Tutorial009.html#htoc100">8.8.6&nbsp;&nbsp;Other features</A>
</UL>
<LI><A HREF="Tutorial009.html#htoc101">8.9&nbsp;&nbsp;Bio.PopGen: Population genetics</A>
<UL><LI>
<A HREF="Tutorial009.html#htoc102">8.9.1&nbsp;&nbsp;GenePop</A>
<LI><A HREF="Tutorial009.html#htoc103">8.9.2&nbsp;&nbsp;Coalescent simulation</A>
<LI><A HREF="Tutorial009.html#htoc104">8.9.3&nbsp;&nbsp;Other applications</A>
<LI><A HREF="Tutorial009.html#htoc105">8.9.4&nbsp;&nbsp;Future Developments</A>
</UL>
<LI><A HREF="Tutorial009.html#htoc106">8.10&nbsp;&nbsp;InterPro</A>
</UL>
<LI><A HREF="Tutorial010.html#htoc107">Chapter&nbsp;9&nbsp;&nbsp;Advanced</A>
<UL><LI>
<A HREF="Tutorial010.html#htoc108">9.1&nbsp;&nbsp;The SeqRecord and SeqFeature classes</A>
<UL><LI>
<A HREF="Tutorial010.html#htoc109">9.1.1&nbsp;&nbsp;Sequence ids and Descriptions -- dealing with SeqRecords</A>
<LI><A HREF="Tutorial010.html#htoc110">9.1.2&nbsp;&nbsp;Features and Annotations -- SeqFeatures</A>
</UL>
<LI><A HREF="Tutorial010.html#htoc111">9.2&nbsp;&nbsp;Regression Testing Framework</A>
<UL><LI>
<A HREF="Tutorial010.html#htoc112">9.2.1&nbsp;&nbsp;Writing a Regression Test</A>
</UL>
<LI><A HREF="Tutorial010.html#htoc113">9.3&nbsp;&nbsp;Parser Design</A>
<UL><LI>
<A HREF="Tutorial010.html#htoc114">9.3.1&nbsp;&nbsp;Design Overview</A>
<LI><A HREF="Tutorial010.html#htoc115">9.3.2&nbsp;&nbsp;Events</A>
<LI><A HREF="Tutorial010.html#htoc116">9.3.3&nbsp;&nbsp;`noevent' EVENT</A>
<LI><A HREF="Tutorial010.html#htoc117">9.3.4&nbsp;&nbsp;Scanners</A>
<LI><A HREF="Tutorial010.html#htoc118">9.3.5&nbsp;&nbsp;Consumers</A>
<LI><A HREF="Tutorial010.html#htoc119">9.3.6&nbsp;&nbsp;BLAST</A>
<LI><A HREF="Tutorial010.html#htoc120">9.3.7&nbsp;&nbsp;Enzyme</A>
<LI><A HREF="Tutorial010.html#htoc121">9.3.8&nbsp;&nbsp;KEGG</A>
<LI><A HREF="Tutorial010.html#htoc122">9.3.9&nbsp;&nbsp;Fasta</A>
<LI><A HREF="Tutorial010.html#htoc123">9.3.10&nbsp;&nbsp;Medline</A>
<LI><A HREF="Tutorial010.html#htoc124">9.3.11&nbsp;&nbsp;Prosite</A>
<LI><A HREF="Tutorial010.html#htoc125">9.3.12&nbsp;&nbsp;SWISS-PROT</A>
<LI><A HREF="Tutorial010.html#htoc126">9.3.13&nbsp;&nbsp;NBRF</A>
<LI><A HREF="Tutorial010.html#htoc127">9.3.14&nbsp;&nbsp;Ndb</A>
<LI><A HREF="Tutorial010.html#htoc128">9.3.15&nbsp;&nbsp;MetaTool</A>
</UL>
<LI><A HREF="Tutorial010.html#htoc129">9.4&nbsp;&nbsp;Substitution Matrices</A>
<UL><LI>
<A HREF="Tutorial010.html#htoc130">9.4.1&nbsp;&nbsp;SubsMat</A>
<LI><A HREF="Tutorial010.html#htoc131">9.4.2&nbsp;&nbsp;FreqTable</A>
</UL>
</UL>
<LI><A HREF="Tutorial011.html#htoc132">Chapter&nbsp;10&nbsp;&nbsp;Where to go from here -- contributing to Biopython</A>
<UL><LI>
<A HREF="Tutorial011.html#htoc133">10.1&nbsp;&nbsp;Maintaining a distribution for a platform</A>
<LI><A HREF="Tutorial011.html#htoc134">10.2&nbsp;&nbsp;Bug Reports + Feature Requests</A>
<LI><A HREF="Tutorial011.html#htoc135">10.3&nbsp;&nbsp;Contributing Code</A>
</UL>
<LI><A HREF="Tutorial012.html#htoc136">Chapter&nbsp;11&nbsp;&nbsp;Appendix: Useful stuff about Python</A>
<UL><LI>
<A HREF="Tutorial012.html#htoc137">11.1&nbsp;&nbsp;What the heck is a handle?</A>
<UL><LI>
<A HREF="Tutorial012.html#htoc138">11.1.1&nbsp;&nbsp;Creating a handle from a string</A>
</UL>
</UL>
</UL>

<HR>
<A HREF="index.html"><IMG SRC ="contents_motif.gif" ALT="Up"></A>
<A HREF="Tutorial002.html"><IMG SRC ="next_motif.gif" ALT="Next"></A>
</BODY>
</HTML>
