<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN"
            "http://www.w3.org/TR/REC-html40/loose.dtd">
<HTML>
<HEAD>

<META http-equiv="Content-Type" content="text/html; charset=US-ASCII">
<META name="GENERATOR" content="hevea 1.10">
<LINK rel="stylesheet" type="text/css" href="Tutorial.css">
<TITLE>Contents</TITLE>
</HEAD>
<BODY >
<A HREF="index.html"><IMG SRC="contents_motif.gif" ALT="Up"></A>
<A HREF="Tutorial002.html"><IMG SRC="next_motif.gif" ALT="Next"></A>
<HR>
<H1 CLASS="chapter">Contents</H1><UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial002.html#htoc1">Chapter&#XA0;1&#XA0;&#XA0;Introduction</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial002.html#htoc2">1.1&#XA0;&#XA0;What is Biopython?</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial002.html#htoc3">1.1.1&#XA0;&#XA0;What can I find in the Biopython package</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial002.html#htoc4">1.2&#XA0;&#XA0;Installing Biopython</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial002.html#htoc5">1.3&#XA0;&#XA0;FAQ</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial003.html#htoc6">Chapter&#XA0;2&#XA0;&#XA0;Quick Start &#X2013; What can you do with Biopython?</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial003.html#htoc7">2.1&#XA0;&#XA0;General overview of what Biopython provides</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial003.html#htoc8">2.2&#XA0;&#XA0;Working with sequences</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial003.html#htoc9">2.3&#XA0;&#XA0;A usage example</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial003.html#htoc10">2.4&#XA0;&#XA0;Parsing sequence file formats</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial003.html#htoc11">2.4.1&#XA0;&#XA0;Simple FASTA parsing example</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial003.html#htoc12">2.4.2&#XA0;&#XA0;Simple GenBank parsing example</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial003.html#htoc13">2.4.3&#XA0;&#XA0;I love parsing &#X2013; please don&#X2019;t stop talking about it!</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial003.html#htoc14">2.5&#XA0;&#XA0;Connecting with biological databases</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial003.html#htoc15">2.6&#XA0;&#XA0;What to do next</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial004.html#htoc16">Chapter&#XA0;3&#XA0;&#XA0;Sequence objects</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial004.html#htoc17">3.1&#XA0;&#XA0;Sequences and Alphabets</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial004.html#htoc18">3.2&#XA0;&#XA0;Sequences act like strings</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial004.html#htoc19">3.3&#XA0;&#XA0;Slicing a sequence</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial004.html#htoc20">3.4&#XA0;&#XA0;Turning Seq objects into strings</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial004.html#htoc21">3.5&#XA0;&#XA0;Nucleotide sequences and (reverse) complements</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial004.html#htoc22">3.6&#XA0;&#XA0;Concatenating or adding sequences</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial004.html#htoc23">3.7&#XA0;&#XA0;MutableSeq objects</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial004.html#htoc24">3.8&#XA0;&#XA0;Transcribing and Translation</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial004.html#htoc25">3.9&#XA0;&#XA0;Working with directly strings</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial005.html#htoc26">Chapter&#XA0;4&#XA0;&#XA0;Sequence Input/Output</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial005.html#htoc27">4.1&#XA0;&#XA0;Parsing or Reading Sequences</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial005.html#htoc28">4.1.1&#XA0;&#XA0;Reading Sequence Files</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial005.html#htoc29">4.1.2&#XA0;&#XA0;Iterating over the records in a sequence file</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial005.html#htoc30">4.1.3&#XA0;&#XA0;Getting a list of the records in a sequence file</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial005.html#htoc31">4.1.4&#XA0;&#XA0;Extracting data</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial005.html#htoc32">4.2&#XA0;&#XA0;Parsing sequences from the net</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial005.html#htoc33">4.2.1&#XA0;&#XA0;Parsing GenBank records from the net</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial005.html#htoc34">4.2.2&#XA0;&#XA0;Parsing SwissProt sequences from the net</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial005.html#htoc35">4.3&#XA0;&#XA0;Sequence files as Dictionaries</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial005.html#htoc36">4.3.1&#XA0;&#XA0;Specifying the dictionary keys</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial005.html#htoc37">4.3.2&#XA0;&#XA0;Indexing a dictionary using the SEGUID checksum</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial005.html#htoc38">4.4&#XA0;&#XA0;Writing Sequence Files</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial005.html#htoc39">4.4.1&#XA0;&#XA0;Converting between sequence file formats</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial005.html#htoc40">4.4.2&#XA0;&#XA0;Converting a file of sequences to their reverse complements</A>
</LI></UL>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial006.html#htoc41">Chapter&#XA0;5&#XA0;&#XA0;Sequence Alignment Input/Output</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial006.html#htoc42">5.1&#XA0;&#XA0;Parsing or Reading Sequence Alignments</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial006.html#htoc43">5.1.1&#XA0;&#XA0;Single Alignments</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial006.html#htoc44">5.1.2&#XA0;&#XA0;Multiple Alignments</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial006.html#htoc45">5.1.3&#XA0;&#XA0;Ambiguous Alignments</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial006.html#htoc46">5.2&#XA0;&#XA0;Writing Alignments</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial006.html#htoc47">5.2.1&#XA0;&#XA0;Converting between sequence alignment file formats</A>
</LI></UL>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial007.html#htoc48">Chapter&#XA0;6&#XA0;&#XA0;BLAST</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial007.html#htoc49">6.1&#XA0;&#XA0;Running BLAST locally</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial007.html#htoc50">6.2&#XA0;&#XA0;Running BLAST over the Internet</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial007.html#htoc51">6.3&#XA0;&#XA0;Saving BLAST output</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial007.html#htoc52">6.4&#XA0;&#XA0;Parsing BLAST output</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial007.html#htoc53">6.5&#XA0;&#XA0;The BLAST record class</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial007.html#htoc54">6.6&#XA0;&#XA0;Deprecated BLAST parsers</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial007.html#htoc55">6.6.1&#XA0;&#XA0;Parsing plain-text BLAST output</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial007.html#htoc56">6.6.2&#XA0;&#XA0;Parsing a file full of BLAST runs</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial007.html#htoc57">6.6.3&#XA0;&#XA0;Finding a bad record somewhere in a huge file</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial007.html#htoc58">6.7&#XA0;&#XA0;Dealing with PSIBlast</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial008.html#htoc59">Chapter&#XA0;7&#XA0;&#XA0;Accessing NCBI&#X2019;s Entrez databases</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial008.html#htoc60">7.1&#XA0;&#XA0;Entrez Guidelines</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial008.html#htoc61">7.2&#XA0;&#XA0;EInfo: Obtaining information about the Entrez databases</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial008.html#htoc62">7.3&#XA0;&#XA0;ESearch: Searching the Entrez databases</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial008.html#htoc63">7.4&#XA0;&#XA0;EPost</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial008.html#htoc64">7.5&#XA0;&#XA0;ESummary: Retrieving summaries from primary IDs</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial008.html#htoc65">7.6&#XA0;&#XA0;EFetch: Downloading full records from Entrez</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial008.html#htoc66">7.7&#XA0;&#XA0;ELink</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial008.html#htoc67">7.8&#XA0;&#XA0;EGQuery: Obtaining counts for search terms</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial008.html#htoc68">7.9&#XA0;&#XA0;ESpell: Obtaining spelling suggestions</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial008.html#htoc69">7.10&#XA0;&#XA0;Examples</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial008.html#htoc70">7.10.1&#XA0;&#XA0;Searching and downloading Entrez Nucleotide records</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial008.html#htoc71">7.10.2&#XA0;&#XA0;Finding the lineage of an organism</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial008.html#htoc72">7.10.3&#XA0;&#XA0;Using the history and WebEnv</A>
</LI></UL>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial009.html#htoc73">Chapter&#XA0;8&#XA0;&#XA0;Swiss-Prot, Prosite, Prodoc, and ExPASy</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial009.html#htoc74">8.1&#XA0;&#XA0;Bio.SwissProt: Parsing Swiss-Prot files</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial009.html#htoc75">8.1.1&#XA0;&#XA0;Parsing Swiss-Prot records</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial009.html#htoc76">8.1.2&#XA0;&#XA0;Parsing the Swiss-Prot keyword and category list</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial009.html#htoc77">8.2&#XA0;&#XA0;Bio.Prosite: Parsing Prosite records</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial009.html#htoc78">8.3&#XA0;&#XA0;Bio.Prosite.Prodoc: Parsing Prodoc records</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial009.html#htoc79">8.4&#XA0;&#XA0;Bio.ExPASy: Accessing the ExPASy server</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial009.html#htoc80">8.4.1&#XA0;&#XA0;Retrieving a Swiss-Prot record</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial009.html#htoc81">8.4.2&#XA0;&#XA0;Searching Swiss-Prot</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial009.html#htoc82">8.4.3&#XA0;&#XA0;Retrieving Prosite and Prodoc records</A>
</LI></UL>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc83">Chapter&#XA0;9&#XA0;&#XA0;Cookbook &#X2013; Cool things to do with it</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial010.html#htoc84">9.1&#XA0;&#XA0;PubMed</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial010.html#htoc85">9.1.1&#XA0;&#XA0;Sending a query to PubMed</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc86">9.1.2&#XA0;&#XA0;Retrieving a PubMed record</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc87">9.2&#XA0;&#XA0;GenBank</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial010.html#htoc88">9.2.1&#XA0;&#XA0;Retrieving GenBank entries from NCBI</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc89">9.2.2&#XA0;&#XA0;Parsing GenBank records</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc90">9.2.3&#XA0;&#XA0;Iterating over GenBank records</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc91">9.3&#XA0;&#XA0;Dealing with alignments</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial010.html#htoc92">9.3.1&#XA0;&#XA0;Clustalw</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc93">9.3.2&#XA0;&#XA0;Calculating summary information</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc94">9.3.3&#XA0;&#XA0;Calculating a quick consensus sequence</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc95">9.3.4&#XA0;&#XA0;Position Specific Score Matrices</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc96">9.3.5&#XA0;&#XA0;Information Content</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc97">9.3.6&#XA0;&#XA0;Translating between Alignment formats</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc98">9.4&#XA0;&#XA0;Substitution Matrices</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial010.html#htoc99">9.4.1&#XA0;&#XA0;Using common substitution matrices</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc100">9.4.2&#XA0;&#XA0;Creating your own substitution matrix from an alignment</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc101">9.5&#XA0;&#XA0;BioSQL &#X2013; storing sequences in a relational database</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc102">9.6&#XA0;&#XA0;BioCorba</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc103">9.7&#XA0;&#XA0;Going 3D: The PDB module</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial010.html#htoc104">9.7.1&#XA0;&#XA0;Structure representation</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc105">9.7.2&#XA0;&#XA0;Disorder</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc106">9.7.3&#XA0;&#XA0;Hetero residues</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc107">9.7.4&#XA0;&#XA0;Some random usage examples</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc108">9.7.5&#XA0;&#XA0;Common problems in PDB files</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc109">9.7.6&#XA0;&#XA0;Other features</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc110">9.8&#XA0;&#XA0;Bio.PopGen: Population genetics</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial010.html#htoc111">9.8.1&#XA0;&#XA0;GenePop</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc112">9.8.2&#XA0;&#XA0;Coalescent simulation</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc113">9.8.3&#XA0;&#XA0;Other applications</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc114">9.8.4&#XA0;&#XA0;Future Developments</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial010.html#htoc115">9.9&#XA0;&#XA0;InterPro</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial011.html#htoc116">Chapter&#XA0;10&#XA0;&#XA0;Advanced</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial011.html#htoc117">10.1&#XA0;&#XA0;The SeqRecord and SeqFeature classes</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial011.html#htoc118">10.1.1&#XA0;&#XA0;Sequence ids and Descriptions &#X2013; dealing with SeqRecords</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial011.html#htoc119">10.1.2&#XA0;&#XA0;Features and Annotations &#X2013; SeqFeatures</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial011.html#htoc120">10.2&#XA0;&#XA0;Regression Testing Framework</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial011.html#htoc121">10.2.1&#XA0;&#XA0;Writing a Regression Test</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial011.html#htoc122">10.3&#XA0;&#XA0;Parser Design</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial011.html#htoc123">10.4&#XA0;&#XA0;Substitution Matrices</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial011.html#htoc124">10.4.1&#XA0;&#XA0;SubsMat</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial011.html#htoc125">10.4.2&#XA0;&#XA0;FreqTable</A>
</LI></UL>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial012.html#htoc126">Chapter&#XA0;11&#XA0;&#XA0;Where to go from here &#X2013; contributing to Biopython</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial012.html#htoc127">11.1&#XA0;&#XA0;Maintaining a distribution for a platform</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial012.html#htoc128">11.2&#XA0;&#XA0;Bug Reports + Feature Requests</A>
</LI><LI CLASS="li-toc"><A HREF="Tutorial012.html#htoc129">11.3&#XA0;&#XA0;Contributing Code</A>
</LI></UL>
</LI><LI CLASS="li-toc"><A HREF="Tutorial013.html#htoc130">Chapter&#XA0;12&#XA0;&#XA0;Appendix: Useful stuff about Python</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial013.html#htoc131">12.1&#XA0;&#XA0;What the heck is a handle?</A>
<UL CLASS="toc"><LI CLASS="li-toc">
<A HREF="Tutorial013.html#htoc132">12.1.1&#XA0;&#XA0;Creating a handle from a string</A>
</LI></UL>
</LI></UL>
</LI></UL><HR>
<A HREF="index.html"><IMG SRC="contents_motif.gif" ALT="Up"></A>
<A HREF="Tutorial002.html"><IMG SRC="next_motif.gif" ALT="Next"></A>
</BODY>
</HTML>
