<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>GEMINI: a flexible framework for exploring genome variation</title>
    
    <link rel="stylesheet" href="_static/labibi.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.16.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="shortcut icon" href="_static/gemini.png"/>
    <link rel="top" title="None" href="#" />
    <link rel="next" title="Installation" href="content/installation.html" />


<script src="_static/labibi.js"></script>



<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', '']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>


  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="content/installation.html" title="Installation"
             accesskey="N">next</a> |</li>
        <li><a href="#">gemini v0.16.3</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  
  <div class="section" id="gemini-a-flexible-framework-for-exploring-genome-variation">
<h1><strong>GEMINI</strong>: <em>a flexible framework for exploring genome variation</em><a class="headerlink" href="#gemini-a-flexible-framework-for-exploring-genome-variation" title="Permalink to this headline">¶</a></h1>
<a class="reference internal image-reference" href="_images/overview.png"><img alt="_images/overview.png" class="align-center" src="_images/overview.png" style="width: 400pt;" /></a>
</div>
<div class="section" id="latest-news">
<h1>Latest news<a class="headerlink" href="#latest-news" title="Permalink to this headline">¶</a></h1>
<div class="section" id="changes-to-inheritance-tools">
<h2>Changes to Inheritance Tools<a class="headerlink" href="#changes-to-inheritance-tools" title="Permalink to this headline">¶</a></h2>
<p>As of version 0.16.0, the built-in Mendelian inheritance tools are more stringent by default (they can be relaxed with the <tt class="docutils literal"><span class="pre">--lenient</span></tt>) option. By default, samples with unknown phenotype will not affect the tools, and
strict requirements are placed on family structure. See the the <a class="reference internal" href="content/tools.html"><em>docs</em></a> for
more info. In addition, the inheritance tools now support multi-generational pedigrees.</p>
</div>
<div class="section" id="much-faster-genotype-queries">
<h2>Much faster genotype queries<a class="headerlink" href="#much-faster-genotype-queries" title="Permalink to this headline">¶</a></h2>
<p>As of version 0.15.0, GEMINI creates auxilliary index files
using the <cite>bcolz</cite> library. This allows queries that filter
on sample genotype information via the <tt class="docutils literal"><span class="pre">--gt-filter</span></tt> option
to be <strong>up to 1000 times faster than identical queries using
version 0.14.0 or earlier</strong>. Details of the implementation and caveats can be found <a class="reference internal" href="content/genotype_query_engines.html"><em>here</em></a>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>In order to expedite your queries with the new bcolz indexing strategy, one must invoke the <tt class="docutils literal"><span class="pre">--use-bcolz</span></tt>
option. We will likely phase this option out over time.
For example:</p>
<div class="last highlight-python"><div class="highlight"><pre>gemini query \
  -q &quot;select variant_id, gts.kid from variants&quot;  \
  --gt-filter &quot;gt_types.mom == HET and gt_types.dad == HET and gts.kid != &#39;HET&#39;&quot; \
  --use-bcolz \
foo.db
</pre></div>
</div>
</div>
</div>
<div class="section" id="new-gemini-workflow">
<h2>New GEMINI Workflow<a class="headerlink" href="#new-gemini-workflow" title="Permalink to this headline">¶</a></h2>
<p>As version 0.12.2 of GEMINI it is required that your input VCF file undergo additional preprocessing such that multi-allelic variants are decomposed and normalized using the <a class="reference external" href="http://genome.sph.umich.edu/wiki/Vt">vt</a> toolset from
the <a class="reference external" href="http://genome.sph.umich.edu/wiki/Main_Page">Abecasis lab</a>. Note that we have also decomposed and normalized all of the VCF-based annotation files (e.g., ExAC, dbSNP, ClinVar, etc.) so that variants and alleles are properly annotated and we minimize false negative and false positive annotations. For a great discussion of why this is necessary, please read <a class="reference external" href="http://www.cureffi.org/2014/04/24/converting-genetic-variants-to-their-minimal-representation/">this blog post</a> from Eric Minikel in Daniel MacArthur&#8217;s lab.</p>
<p>Essentially, VCF preprocessing for GEMINI now boils down to the following steps.</p>
<ol class="arabic simple" start="0">
<li>If working with GATK VCFs, you need to correct the AD INFO tag definition to play nicely with <cite>vt</cite>.</li>
<li><a class="reference external" href="http://genome.sph.umich.edu/wiki/Vt#Decompose">Decompose</a> the original VCF such that variants with multiple alleles are expanded into distinct variant records; one record for each REF/ALT combination.</li>
<li><a class="reference external" href="http://genome.sph.umich.edu/wiki/Vt#Normalization">Normalize</a> the decomposed VCF so that variants are left aligned and represented using the most parsimonious alleles.</li>
<li>Annotate with VEP or snpEff.</li>
<li>bgzip and tabix.</li>
</ol>
<p>A workflow for the above steps is given below.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c"># setup</span>
<span class="nv">VCF</span><span class="o">=</span>/path/to/my.vcf
<span class="nv">NORMVCF</span><span class="o">=</span>/path/to/my.norm.vcf
<span class="nv">REF</span><span class="o">=</span>/path/to/human.b37.fasta
<span class="nv">SNPEFFJAR</span><span class="o">=</span>/path/to/snpEff.jar

<span class="c"># decompose, normalize and annotate VCF with snpEff.</span>
<span class="c"># NOTE: can also swap snpEff with VEP</span>
<span class="c">#NOTE: -classic and -formatEff flags needed with snpEff &gt;= v4.1</span>
zless <span class="nv">$VCF</span> <span class="se">\</span>
   | sed <span class="s1">&#39;s/ID=AD,Number=./ID=AD,Number=R/&#39;</span>
   | vt decompose -s - <span class="se">\</span>
   | vt normalize -r <span class="nv">$REF</span> - <span class="se">\</span>
   | java -Xmx4G -jar <span class="nv">$SNPEFFJAR</span> -formatEff -classic GRCh37.75  <span class="se">\</span>
   | bgzip -c &gt; <span class="nv">$NORMVCF</span>
tabix <span class="nv">$NORMVCF</span>

<span class="c"># load the pre-processed VCF into GEMINI</span>
gemini load --cores 3 -t snpEff -v <span class="nv">$NORMVCF</span> <span class="nv">$db</span>

<span class="c"># query away</span>
gemini query -q <span class="s2">&quot;select chrom, start, end, ref, alt, (gts).(*) from variants&quot;</span> <span class="se">\</span>
             --gt-filter <span class="s2">&quot;gt_types.mom == HET and \</span>
<span class="s2">                          gt_types.dad == HET and \</span>
<span class="s2">                          gt_types.kid == HOM_ALT&quot;</span> <span class="se">\</span>
             <span class="nv">$db</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="overview">
<h1>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h1>
<p>GEMINI (GEnome MINIng) is designed to be a flexible framework for exploring genetic variation
in the context of the wealth of genome annotations available for the human genome.
By placing genetic variants, sample genotypes, and useful genome annotations into
an integrated database framework, <tt class="docutils literal"><span class="pre">GEMINI</span></tt> provides a simple, flexible, yet
very powerful system for exploring genetic variation for disease and
population genetics.</p>
<p>Using the GEMINI framework begins by loading a VCF file (and an optional PED file) into a
database.  Each variant is automatically annotated by comparing it to several
genome annotations from source such as ENCODE tracks, UCSC tracks, OMIM, dbSNP,
KEGG, and HPRD.  All of this information is stored in portable
SQLite database that allows one to explore and interpret both coding and
non-coding variation using &#8220;off-the-shelf&#8221; tools or an enhanced SQL engine.</p>
<p>Please also see the original <a class="reference external" href="http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003153">manuscript</a>.</p>
<p>This <a class="reference external" href="http://www.youtube.com/watch?v=p-UWmDG6yj4&amp;t=0m40s">video</a> provides more details about GEMINI&#8217;s aims and
utility.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ol class="last arabic simple">
<li>GEMINI solely supports human genetic variation mapped to build 37 (aka hg19) of the human genome.</li>
<li>GEMINI is very strict about adherence to VCF format 4.1.</li>
<li>For best performance, load and query GEMINI databases on the fastest hard drive to which you have access.</li>
</ol>
</div>
</div>
<div class="section" id="citation">
<h1>Citation<a class="headerlink" href="#citation" title="Permalink to this headline">¶</a></h1>
<p>If you use GEMINI in your research, please cite the following manuscript:</p>
<div class="highlight-python"><div class="highlight"><pre>Paila U, Chapman BA, Kirchner R, Quinlan AR (2013)
GEMINI: Integrative Exploration of Genetic Variation and Genome Annotations.
PLoS Comput Biol 9(7): e1003153. doi:10.1371/journal.pcbi.1003153
</pre></div>
</div>
</div>
<div class="section" id="table-of-contents">
<h1>Table of contents<a class="headerlink" href="#table-of-contents" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="content/installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/installation.html#automated-installation">Automated installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/installation.html#software-dependencies">Software dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/installation.html#manual-installation">Manual installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/installation.html#installing-annotation-files">Installing annotation files</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/installation.html#using-previously-installed-annotation-files">Using previously installed annotation files</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/installation.html#running-the-testing-suite">Running the testing suite</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/quick_start.html">Quick start</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/quick_start.html#importing-vcf-files-into-gemini">Importing VCF files into gemini.</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/quick_start.html#querying-the-gemini-database">Querying the <cite>gemini</cite> database.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/functional_annotation.html">Annotation with snpEff or VEP</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/functional_annotation.html#stepwise-installation-and-usage-of-vep">Stepwise installation and usage of VEP</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/functional_annotation.html#stepwise-installation-and-usage-of-snpeff">Stepwise installation and usage of SnpEff</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/functional_annotation.html#columns-populated-by-snpeff-vep-tools">Columns populated by snpEff/VEP tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/functional_annotation.html#standardizing-impact-definitions-for-gemini">Standardizing <tt class="docutils literal"><span class="pre">impact</span></tt> definitions for GEMINI</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/functional_annotation.html#so-impact-definitions-in-gemini">SO <tt class="docutils literal"><span class="pre">impact</span></tt> definitions in GEMINI</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/preprocessing.html">Preprocessing and Loading a VCF file into GEMINI</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/preprocessing.html#step-1-split-left-align-and-trim-variants">Step 1. split, left-align, and trim variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/preprocessing.html#step-2-annotate-with-snpeff-or-vep">Step 2. Annotate with snpEff or VEP</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/preprocessing.html#the-basics">The basics</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/preprocessing.html#using-multiple-cpus-for-loading">Using multiple CPUs for loading</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/preprocessing.html#using-lsf-sge-slurm-and-torque-schedulers">Using LSF, SGE, SLURM and Torque schedulers</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/preprocessing.html#describing-samples-with-a-ped-file">Describing samples with a PED file</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/preprocessing.html#load-gerp-base-pair-conservation-scores">Load GERP base pair conservation scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/preprocessing.html#load-cadd-scores-for-deleterious-variants">Load CADD scores for deleterious variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/preprocessing.html#loading-vcfs-without-genotypes">Loading VCFs without genotypes.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/querying.html">Querying the GEMINI database</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#basic-queries">Basic queries</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#selecting-sample-genotypes">Selecting sample genotypes</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#selecting-sample-genotypes-based-on-wildcards">Selecting sample genotypes based on &#8220;wildcards&#8221;.</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#gt-filter-filtering-on-genotypes"><tt class="docutils literal"><span class="pre">--gt-filter</span></tt> Filtering on genotypes</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#gt-filter-wildcard-filtering-on-genotype-columns"><tt class="docutils literal"><span class="pre">--gt-filter</span></tt> Wildcard filtering on genotype columns.</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#show-samples-finding-out-which-samples-have-a-variant"><tt class="docutils literal"><span class="pre">--show-samples</span></tt> Finding out which samples have a variant</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#show-samples-format-sampledetail-provide-a-flattened-view-of-samples"><tt class="docutils literal"><span class="pre">--show-samples</span> <span class="pre">--format</span> <span class="pre">sampledetail</span></tt> Provide a flattened view of samples</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#show-families-finding-out-which-families-have-a-variant"><tt class="docutils literal"><span class="pre">--show-families</span></tt> Finding out which families have a variant</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#region-restrict-a-query-to-a-specified-region"><tt class="docutils literal"><span class="pre">--region</span></tt> Restrict a query to a specified region</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#sample-filter-restrict-a-query-to-specified-samples"><tt class="docutils literal"><span class="pre">--sample-filter</span></tt> Restrict a query to specified samples</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#sample-delim-changing-the-sample-list-delimiter"><tt class="docutils literal"><span class="pre">--sample-delim</span></tt> Changing the sample list delimiter</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#format-reporting-query-output-in-an-alternate-format"><tt class="docutils literal"><span class="pre">--format</span></tt> Reporting query output in an alternate format.</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#carrier-summary-by-phenotype-summarize-carrier-status"><tt class="docutils literal"><span class="pre">--carrier-summary-by-phenotype</span></tt> Summarize carrier status</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/querying.html#querying-the-gene-tables">Querying the gene tables</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#using-the-gene-detailed-gene-summary-tables">Using the <tt class="docutils literal"><span class="pre">gene_detailed</span></tt> &amp; <tt class="docutils literal"><span class="pre">gene_summary</span></tt> tables</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#restrict-analysis-to-transcripts-with-a-valid-ccds-id">Restrict analysis to transcripts with a valid <tt class="docutils literal"><span class="pre">CCDS_ID</span></tt></a></li>
<li class="toctree-l2"><a class="reference internal" href="content/querying.html#what-if-i-don-t-see-my-gene-in-the-database">What if I don&#8217;t see my <tt class="docutils literal"><span class="pre">gene</span></tt> in the database?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/tools.html">Built-in analysis tools</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#common-args-common-arguments"><tt class="docutils literal"><span class="pre">common_args</span></tt>: common arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#comp-hets-identifying-potential-compound-heterozygotes"><tt class="docutils literal"><span class="pre">comp_hets</span></tt>: Identifying potential compound heterozygotes</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#mendelian-error-identify-non-mendelian-transmission"><tt class="docutils literal"><span class="pre">mendelian_error</span></tt>: Identify non-mendelian transmission.</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#de-novo-identifying-potential-de-novo-mutations"><tt class="docutils literal"><span class="pre">de_novo</span></tt>: Identifying potential de novo mutations.</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#autosomal-recessive-find-variants-meeting-an-autosomal-recessive-model"><tt class="docutils literal"><span class="pre">autosomal_recessive</span></tt>: Find variants meeting an autosomal recessive model.</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#autosomal-dominant-find-variants-meeting-an-autosomal-dominant-model"><tt class="docutils literal"><span class="pre">autosomal_dominant</span></tt>: Find variants meeting an autosomal dominant model.</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#pathways-map-genes-and-variants-to-kegg-pathways"><tt class="docutils literal"><span class="pre">pathways</span></tt>: Map genes and variants to KEGG pathways.</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#interactions-find-genes-among-variants-that-are-interacting-partners"><tt class="docutils literal"><span class="pre">interactions</span></tt>: Find genes among variants that are interacting partners.</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#lof-sieve-filter-lof-variants-by-transcript-position-and-type"><tt class="docutils literal"><span class="pre">lof_sieve</span></tt>: Filter LoF variants by transcript position and type</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#annotate-adding-your-own-custom-annotations"><tt class="docutils literal"><span class="pre">annotate</span></tt>: adding your own custom annotations</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#region-extracting-variants-from-specific-regions-or-genes"><tt class="docutils literal"><span class="pre">region</span></tt>: Extracting variants from specific regions or genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#windower-conducting-analyses-on-genome-windows"><tt class="docutils literal"><span class="pre">windower</span></tt>: Conducting analyses on genome &#8220;windows&#8221;.</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#stats-compute-useful-variant-statistics"><tt class="docutils literal"><span class="pre">stats</span></tt>: Compute useful variant statistics.</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#burden-perform-sample-wise-gene-level-burden-calculations"><tt class="docutils literal"><span class="pre">burden</span></tt>: perform sample-wise gene-level burden calculations</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#roh-identifying-runs-of-homozygosity"><tt class="docutils literal"><span class="pre">ROH</span></tt>: Identifying runs of homozygosity</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#set-somatic-flag-somatic-variants"><tt class="docutils literal"><span class="pre">set_somatic</span></tt>: Flag somatic variants</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#actionable-mutations-report-actionable-somatic-mutations-and-drug-gene-interactions"><tt class="docutils literal"><span class="pre">actionable_mutations</span></tt>: Report actionable somatic mutations and drug-gene interactions</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#fusions-report-putative-gene-fusions"><tt class="docutils literal"><span class="pre">fusions</span></tt>: Report putative gene fusions</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/tools.html#db-info-list-the-gemini-database-tables-and-columns"><tt class="docutils literal"><span class="pre">db_info</span></tt>: List the gemini database tables and columns</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/browser.html">The GEMINI browser interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/database_schema.html">The GEMINI database schema</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/database_schema.html#the-variants-table">The <tt class="docutils literal"><span class="pre">variants</span></tt> table</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/database_schema.html#the-variant-impacts-table">The <tt class="docutils literal"><span class="pre">variant_impacts</span></tt> table</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/database_schema.html#details-of-the-impact-and-impact-severity-columns">Details of the <tt class="docutils literal"><span class="pre">impact</span></tt> and <tt class="docutils literal"><span class="pre">impact_severity</span></tt> columns</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/database_schema.html#the-samples-table">The <tt class="docutils literal"><span class="pre">samples</span></tt> table</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/database_schema.html#the-resources-table">The <tt class="docutils literal"><span class="pre">resources</span></tt> table</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/database_schema.html#the-version-table">The <tt class="docutils literal"><span class="pre">version</span></tt> table</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/database_schema.html#the-gene-detailed-table">The <tt class="docutils literal"><span class="pre">gene_detailed</span></tt> table</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/database_schema.html#the-gene-summary-table">The <tt class="docutils literal"><span class="pre">gene_summary</span></tt> table</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/api.html">Using the GEMINI API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/api.html#the-geminiquery-class">The GeminiQuery class</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/genotype_query_engines.html">Speeding genotype queries</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/genotype_query_engines.html#design-of-gemini">Design of Gemini</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/genotype_query_engines.html#bcolz-indexes">bcolz indexes</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/genotype_query_engines.html#usage">Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/genotype_query_engines.html#design-of-genotype-query-engines">Design of Genotype Query Engines</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/acknowledgements.html">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/history.html">Release History</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#future">(Future)</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id1">0.16.3</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id2">0.16.2</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id3">0.16.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id4">0.16.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id5">0.15.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id6">0.15.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id7">0.14.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#apr-09">0.13.1 (2015-Apr-09)</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id8">0.12.2</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id9">0.12.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id10">0.11.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id11">0.10.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id12">0.10.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id13">0.8.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id14">0.7.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id15">0.7.0</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#id16">0.6.6</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#jan-03">0.6.4 (2014-Jan-03)</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#dec-10">0.6.3.2 (2013-Dec-10)</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#nov-19">0.6.3.1 (2013-Nov-19)</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#nov-7">0.6.3 (2013-Nov-7)</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#oct-7">0.6.2 (2013-Oct-7)</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#sep-09">0.6.1 (2013-Sep-09)</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#sep-02">0.6.0 (2013-Sep-02)</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#b-2013-jul-23">0.5.0b (2013-Jul-23)</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#b-2013-jun-12">0.4.0b (2013-Jun-12)</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/history.html#b">0.3.0b</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/faq.html">F.A.Q.</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/faq.html#does-gemini-work-with-non-human-genomes">Does GEMINI work with non-human genomes?</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/faq.html#what-versions-of-the-human-genome-does-gemini-support">What versions of the human genome does GEMINI support?</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/faq.html#how-can-i-use-plink-files-with-gemini">How can I use PLINK files with GEMINI?</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/other.html">Other information</a></li>
</ul>
</div>
</div>




  

<div id="disqus_thread"></div>
 <script type="text/javascript">
     /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
     var disqus_shortname = 'geminidocs'; // required: replace example with your forum shortname

     /* * * DON'T EDIT BELOW THIS LINE * * */
     (function() {
         var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
         dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
         (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
     })();
 </script>
 <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
 <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>






          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="#">
              <img class="logo" src="_static/gemini.png" alt="Logo"/>
            </a></p><p>
   GEMINI is a flexible framework for exploring genome variation.
</p>
<h3>GEMINI links</h3>
<ul>
  <li><a target="_blank" href="https://github.com/arq5x/gemini/issues">Issue Tracker</a></li>
  <li><a target="_blank" href="https://github.com/arq5x/gemini">Source @ GitHub</a></li>
  <li><a target="_blank" href="http://groups.google.com/group/gemini-variation">Mailing list @ Google Groups</a></li>
  <li><a target="_blank" href="http://quinlanlab.org">Quinlan lab @ UVa</a></li>

</ul>

<h3>Sources</h3>
<p><a href="https://github.com/arq5x/gemini">Browse source @ GitHub</a>.</p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/index.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="content/installation.html" title="Installation"
             >next</a> |</li>
        <li><a href="#">gemini v0.16.3</a> &raquo;</li> 
      </ul>
    </div>

    <div class="footer">
        &copy; Copyright 2012,2013,2014,2015.
      Last updated on Jul 22, 2015.
    <a href="http://www.gnu.org/licenses/gpl-3.0.txt"><p>GPL3 licensed</p></a>
    </div>

<!-- 
<div id="editor-trap">
      <h3>Edit and improve this document!</h3>

      <p>
        This file can be edited directly through the Web. Anyone can
        update and fix errors in this document with few clicks --
        no downloads needed.
      <p>

      <ol>

        <li>
          Go to
        <a href="https://github.com/arq5x/gemini/blob/master/docs/index.rst">
          <strong>GEMINI</strong>: <em>a flexible framework for exploring genome variation</em>
        </a> on GitHub.
      </li>

        <li>
        <b>Edit</b> files using GitHub's text editor in your web browser (see the 'Edit' tab on the top right of the file)
      </li>

      <li>
        Fill in the <b>Commit message</b> text box at the bottom of the page describing <i>why</i>
        you made the changes. Press the <b>Propose file change</b> button next to it when done.
      </li>

      <li>
        Then click <b>Send a pull request</b>.
      </li>

      <li>
        Your changes are now queued for review under the project's <a href="https://github.com/arq5x/gemini/pulls">Pull requests</a> tab on GitHub!
      </li>
      </ol>

      <p>
        For an introduction to the documentation format please see <a href="http://docutils.sourceforge.net/docs/user/rst/quickstart.html">the reST primer</a>.
      </p>

  </div>
 -->
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', UA-39150563-1]);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>


  </body>
</html>