<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN"
            "http://www.w3.org/TR/REC-html40/loose.dtd">
<HTML>
<HEAD>

<META http-equiv="Content-Type" content="text/html; charset=ISO-8859-1">
<META name="GENERATOR" content="hevea 1.07">
<TITLE>
BLAST
</TITLE>
</HEAD>
<BODY >
<A HREF="Tutorial003.html"><IMG SRC ="previous_motif.gif" ALT="Previous"></A>
<A HREF="index.html"><IMG SRC ="contents_motif.gif" ALT="Up"></A>
<A HREF="Tutorial005.html"><IMG SRC ="next_motif.gif" ALT="Next"></A>
<HR>

<H1><A NAME="htoc20">Chapter&nbsp;3</A>&nbsp;&nbsp;BLAST</H1>
<A NAME="sec:blast"></A>
Hey, everybody loves BLAST right? I mean, geez, how can get it get any easier to do comparisons between one of your sequences and every other sequence in the known world? Heck, if I was writing the code to do that it would probably take about a day and a half to complete, and the results still wouldn't be as good. But, of course, this section isn't about how cool BLAST is, since we already know that. It is about the problem with BLAST -- it can be really difficult to deal with the volume of data generated by large runs, and to automate BLAST runs in general.<BR>
<BR>
Fortunately, the Biopython folks know this only too well, so they've developed lots of tools for dealing with BLAST and making things much easier. This section details how to use these tools and do useful things with 'em. <BR>
<BR>
<A NAME="toc10"></A>
<H2><A NAME="htoc21">3.1</A>&nbsp;&nbsp;Running BLAST locally</H2>
<A NAME="sec:running-local-blast"></A>
Running BLAST locally (as opposed to over the internet, see
Section&nbsp;<A HREF="#sec:running-www-blast">3.2</A>) has two advantages:
<UL><LI>
Local BLAST may be faster than BLAST over the internet;
<LI>Local BLAST allows you to make your own database to search for sequences against.
</UL>
Dealing with proprietary or unpublished sequence data can be another reason to run BLAST locally. You may not be allowed to redistribute the sequences, so submitting them to the NCBI as a BLAST query would not be an option. <BR>
<BR>
Biopython provides lots of nice code to enable you to call local BLAST executables from your scripts, and have full access to the many command line options that these executables provide. You can obtain local BLAST precompiled for a number of platforms at <A HREF="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/"><TT>ftp://ftp.ncbi.nlm.nih.gov/blast/executables/</TT></A>, or can compile it yourself in the NCBI toolbox (<A HREF="ftp://ftp.ncbi.nlm.nih.gov/toolbox/"><TT>ftp://ftp.ncbi.nlm.nih.gov/toolbox/</TT></A>).<BR>
<BR>
The code for dealing with local BLAST is found in <CODE>Bio.Blast.NCBIStandalone</CODE>, specifically in the functions <CODE>blastall</CODE> and <CODE>blastpgp</CODE>, which correspond with the BLAST executables that their names imply.<BR>
<BR>
Let's use these functions to run a <CODE>blastall</CODE> against a local database and return the results. First, we want to set up the paths to everything that we'll need to do the BLAST. What we need to know is the path to the database (which should have been prepared using <CODE>formatdb</CODE>, see <A HREF="ftp://ftp.ncbi.nlm.nih.gov/blast/documents/formatdb.html"><TT>ftp://ftp.ncbi.nlm.nih.gov/blast/documents/formatdb.html</TT></A>) to search against, the path to the file we want to search, and the path to the <CODE>blastall</CODE> executable.
<PRE>
&gt;&gt;&gt; my_blast_db = "/home/mdehoon/Data/Genomes/Databases/bsubtilis"
# I used formatdb to create a BLAST database named bsubtilis
# (for Bacillus subtilis) consisting of the following three files:
# /home/mdehoon/Data/Genomes/Databases/bsubtilis.nhr
# /home/mdehoon/Data/Genomes/Databases/bsubtilis.nin
# /home/mdehoon/Data/Genomes/Databases/bsubtilis.nsq

&gt;&gt;&gt; my_blast_file = "m_cold.fasta"
# A FASTA file with the sequence I want to BLAST

&gt;&gt;&gt; my_blast_exe = "/usr/local/blast/bin/blastall"
# The name of my BLAST executable
</PRE>
The FASTA file used in this example is available
<A HREF="examples/m_cold.fasta">here</A> as well as
<A HREF="http://biopython.org/DIST/docs/tutorial/examples/m_cold.fasta">online</A>.<BR>
<BR>
Now that we've got that all set, we are ready to run the BLAST and collect the results. We can do this with two lines:
<PRE>
&gt;&gt;&gt; from Bio.Blast import NCBIStandalone
&gt;&gt;&gt; result_handle, error_info = NCBIStandalone.blastall(my_blast_exe, "blastn",
                                                    my_blast_db, my_blast_file)
</PRE>
Note that the Biopython interfaces to local blast programs returns two values. The first is a handle to the blast output, which is ready to either be saved or passed to a parser. The second is the possible error output generated by the blast command.<BR>
<BR>
The error info can be hard to deal with, because if you try to do a <CODE>error_info.read()</CODE> and there was no error info returned, then the <CODE>read()</CODE> call will block and not return, locking your script. In my opinion, the best way to deal with the error is only to print it out if you are not getting <CODE>blast_out</CODE> results to be parsed, but otherwise to leave it alone.<BR>
<BR>
This command will generate BLAST output in XML format, as that is the format expected by the XML parser, described in Section&nbsp;<A HREF="#sec:parsing-blast">3.4</A>. For plain text output, use the <CODE>align_view='0'</CODE> keyword. To parse text output instead of XML output, see the Section&nbsp;<A HREF="#sec:parsing-blast-deprecated">3.6</A> below. However, parsing text output is not recommended, as the BLAST plain text output changes frequently, breaking our parsers.<BR>
<BR>
If you are interested in saving your results to a file before parsing them,
see Section&nbsp;<A HREF="#sec:saving-blast-output">3.3</A>. To find out how to parse the BLAST
results, go to Section&nbsp;<A HREF="#sec:parsing-blast">3.4</A><BR>
<BR>
<A NAME="toc11"></A>
<H2><A NAME="htoc22">3.2</A>&nbsp;&nbsp;Running BLAST over the Internet</H2>
<A NAME="sec:running-www-blast"></A>
The first step in automating BLASTing is to make everything accessible
from python scripts. So, Biopython contains code that allows you to
run the WWW version of BLAST
(<A HREF="http://www.ncbi.nlm.nih.gov/BLAST/"><TT>http://www.ncbi.nlm.nih.gov/BLAST/</TT></A>) directly from
your python scripts. This is very nice, especially since BLAST can be
a real pain to deal with from scripts, especially with the whole BLAST
queue thing and the separate results page. Keeping the Biopython code
up to date with all of the changes at NCBI is hard enough!<BR>
<BR>
The code to deal with the WWW version of BLAST is found in the
<CODE>Bio.Blast.NCBIWWW</CODE> module, and the <CODE>qblast</CODE> function. Let's
say we want to BLAST info we have in a FASTA formatted file against
the database. First, we need to get the info in the FASTA file. The
easiest way to do this is to use the <CODE>Bio.SeqIO</CODE> module.
In this example, we'll use <CODE>Bio.SeqIO.parse</CODE> to parse the FASTA file
and store the first FASTA record in the file in a SeqRecord object
(section&nbsp;<A HREF="Tutorial003.html#sec:fasta-parsing">2.4.1</A> explains <CODE>Bio.SeqIO.parse</CODE> in
more detail).
<PRE>
&gt;&gt;&gt; from Bio import SeqIO
&gt;&gt;&gt; file_handle = open("m_cold.fasta")
&gt;&gt;&gt; records = SeqIO.parse(file_handle, format="fasta")
&gt;&gt;&gt; record = records.next()
</PRE>
Now we take the sequence as a plain string from the SeqRecord:
<PRE>
&gt;&gt;&gt; sequence = record.seq.data
</PRE>and run BLAST on it. The code to do the simplest possible BLAST
(a simple blastn of the FASTA file against all of the non-redundant
databases) is:
<PRE>
&gt;&gt;&gt; from Bio.Blast import NCBIWWW
&gt;&gt;&gt; result_handle = NCBIWWW.qblast("blastn", "nr", sequence)
</PRE>
The <CODE>qblast</CODE> function also take a number of other option arguments
which are basically analogous to the different parameters you can set
on the basic BLAST page
(<A HREF="http://www.ncbi.nlm.nih.gov/blast/blast.cgi?Jform=0"><TT>http://www.ncbi.nlm.nih.gov/blast/blast.cgi?Jform=0</TT></A>),
but for this I'll just talk about the first few arguments, which are
the most important. The first three are non-optional.
<UL><LI>
The first argument is the blast program to use for the search. The
options and descriptions of the programs are available at
<A HREF="http://www.ncbi.nlm.nih.gov/BLAST/blast_program.html"><TT>http://www.ncbi.nlm.nih.gov/BLAST/blast_program.html</TT></A>.
Currently <CODE>qblast</CODE> only works with blastn and blastp as program
arguments --- let us known if you want to use one of the other blast
programs instead.
<LI>The second argument specifies the databases to search against. Again,
the options for this are available on the NCBI web pages at
<A HREF="http://www.ncbi.nlm.nih.gov/BLAST/blast_databases.html"><TT>http://www.ncbi.nlm.nih.gov/BLAST/blast_databases.html</TT></A>.
<LI>The third argument is your Fasta sequence as a plain string.
<LI>The <CODE>qblast</CODE> function can return the BLAST results in various
formats, which you can choose with the optional <CODE>format_type</CODE> keyword:
<CODE>"HTML"</CODE>, <CODE>"Text"</CODE>, <CODE>"ASN.1"</CODE>, or <CODE>"XML"</CODE>.
The default is <CODE>"XML"</CODE>, as that is the format expected by the parser,
described in section&nbsp;<A HREF="#sec:parsing-blast">3.4</A> below.
</UL>
After you have set the search options, you are all ready to BLAST.
Biopython takes care of worrying about when the results are available,
and will pause until it can get the results and return them.<BR>
<BR>
<A NAME="toc12"></A>
<H2><A NAME="htoc23">3.3</A>&nbsp;&nbsp;Saving BLAST output</H2>
<A NAME="sec:saving-blast-output"></A>
Before parsing the results, it is often useful to save them into a
file so that you can use them later without having to go back and
re-blast everything. I find this especially useful when debugging my
code that extracts info from the BLAST files, but it could also be
useful just for making backups of things you've done.<BR>
<BR>
If you don't want to save the BLAST output, you can skip to
section&nbsp;<A HREF="#sec:parsing-blast">3.4</A>. If you do, read on.<BR>
<BR>
We need to be a bit careful since we can use <CODE>result_handle.read()</CODE> to
read the BLAST output only once -- calling <CODE>result_handle.read()</CODE> again
returns an empty string. First, we use <CODE>read()</CODE> and store all of
the information from the handle into a string:
<PRE>
&gt;&gt;&gt; blast_results = result_handle.read()
</PRE>
Next, we save this string in a file:
<PRE>
&gt;&gt;&gt; save_file = open("my_blast.xml", "w")
&gt;&gt;&gt; save_file.write(blast_results)
&gt;&gt;&gt; save_file.close()
</PRE>
After doing this, the results are in the file <CODE>my_blast.xml</CODE> and the
variable <CODE>blast_results</CODE> contains the BLAST results in a string
form. However, the <CODE>parse</CODE> function of the BLAST parser (described
in&nbsp;<A HREF="#sec:parsing-blast">3.4</A>) takes a file-handle-like object, not a
plain string. To get a handle, there are two things you can do:
<UL><LI>
Use the Python standard library module <CODE>cStringIO</CODE>. The
following code will turn the plain string into a handle, which we can
feed directly into the BLAST parser:
<PRE>
&gt;&gt;&gt; import cStringIO
&gt;&gt;&gt; result_handle = cStringIO.StringIO(blast_results)
</PRE><LI>Open the saved file for reading. Duh.
<PRE>
&gt;&gt;&gt; result_handle = open("my_blast.xml")
</PRE></UL>
Now that we've got the BLAST results, we are ready to do
something with them, so this leads us right into the parsing section.<BR>
<BR>
<A NAME="toc13"></A>
<H2><A NAME="htoc24">3.4</A>&nbsp;&nbsp;Parsing BLAST output</H2>
<A NAME="sec:parsing-blast"></A>
As mentioned above, BLAST can generate output in various formats,
such as XML, HTML, and plain text. Originally, Biopython had a parser for
BLAST plain text and HTML output, as these were the output formats supported
by BLAST. Unfortunately, the BLAST output in these formats kept changing,
each time breaking the Biopython parsers. As keeping up with changes in BLAST
became a hopeless endeavor, especially with users running different BLAST
versions, we now recommend to parse the output in XML format, which can be
generated by recent versions of BLAST. Not only is the XML output more stable
than the plain text and HTML output, it is also much easier to parse
automatically, making Biopython a whole lot more stable.<BR>
<BR>
Though deprecated, the parsers for BLAST output in plain text or HTML output
are still available in Biopython
(see Section&nbsp;<A HREF="#sec:parsing-blast-deprecated">3.6</A>).
Use them at your own risk: they may or may not work, depending on which BLAST
version you're using.<BR>
<BR>
You can get BLAST output in XML format in various ways. For the parser, it
doesn't matter how the output was generated, as long as it is in the XML format.
<UL><LI>
You can use Biopython to run BLAST locally, as described in
section&nbsp;<A HREF="#sec:running-local-blast">3.1</A>.
<LI>You can use Biopython to run BLAST over the internet, as described in
section&nbsp;<A HREF="#sec:running-www-blast">3.2</A>.
<LI>You can do the BLAST seach yourself on the NCBI site through your
web browser, and then save the results. You need to choose XML as the format
in which to receive the results, and save the final BLAST page you get
(you know, the one with all of the interesting results!) to a file.
<LI>You can also run BLAST locally without using Biopython, and save
the output in a file. Again, you need to choose XML as the format in which
to receive the results.
</UL>
The important point is that you do not have to use Biopython
scripts to fetch the data in order to be able to parse it.<BR>
<BR>
Doing things in one of these ways, you then need to get a handle
to the results. In python, a handle is just a nice general way of
describing input to any info source so that the info can be retrieved
using <CODE>read()</CODE> and <CODE>readline()</CODE> functions. This is the type
of input the BLAST parser (and the other Biopython parsers take). <BR>
<BR>
If you followed the code above for interacting with BLAST through a
script, then you already have <CODE>result_handle</CODE>, the handle to the
BLAST results. If instead you ran BLAST some other way, and have the
BLAST output (in XML format) in the file <CODE>my_blast.xml</CODE>, all you
need to do is to open the file for reading:
<PRE>
&gt;&gt;&gt; result_handle = open("my_blast.xml")
</PRE>
Now that we've got a handle, we are ready to parse the output. The
code to parse it is really quite small:
<PRE>
&gt;&gt;&gt; from Bio.Blast import NCBIXML
&gt;&gt;&gt; blast_records = NCBIXML.parse(result_handle)
</PRE>
To understand what <CODE>NCBIXML.parse</CODE> returns, there are two things
that you need to keep in mind:
<UL><LI>
The BLAST output may contain the output of more than
one BLAST search. This will for example be the case if you ran BLAST locally
on a Fasta file containing more than one sequence. For each sequence, the
BLAST parser will return one BLAST record.
<LI>The BLAST output may therefore be huge.
</UL>
To be able to handle these situations, <CODE>NCBIXML.parse</CODE> returns an
iterator (just like <CODE>Bio.SeqIO.parse</CODE>). In plain English, an iterator
allows you to step through the BLAST output, retrieving BLAST records one
by one for each BLAST search:
<PRE>
&gt;&gt;&gt; blast_record = blast_records.next()
# ... do something with blast_record
&gt;&gt;&gt; blast_record = blast_records.next()
# ... do something with blast_record
&gt;&gt;&gt; blast_record = blast_records.next()
# ... do something with blast_record
&gt;&gt;&gt; blast_record = blast_records.next()
Traceback (most recent call last):
  File "&lt;stdin&gt;", line 1, in &lt;module&gt;
StopIteration
# No further records
</PRE>
Or, you can use a <CODE>for</CODE>-loop:
<PRE>
&gt;&gt;&gt; for blast_record in blast_records:
...     # Do something with blast_record
</PRE>
Note though that you can step through the BLAST records only once.
Usually, from each BLAST record you would save the information that
you are interested in. If you want to save all returned BLAST records,
you can convert the iterator into a list:
<PRE>
&gt;&gt;&gt; blast_records = list(blast_records)
</PRE>Now you can access each BLAST record in the list with an index as usual.
If your BLAST file is huge though, you may run into problems trying to
save them all in a list.<BR>
<BR>
Usually, you'll be running one BLAST search at a time. Then, all you need
to do is to pick up the first (and only) BLAST record in <CODE>blast_records</CODE>:
<PRE>
&gt;&gt;&gt; blast_record = blast_records.next()
</PRE>
I guess by now you're wondering what is in a BLAST record.<BR>
<BR>
<A NAME="toc14"></A>
<H2><A NAME="htoc25">3.5</A>&nbsp;&nbsp;The BLAST record class</H2>
A BLAST Record contains everything you might ever want to extract from the
BLAST output. Right now we'll just show
an example of how to get some info out of the BLAST report, but if you
want something in particular that is not described here, look at the
info on the record class in detail, and take a gander into the code or
automatically generated documentation -- the docstrings have lots of
good info about what is stored in each piece of information.<BR>
<BR>
To continue with our example, let's just print out some summary info
about all hits in our blast report greater than a particular
threshold. The following code does this:
<PRE>
&gt;&gt;&gt; E_VALUE_THRESH = 0.04

&gt;&gt;&gt; for alignment in blast_record.alignments:
...     for hsp in alignment.hsps:
...         if hsp.expect &lt; E_VALUE_THRESH:
...             print '****Alignment****'
...             print 'sequence:', alignment.title
...             print 'length:', alignment.length
...             print 'e value:', hsp.expect
...             print hsp.query[0:75] + '...'
...             print hsp.match[0:75] + '...'
...             print hsp.sbjct[0:75] + '...'
</PRE>
This will print out summary reports like the following:
<PRE>
****Alignment****
sequence: &gt;gb|AF283004.1|AF283004 Arabidopsis thaliana cold acclimation protein WCOR413-like protein
alpha form mRNA, complete cds
length: 783
e value: 0.034
tacttgttgatattggatcgaacaaactggagaaccaacatgctcacgtcacttttagtcccttacatattcctc...
||||||||| | ||||||||||| || ||||  || || |||||||| |||||| |  | |||||||| ||| ||...
tacttgttggtgttggatcgaaccaattggaagacgaatatgctcacatcacttctcattccttacatcttcttc...
</PRE>
Basically, you can do anything you want to with the info in the BLAST
report once you have parsed it. This will, of course, depend on what
you want to use it for, but hopefully this helps you get started on
doing what you need to do!<BR>
<BR>
An important consideration for extracting information from a BLAST report is the type of objects that the information is stored in. In Biopython, the parsers return <CODE>Record</CODE> objects, either <CODE>Blast</CODE> or <CODE>PSIBlast</CODE> depending on what you are parsing. These objects are defined in <CODE>Bio.Blast.Record</CODE> and are quite complete.<BR>
<BR>
Here are my attempts at UML class diagrams for the <CODE>Blast</CODE> and <CODE>PSIBlast</CODE> record classes. If you are good at UML and see mistakes/improvements that can be made, please let me know. The Blast class diagram is shown in Figure&nbsp;<A HREF="#fig:blastrecord">3.5</A>.<BR>
<BR>

<A NAME="fig:blastrecord"></A>
<IMG SRC="images/BlastRecord.png" width=650, height=750>
<BR>
<BR>
The PSIBlast record object is similar, but has support for the rounds that are used in the iteration steps of PSIBlast. The class diagram for PSIBlast is shown in Figure&nbsp;<A HREF="#fig:psiblastrecord">3.5</A>.<BR>
<BR>

<A NAME="fig:psiblastrecord"></A>
<IMG SRC="images/PSIBlastRecord.png" width=650, height=750>
<BR>
<BR>
<A NAME="toc15"></A>
<H2><A NAME="htoc26">3.6</A>&nbsp;&nbsp;Deprecated BLAST parsers</H2>
<A NAME="sec:parsing-blast-deprecated"></A>
Older versions of Biopython had parsers for BLAST output in plain text or HTML
format. Over the years, we discovered that it is very hard to maintain these
parsers in working order. Basically, any small change to the BLAST output in
newly released BLAST versions tends to cause the plain text and HTML parsers
to break. We therefore recommend parsing BLAST output in XML format, as
described in section&nbsp;<A HREF="#sec:parsing-blast">3.4</A>.
Whereas the plain text and HTML parsers are still available in
Biopython, use them at your own risk. They may or may not work, depending on
which BLAST versions you're using.<BR>
<BR>

<H3><A NAME="htoc27">3.6.1</A>&nbsp;&nbsp;Parsing plain-text BLAST output</H3>
The plain text BLAST parser is located in <CODE>Bio.Blast.NCBIStandalone</CODE>.
As with the XML parser, we need to have a handle object that we can pass to the parser. The handle must implement the <CODE>readline()</CODE> method and do this properly. The common ways to get such a handle are to either use the provided <CODE>blastall</CODE> or <CODE>blastpgp</CODE> functions to run the local blast, or to run a local blast via the command line, and then do something like the following:
<PRE>
&gt;&gt;&gt; result_handle = open("my_file_of_blast_output.txt")
</PRE>
Well, now that we've got a handle (which we'll call <CODE>result_handle</CODE>),
we are ready to parse it. This can be done with the following code:
<PRE>
&gt;&gt;&gt; from Bio.Blast import NCBIStandalone
&gt;&gt;&gt; blast_parser = NCBIStandalone.BlastParser()
&gt;&gt;&gt; blast_record = blast_parser.parse(result_handle)
</PRE>
This will parse the BLAST report into a Blast Record class (either a Blast or a PSIBlast record, depending on what you are parsing) so that you can extract the information from it. In our case, let's just use print out a quick summary of all of the alignments greater than some threshold value.
<PRE>
&gt;&gt;&gt; E_VALUE_THRESH = 0.04
&gt;&gt;&gt; for alignment in b_record.alignments:
...     for hsp in alignment.hsps:
...         if hsp.expect &lt; E_VALUE_THRESH:
...             print '****Alignment****'
...             print 'sequence:', alignment.title
...             print 'length:', alignment.length
...             print 'e value:', hsp.expect
...             print hsp.query[0:75] + '...'
...             print hsp.match[0:75] + '...'
...             print hsp.sbjct[0:75] + '...'
</PRE>
If you also read the section&nbsp;<A HREF="#sec:parsing-blast">3.4</A> on parsing BLAST XML output, you'll notice that the above code is identical to what is found in that section. Once you parse something into a record class you can deal with it independent of the format of the original BLAST info you were parsing. Pretty snazzy!<BR>
<BR>
Sure, parsing one record is great, but I've got a BLAST file with tons of records -- how can I parse them all? Well, fear not, the answer lies in the very next section.<BR>
<BR>

<H3><A NAME="htoc28">3.6.2</A>&nbsp;&nbsp;Parsing a file full of BLAST runs</H3>
Of course, local blast is cool because you can run a whole bunch of sequences against a database and get back a nice report on all of it. So, Biopython definitely has facilities to make it easy to parse humongous files without memory problems. <BR>
<BR>
We can do this using the blast iterator. To set up an iterator, we first set up a parser, to parse our blast reports in Blast Record objects:
<PRE>
&gt;&gt;&gt; from Bio.Blast import NCBIStandalone
&gt;&gt;&gt; blast_parser = NCBIStandalone.BlastParser()
</PRE>
Then we will assume we have a handle to a bunch of blast records, which we'll call <CODE>result_handle</CODE>. Getting a handle is described in full detail above in the blast parsing sections. <BR>
<BR>
Now that we've got a parser and a handle, we are ready to set up the iterator with the following command:
<PRE>
&gt;&gt;&gt; blast_iterator = NCBIStandalone.Iterator(blast_handle, blast_parser)
</PRE>
The second option, the parser, is optional. If we don't supply a parser, then the iterator will just return the raw BLAST reports one at a time.<BR>
<BR>
Now that we've got an iterator, we start retrieving blast records (generated by our parser) using <CODE>next()</CODE>:
<PRE>
&gt;&gt;&gt; blast_record = blast_iterator.next()
</PRE>
Each call to next will return a new record that we can deal with. Now we can iterate through this records and generate our old favorite, a nice little blast report:
<PRE>
&gt;&gt;&gt; while 1:
...     b_record = b_iterator.next()

...     if b_record is None:
...         break

...     E_VALUE_THRESH = 0.04
...     for alignment in b_record.alignments:
...         for hsp in alignment.hsps:
...             if hsp.expect &lt; E_VALUE_THRESH:
...                 print '****Alignment****'
...                 print 'sequence:', alignment.title
...                 print 'length:', alignment.length
...                 print 'e value:', hsp.expect

...                 if len(hsp.query) &gt; 75:
...                     dots = '...'
...                 else:
...                     dots = ''
                
...                 print hsp.query[0:75] + dots
...                 print hsp.match[0:75] + dots
...                 print hsp.sbjct[0:75] + dots
</PRE>
Notice that <CODE>b_iterator.next()</CODE> will return <CODE>None</CODE> when it runs out of records to parse, so it is easy to iterate through the entire file with a while loop that checks for the existence of a record.<BR>
<BR>
The iterator allows you to deal with huge blast records without any memory problems, since things are read in one at a time. I have parsed tremendously huge files without any problems using this.<BR>
<BR>

<H3><A NAME="htoc29">3.6.3</A>&nbsp;&nbsp;Finding a bad record somewhere in a huge file</H3>
One really ugly problem that happens to me is that I'll be parsing a huge blast file for a while, and the parser will bomb out with a SyntaxError. This is a serious problem, since you can't tell if the SyntaxError is due to a parser problem, or a problem with the BLAST. To make it even worse, you have no idea where the parse failed, so you can't just ignore the error, since this could be ignoring an important data point.<BR>
<BR>
We used to have to make a little script to get around this problem, but the <CODE>Bio.Blast</CODE> module now includes a <CODE>BlastErrorParser</CODE> which really helps make this easier. The <CODE>BlastErrorParser</CODE> works very similar to the regular <CODE>BlastParser</CODE>, but it adds an extra layer of work by catching SyntaxErrors that are generated by the parser, and attempting to diagnose the errors.<BR>
<BR>
Let's take a look at using this parser -- first we define the file we are going to parse and the file to write the problem reports to:
<PRE>
&gt;&gt;&gt; import os
 
&gt;&gt;&gt; blast_file = os.path.join(os.getcwd(), "blast_out", "big_blast.out")
&gt;&gt;&gt; error_file = os.path.join(os.getcwd(), "blast_out", "big_blast.problems")
</PRE>
Now we want to get a <CODE>BlastErrorParser</CODE>:
<PRE>
&gt;&gt;&gt; from Bio.Blast import NCBIStandalone

&gt;&gt;&gt; error_handle = open(error_file, "w")

&gt;&gt;&gt; blast_error_parser = NCBIStandalone.BlastErrorParser(error_handle)
</PRE>
Notice that the parser take an optional argument of a handle. If a handle is passed, then the parser will write any blast records which generate a SyntaxError to this handle. Otherwise, these records will not be recorded.<BR>
<BR>
Now we can use the <CODE>BlastErrorParser</CODE> just like a regular blast parser. Specifically, we might want to make an iterator that goes through our blast records one at a time and parses them with the error parser:
<PRE>
&gt;&gt;&gt; result_handle = open(blast_file)
&gt;&gt;&gt; iterator = NCBIStandalone.Iterator(result_handle, blast_error_parser)
</PRE>
Just like before, we can read these records one a time with, but now we can catch and deal with errors that are due to problems with Blast (and not with the parser itself):
<PRE>
&gt;&gt;&gt; try:
...     next_record = iterator.next()
... except NCBIStandalone.LowQualityBlastError, info:
...     print "LowQualityBlastError detected in id %s" % info[1]
</PRE>
Right now the <CODE>BlastErrorParser</CODE> can generate the following errors:
<UL><LI>
 <CODE>SyntaxError</CODE> -- This is the same error generated by the regular BlastParser, and is due to the parser not being able to parse a specific file. This is normally either due to a bug in the parser, or some kind of discrepancy between the version of BLAST you are using and the versions the parser is able to handle.<BR>
<BR>
<LI><CODE>LowQualityBlastError</CODE> -- When BLASTing a sequence that is of really bad quality (for example, a short sequence that is basically a stretch of one nucleotide), it seems that Blast ends up masking out the entire sequence and ending up with nothing to parse. In this case it will produce a truncated report that causes the parser to generate a SyntaxError. <CODE>LowQualityBlastError</CODE> is reported in these cases. This error returns an info item with the following information:
 <UL><LI>
 <CODE>item[0]</CODE> -- The error message
 <LI><CODE>item[1]</CODE> -- The id of the input record that caused the error. This is really useful if you want to record all of the records that are causing problems.
 </UL>
</UL>
As mentioned, with each error generated, the BlastErrorParser will write the offending record to the specified <CODE>error_handle</CODE>. You can then go ahead and look and these and deal with them as you see fit. Either you will be able to debug the parser with a single blast report, or will find out problems in your blast runs. Either way, it will definitely be a useful experience!<BR>
<BR>
Hopefully the <CODE>BlastErrorParser</CODE> will make it much easier to debug and deal with large Blast files.<BR>
<BR>
<A NAME="toc16"></A>
<H2><A NAME="htoc30">3.7</A>&nbsp;&nbsp;Dealing with PSIBlast</H2>
We should write some stuff to make it easier to deal directly with PSIBlast from scripts (i.&nbsp;e.&nbsp;output the align file in the proper format from an alignment). I need to look at PSIBlast more and come up with some good ways of going this...<BR>
<BR>
<HR>
<A HREF="Tutorial003.html"><IMG SRC ="previous_motif.gif" ALT="Previous"></A>
<A HREF="index.html"><IMG SRC ="contents_motif.gif" ALT="Up"></A>
<A HREF="Tutorial005.html"><IMG SRC ="next_motif.gif" ALT="Next"></A>
</BODY>
</HTML>
