<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Built-in analysis tools</title>
    
    <link rel="stylesheet" href="../_static/labibi.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.16.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/gemini.png"/>
    <link rel="top" title="None" href="../index.html" />
    <link rel="next" title="The GEMINI browser interface" href="browser.html" />
    <link rel="prev" title="Querying the GEMINI database" href="querying.html" />


<script src="../_static/labibi.js"></script>



<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', '']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>


  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="browser.html" title="The GEMINI browser interface"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="querying.html" title="Querying the GEMINI database"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">gemini v0.16.3</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  
  <div class="section" id="built-in-analysis-tools">
<h1>Built-in analysis tools<a class="headerlink" href="#built-in-analysis-tools" title="Permalink to this headline">¶</a></h1>
<p>As of version 0.14, the tools have a standardized output that is different
from previous versions.
Requested <cite>&#8211;columns</cite> will come first followed by a standard set of columns:</p>
<blockquote>
<div><ul class="simple">
<li>variant_id - unique id from the databse</li>
<li>family_id - family id for this row</li>
<li>family_members - which family members were tested</li>
<li>family_genotypes - genotypes of this family</li>
<li>samples - samples contributing to this row appearing in the results</li>
<li>family_count - number of families with this effect</li>
</ul>
</div></blockquote>
<p>Other tools such as <cite>mendel_errors</cite> additional columns at the end.</p>
<p>When queries are limited to variants in genes, the output will be in
<cite>gene</cite>, <cite>chrom</cite> order as opposed to the usual <cite>gene</cite>, <cite>position</cite> order.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>As of version 0.16.0, the inheritance tools (autsomal_dominant,
autosomal_recessive, comp_het, mendel_errors, de_novo) are now
more strict by default.</p>
<p>A &#8211;lenient flag allows, e.g. allows some samples to be of unknown
phenotype or to not have both parents of known phenotype.</p>
<p>The &#8211;allow-unaffected flag will result in reporting variants where
unaffected samples share the same variants. The default will only
report variants that are unique to affected samples.</p>
<p class="last">The design decisions for this change are described here:
<a class="reference external" href="https://github.com/arq5x/gemini/issues/388">https://github.com/arq5x/gemini/issues/388</a>
a visual representation is here:
<a class="reference external" href="https://github.com/arq5x/gemini/blob/master/inheritance.ipynb">https://github.com/arq5x/gemini/blob/master/inheritance.ipynb</a></p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Candidate variants reported by the built-in inheritance model tools
will appear in order by chromosome, then alphabetically by gene.
In other words, they will not be in strict positional order for each chromosome.
This is in an effort to group all candidate variants by gene since the gene
is typically the atomic unit of interest.</p>
</div>
<div class="section" id="common-args-common-arguments">
<h2><tt class="docutils literal"><span class="pre">common_args</span></tt>: common arguments<a class="headerlink" href="#common-args-common-arguments" title="Permalink to this headline">¶</a></h2>
<p>The inheritance tools share a common set of arguments. We will
describe them here and refer to them in the corresponding sections:</p>
<div class="section" id="columns">
<h3><tt class="docutils literal"><span class="pre">--columns</span></tt><a class="headerlink" href="#columns" title="Permalink to this headline">¶</a></h3>
<p>This flag is followed by a comma-delimited list of columns the user is
requestin in the output.</p>
</div>
<div class="section" id="min-kindreds-1">
<h3><tt class="docutils literal"><span class="pre">--min-kindreds</span> <span class="pre">1</span></tt><a class="headerlink" href="#min-kindreds-1" title="Permalink to this headline">¶</a></h3>
<p>This is the number of families required to have a variant in the same gene
in order for it to be reported. For example, we may only be interested in
candidates where at least 4 families have a variant in that gene.</p>
</div>
<div class="section" id="families">
<h3><tt class="docutils literal"><span class="pre">--families</span></tt><a class="headerlink" href="#families" title="Permalink to this headline">¶</a></h3>
<p>By default, candidate variants are reported for all families in the database.
One can restrict the analysis to variants in specific familes with the
<tt class="docutils literal"><span class="pre">--families</span></tt> option.  Families should be provided as a comma-separated list</p>
</div>
<div class="section" id="filter">
<h3><tt class="docutils literal"><span class="pre">--filter</span></tt><a class="headerlink" href="#filter" title="Permalink to this headline">¶</a></h3>
<p>By default, each tool will report all variants regardless of their putative
functional impact. In order to apply additional constraints on the variants
returned, one can use the <tt class="docutils literal"><span class="pre">--filter</span></tt> option. Using SQL syntax, conditions
applied with the <tt class="docutils literal"><span class="pre">--filter</span></tt> options become WHERE clauses in the query issued to
the GEMINI database.</p>
</div>
<div class="section" id="d-0-depth">
<h3><tt class="docutils literal"><span class="pre">-d</span> <span class="pre">[0]</span></tt> (depth)<a class="headerlink" href="#d-0-depth" title="Permalink to this headline">¶</a></h3>
<p>Unfortunately, spurious inherited variants can often appear due to
insufficient sequence coverage
One simple way to filter such artifacts is to enforce a minimum sequence
depth (default: 0) for each sample. We can do that with this flag.</p>
</div>
<div class="section" id="allow-unaffected">
<h3><tt class="docutils literal"><span class="pre">--allow-unaffected</span></tt><a class="headerlink" href="#allow-unaffected" title="Permalink to this headline">¶</a></h3>
<p>By default, candidates that also appear in unaffected samples are not reported
if this flag is specified, such variants will be reported.</p>
</div>
<div class="section" id="lenient">
<h3><tt class="docutils literal"><span class="pre">--lenient</span></tt><a class="headerlink" href="#lenient" title="Permalink to this headline">¶</a></h3>
<p>Loosen the restrictions on family structure. This will allow, for example,
finding compound_hets in unaffected samples.</p>
</div>
<div class="section" id="gt-pl-max">
<h3><tt class="docutils literal"><span class="pre">--gt-pl-max</span></tt><a class="headerlink" href="#gt-pl-max" title="Permalink to this headline">¶</a></h3>
<p>In order to eliminate less confident genotypes, it is possible to enforce a maximum PL value
for each sample. On this scale, lower values indicate more confidence that the called genotype
is correct. 10 is a reasonable value:</p>
</div>
</div>
<div class="section" id="comp-hets-identifying-potential-compound-heterozygotes">
<h2><tt class="docutils literal"><span class="pre">comp_hets</span></tt>: Identifying potential compound heterozygotes<a class="headerlink" href="#comp-hets-identifying-potential-compound-heterozygotes" title="Permalink to this headline">¶</a></h2>
<p>Many recessive disorders are caused by compound heterozygotes. Unlike canonical
recessive sites where the same recessive allele is inherited from both parents
at the _same_ site in the gene, compound heterozygotes occur when
the individual&#8217;s phenotype is caused by two heterozygous recessive alleles at
_different_ sites in a particular gene.</p>
<p>We are looking for two (typically loss-of-function (LoF))
heterozygous variants impacting the same gene at different loci.  The
complicating factor is that this is _recessive_ and as such, we must also
require that the consequential alleles at each heterozygous site were
inherited on different chromosomes (one from each parent).
Where possible, <cite>comp_hets</cite> will phase by transmission. Once this has been
done, the <cite>comp_hets</cite> tool will provide a report of candidate compound
heterozygotes for each sample/gene.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>As of version 0.16.0 the <tt class="docutils literal"><span class="pre">comp_het</span></tt> tool will perform family-based phasing
by default in order to provide better candidates even in the absence of
unphased genotypes. Any candidate that could be one element of a comp_het
will also be phaseable as long as the parents and their genotypes are known.</p>
<p>As of version 0.16.1, the &#8211;ignore-phasing option is removed and there is no
&#8211;lenient option.</p>
<p class="last">In 0.16.2, a &#8211;pattern-only flag was added to find compound hets by inheritance
pattern without regard to affection status. A priority code was also added where
variants with priority 1 are much more informative. See docs below for further
information.</p>
</div>
<div class="section" id="genotype-requirements">
<h3>Genotype Requirements<a class="headerlink" href="#genotype-requirements" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>All affected individuals must be heterozygous at both sites.</li>
<li>No unaffected can be heterozygous at both sites.</li>
<li>No unaffected can be homozygous alterate at either site.</li>
<li>No phased-unaffected can be heterozygous at both sites.<ol class="loweralpha">
<li><cite>&#8211;allow-unaffected</cite> keeps sites where a phased unaffected shares the het-pair</li>
<li>unphased, unaffected that share the het pair are counted and reported for each candidate pair.</li>
</ol>
</li>
<li>Remove candidates where an affected from the same family does NOT share the same het pair.</li>
<li>Sites are automatically phased by transmission when parents are present in order to remove false positive candidates.</li>
</ul>
<p>we prioritize with these rules:</p>
<table border="1" class="docutils">
<colgroup>
<col width="4%" />
<col width="4%" />
<col width="5%" />
<col width="12%" />
<col width="11%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">mom</th>
<th class="head">dad</th>
<th class="head">kid</th>
<th class="head">phaseable</th>
<th class="head">priority</th>
<th class="head">notes</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>R-H</td>
<td>H-R</td>
<td>H-H</td>
<td>both</td>
<td>1</td>
<td>both sites phaseable and alts on opposite chroms</td>
</tr>
<tr class="row-odd"><td>R-H</td>
<td>H-H</td>
<td>H-H</td>
<td>one</td>
<td>2</td>
<td>should be a rare occurrence</td>
</tr>
<tr class="row-even"><td>H-H</td>
<td>H-H</td>
<td>H-H</td>
<td>NO</td>
<td>2</td>
<td>should be a rare occurrence</td>
</tr>
<tr class="row-odd"><td>A-R</td>
<td>H-H</td>
<td>H-H</td>
<td>both</td>
<td>NA</td>
<td>exclude hom-alts from un-affecteds</td>
</tr>
<tr class="row-even"><td>R-R</td>
<td>H-H</td>
<td>H-H</td>
<td>both</td>
<td>NA</td>
<td>phaseable, but alts are on the same chroms.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">candidates of priority != 1 are very unlikely (&lt; 1%) to be real
(see: <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3734130/">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3734130/</a>); we report them
for completeness, but strongly recommend using priority 1 only.</p>
</div>
</div>
<div class="section" id="pattern-only">
<h3>Pattern Only<a class="headerlink" href="#pattern-only" title="Permalink to this headline">¶</a></h3>
<p>To find compound heterozygotes by inheritance pattern only, without regard to affections, the
following rules are used (with &#8211;pattern-only):</p>
<ul class="simple">
<li>Kid must be HET at both sites.</li>
<li>Kid must have alts on different chromosomes.</li>
<li>Neither parent can be HOM_ALT at either site.</li>
<li>If either parent is phased at both sites and matches the kid, it&#8217;s excluded.</li>
<li>If either parent is HET at both sites, priority is reduced.</li>
<li>When the above criteria are met, and both parents and kid are phased or parents are HET at different sites, the priority is 1.</li>
<li>If both parents are not phased, the priority is 2.</li>
<li>For every parent that&#8217;s a het at both sites, the priority is incremented by 1.</li>
<li>The priority in a family is the minimum found among all kids.</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Each pair of consecutive lines in the output represent the two variants
for a compound heterozygote in a give sample.  The third column,
<cite>comp_het_id</cite>, tracks the distinct compound heterozygote variant pairs.</p>
</div>
<p>Example usage with a subset of columns:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini comp_hets my.db --columns <span class="s2">&quot;chrom, start, end&quot;</span> test.comp_het_default.2.db
chrom   start   end gene    alt variant_id  family_id   family_members  family_genotypes    samples family_count    comp_het_id
chr1    17362   17366   WASH7P  T   1   3   dad_3<span class="o">(</span>dad;unaffected<span class="o">)</span>,mom_3<span class="o">(</span>mom;unaffected<span class="o">)</span>,child_3<span class="o">(</span>child;affected<span class="o">)</span> TTCT|T,TTCT|TTCT,TTCT|T child_3 2   1
chr1    17729   17730   WASH7P  A   2   3   dad_3<span class="o">(</span>dad;unaffected<span class="o">)</span>,mom_3<span class="o">(</span>mom;unaffected<span class="o">)</span>,child_3<span class="o">(</span>child;affected<span class="o">)</span> C|A,C|A,A|C child_3 2   1
chr1    17362   17366   WASH7P  T   1   4   dad_4<span class="o">(</span>dad;unaffected<span class="o">)</span>,mom_4<span class="o">(</span>mom;unaffected<span class="o">)</span>,child_4<span class="o">(</span>child;affected<span class="o">)</span> TTCT|T,TTCT|TTCT,TTCT|T child_4 2   1
chr1    17729   17730   WASH7P  A   2   4   dad_4<span class="o">(</span>dad;unaffected<span class="o">)</span>,mom_4<span class="o">(</span>mom;unaffected<span class="o">)</span>,child_4<span class="o">(</span>child;affected<span class="o">)</span> C|A,C|A,A|C child_4 2   1
</pre></div>
</div>
<p>This indicates that samples child_3 and child_4 have a candidate compound heterozygotes in WASH.</p>
<p>the following command would further restrict candidate genes to those genes with a compound heterozygote in at least two families:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini comp_hets -d 50 <span class="se">\</span>
      --columns <span class="s2">&quot;chrom, start, end, ref, alt&quot;</span> <span class="se">\</span>
      --filter <span class="s2">&quot;impact_severity = &#39;HIGH&#39;&quot;</span> <span class="se">\</span>
      --allow-unaffected <span class="se">\</span>
      --min-kindreds 2 <span class="se">\</span>
      my.db
</pre></div>
</div>
<p>Now, this does not require that the family members are necessarily restricted to solely
those that are affected. To impose this restriction, we remove the <tt class="docutils literal"><span class="pre">--allow-unaffected</span></tt>
flag</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini comp_hets -d 50 <span class="se">\</span>
      --columns <span class="s2">&quot;chrom, start, end, ref, alt&quot;</span> <span class="se">\</span>
      --filter <span class="s2">&quot;impact_severity = &#39;HIGH&#39;&quot;</span> <span class="se">\</span>
      --min-kindreds 2 <span class="se">\</span>
      my.db
</pre></div>
</div>
<p>We may also specify the families of interest:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini comp_hets --families 1 my.db
<span class="nv">$ </span>gemini comp_hets --families 1,7 my.db
</pre></div>
</div>
</div>
</div>
<div class="section" id="mendelian-error-identify-non-mendelian-transmission">
<h2><tt class="docutils literal"><span class="pre">mendelian_error</span></tt>: Identify non-mendelian transmission.<a class="headerlink" href="#mendelian-error-identify-non-mendelian-transmission" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>This tool requires that you identify familial relationships via a PED file
when loading your VCF into gemini via:</p>
<p class="last"><tt class="docutils literal"><span class="pre">gemini</span> <span class="pre">load</span> <span class="pre">-v</span> <span class="pre">my.vcf</span> <span class="pre">-p</span> <span class="pre">my.ped</span> <span class="pre">my.db</span></tt></p>
</div>
<p>We can query for mendelian errors in trios including:</p>
<ul class="simple">
<li>loss of heterozygosity</li>
<li>implausible de-novo mutations</li>
<li>de-novo mutations</li>
<li>uniparental disomy</li>
</ul>
<div class="section" id="id1">
<h3>Genotype Requirements<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>(LOH) kind and one parent are opposite homozygotes; other parent is HET</li>
<li>(uniparental disomy) parents are opposite homozygotes; kid is homozygote;</li>
<li>(plausible de novo) kid is het. parents are same homozygotes</li>
<li>(implausible de novo) kid is homozygoes. parents are same homozygotes and opposite to kid.</li>
</ul>
<p>If allow <cite>&#8211;only-affected</cite> is used, then the tools will only consider samples that have parents
<strong>and</strong> are affected. The default is to consider any sample with parents.</p>
<p>This tool will report the probability of a mendelian error in the final column
that is derived from the genotype likelihoods if they are available.</p>
<p>Example:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini mendel_errors --columns <span class="s2">&quot;chrom,start,end&quot;</span> test.mendel.db --gt-pl-max 1
chrom       start   end     variant_id      family_id       family_members  family_genotypes        samples family_count    violation       violation_prob
chr1        10670   10671   1       CEPH1463        NA12889<span class="o">(</span>dad;unknown<span class="o">)</span>,NA12890<span class="o">(</span>mom;unknown<span class="o">)</span>,NA12877<span class="o">(</span>child;unknown<span class="o">)</span>        G/G,G/G,G/C     NA12877 1       plausible de novo       0.962
chr1        28493   28494   2       CEPH1463        NA12889<span class="o">(</span>dad;unknown<span class="o">)</span>,NA12890<span class="o">(</span>mom;unknown<span class="o">)</span>,NA12877<span class="o">(</span>child;unknown<span class="o">)</span>        T/C,T/T,C/C     NA12877 1       loss of heterozygosity  0.660
chr1        28627   28628   3       CEPH1463        NA12889<span class="o">(</span>dad;unknown<span class="o">)</span>,NA12890<span class="o">(</span>mom;unknown<span class="o">)</span>,NA12877<span class="o">(</span>child;unknown<span class="o">)</span>        C/C,C/C,C/T     NA12877 1       plausible de novo       0.989
chr1        267558  267560  5       CEPH1463        NA12889<span class="o">(</span>dad;unknown<span class="o">)</span>,NA12890<span class="o">(</span>mom;unknown<span class="o">)</span>,NA12877<span class="o">(</span>child;unknown<span class="o">)</span>        C/C,C/C,CT/C    NA12877 1       plausible de novo       0.896
chr1        537969  537970  7       CEPH1463        NA12889<span class="o">(</span>dad;unknown<span class="o">)</span>,NA12890<span class="o">(</span>mom;unknown<span class="o">)</span>,NA12877<span class="o">(</span>child;unknown<span class="o">)</span>        C/C,C/C,C/T     NA12877 1       plausible de novo       0.928
chr1        547518  547519  11      CEPH1463        NA12889<span class="o">(</span>dad;unknown<span class="o">)</span>,NA12890<span class="o">(</span>mom;unknown<span class="o">)</span>,NA12877<span class="o">(</span>child;unknown<span class="o">)</span>        G/G,G/G,G/T     NA12877 1       plausible de novo       1.000
chr1        589081  589086  14      CEPH1463        NA12889<span class="o">(</span>dad;unknown<span class="o">)</span>,NA12890<span class="o">(</span>mom;unknown<span class="o">)</span>,NA12877<span class="o">(</span>child;unknown<span class="o">)</span>        G/G,GAGAA/GAGAA,G/G     NA12877 1       uniparental disomy      0.940
chr1        749688  749689  16      CEPH1463        NA12889<span class="o">(</span>dad;unknown<span class="o">)</span>,NA12890<span class="o">(</span>mom;unknown<span class="o">)</span>,NA12877<span class="o">(</span>child;unknown<span class="o">)</span>        T/T,T/T,G/G     NA12877 1       implausible de novo     0.959
chr1        788944  788945  17      CEPH1463        NA12889<span class="o">(</span>dad;unknown<span class="o">)</span>,NA12890<span class="o">(</span>mom;unknown<span class="o">)</span>,NA12877<span class="o">(</span>child;unknown<span class="o">)</span>        C/C,G/G,G/G     NA12877 1       uniparental disomy      0.914
chr1        1004248 1004249 22      CEPH1463        NA12889<span class="o">(</span>dad;unknown<span class="o">)</span>,NA12890<span class="o">(</span>mom;unknown<span class="o">)</span>,NA12877<span class="o">(</span>child;unknown<span class="o">)</span>        G/G,G/G,G/C     NA12877 1       plausible de novo       1.000
</pre></div>
</div>
<p>Where, here, we have required the called genotype to have at most a PL of 1 (lower is more confident).
Note that the &#8220;violation&#8221; column indicates the type of mendelian error and the final column can be used for further filtering,
with higher numbers indicating a greater probability of mendelian error. We have found &gt; 0.99 to be a reasonable
cutoff.</p>
<p>Arguments are similar to the other tools:</p>
<div class="highlight-bash"><div class="highlight"><pre>positional arguments:
  db                    The name of the database to be queried.

optional arguments:
  -h, --help            show this help message and exit
  --columns STRING      A list of columns that you would like returned. Def. =
                        &quot;*&quot;
  --filter STRING       Restrictions to apply to variants (SQL syntax)
  --min-kindreds MIN_KINDREDS
                        The min. number of kindreds that must have a candidate
                        variant in a gene.
  --families FAMILIES   Restrict analysis to a specific set of 1 or more
                        (comma) separated) families
  -d MIN_SAMPLE_DEPTH   The minimum aligned sequence depth (genotype DP) req&#39;d
                        for each sample (def. = 0)
  --gt-pl-max GT_PHRED_LL
                        The maximum phred-scaled genotype likelihod (PL)
                        allowed for each sample.
  --allow-unaffected    consider candidates that also appear in unaffected samples.
</pre></div>
</div>
</div>
</div>
<div class="section" id="de-novo-identifying-potential-de-novo-mutations">
<h2><tt class="docutils literal"><span class="pre">de_novo</span></tt>: Identifying potential de novo mutations.<a class="headerlink" href="#de-novo-identifying-potential-de-novo-mutations" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>1. This tool requires that you identify familial relationships via a PED file
when loading your VCF into gemini via:</p>
<p class="last"><tt class="docutils literal"><span class="pre">gemini</span> <span class="pre">load</span> <span class="pre">-v</span> <span class="pre">my.vcf</span> <span class="pre">-p</span> <span class="pre">my.ped</span> <span class="pre">my.db</span></tt></p>
</div>
<div class="section" id="id2">
<h3>Genotype Requirements<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>all affecteds must be het</li>
<li>[affected] all unaffected must be homref or homalt</li>
<li>at least 1 affected kid must have unaffected parents</li>
<li>[strict] if an affected has affected parents, it&#8217;s not de_novo</li>
<li>[strict] all affected kids must have unaffected (or no) parents</li>
<li>[strict] warning if none of the affected samples have parents.</li>
</ul>
<p>The last 3 items, prefixed with [strict] can be turned off with <cite>&#8211;lenient</cite></p>
<p>If <cite>&#8211;allow-unaffected</cite> is specified, then the item prefixed [affected] is not
required.</p>
<p><cite>Example PED file format for GEMINI</cite></p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c">#Family_ID      Individual_ID   Paternal_ID     Maternal_ID     Sex     Phenotype       Ethnicity</span>
1       S173    S238    S239    1       2       caucasian
1       S238    -9      -9      1       1       caucasian
1       S239    -9      -9      2       1       caucasian
2       S193    S230    S231    1       2       caucasian
2       S230    -9      -9      1       1       caucasian
2       S231    -9      -9      2       1       caucasian
3       S242    S243    S244    1       2       caucasian
3       S243    -9      -9      1       1       caucasian
3       S244    -9      -9      2       1       caucasian
4       S253    S254    S255    1       2       caucasianNEuropean
4       S254    -9      -9      1       1       caucasianNEuropean
4       S255    -9      -9      2       1       caucasianNEuropean
</pre></div>
</div>
<p>Assuming you have defined the familial relationships between samples when loading
your VCF into GEMINI, one can leverage a built-in tool for identifying de novo
(a.k.a spontaneous) mutations that arise in offspring.</p>
</div>
<div class="section" id="example">
<h3>example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini de_novo --columns <span class="s2">&quot;chrom,start,end&quot;</span> test.de_novo.db
chrom       start   end     variant_id      family_id       family_members  family_genotypes        samples family_count
chr10       1142207 1142208 1       1       1_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,1_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,1_kid<span class="o">(</span>child;affected<span class="o">)</span>       T/T,T/T,T/C     1_kid   1
chr10       48003991        48003992        2       2       2_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,2_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,2_kid<span class="o">(</span>child;affected<span class="o">)</span>       C/C,C/C,C/T     2_kid   1
chr10       48004991        48004992        3       3       3_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,3_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,3_kid<span class="o">(</span>child;affected<span class="o">)</span>       C/C,C/C,C/T     3_kid   1
chr10       135336655       135336656       4       4       1_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,1_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,1_kid<span class="o">(</span>child;affected<span class="o">)</span>       G/G,G/G,G/A     1_kid   2
chr10       135336655       135336656       4       4       2_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,2_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,2_kid<span class="o">(</span>child;affected<span class="o">)</span>       G/G,G/G,G/A     2_kid   2
chr10       135369531       135369532       5       5       1_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,1_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,1_kid<span class="o">(</span>child;affected<span class="o">)</span>       T/T,T/T,T/C     1_kid   3
chr10       135369531       135369532       5       5       2_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,2_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,2_kid<span class="o">(</span>child;affected<span class="o">)</span>       T/T,T/T,T/C     2_kid   3
chr10       135369531       135369532       5       5       3_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,3_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,3_kid<span class="o">(</span>child;affected<span class="o">)</span>       T/T,T/T,T/C     3_kid   3
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The output will always start with the the requested columns followed by
the 5 columns enumerated at the start of this document.</p>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini de_novo -d 50 --columns <span class="s2">&quot;chrom,start,end&quot;</span> test.de_novo.db
chrom       start   end     variant_id      family_id       family_members  family_genotypes        samples family_count
chr10       135369531       135369532       5       5       3_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,3_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,3_kid<span class="o">(</span>child;affected<span class="o">)</span>       T/T,T/T,T/C     3_kid   1
</pre></div>
</div>
</div>
<div class="section" id="id3">
<h3><tt class="docutils literal"><span class="pre">example</span></tt><a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<p>if we wanted to restrict candidate variants
to solely those with a HIGH predicted functional consequence, we could use the
following:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini de_novo <span class="se">\</span>
      --columns <span class="s2">&quot;chrom, start, end, ref, alt&quot;</span> <span class="se">\</span>
      --filter <span class="s2">&quot;impact_severity = &#39;HIGH&#39;&quot;</span> <span class="se">\</span>
      test.de_novo.db
chrom       start   end     ref     alt     variant_id      family_id       family_members  family_genotypes        samples family_count
chr10       1142207 1142208 T       C       1       1       1_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,1_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,1_kid<span class="o">(</span>child;affected<span class="o">)</span>       T/T,T/T,T/C     1_kid   1
</pre></div>
</div>
</div>
<div class="section" id="id4">
<h3><tt class="docutils literal"><span class="pre">example</span></tt><a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<p>the following command would further restrict candidate genes to those genes with a de novo variant in at least two families:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini de_novo <span class="se">\</span>
      --columns <span class="s2">&quot;chrom, start, end, ref, alt&quot;</span> <span class="se">\</span>
      --filter <span class="s2">&quot;impact_severity = &#39;HIGH&#39;&quot;</span> <span class="se">\</span>
      --min-kindreds 2 <span class="se">\</span>
      test.de_novo.db
</pre></div>
</div>
</div>
<div class="section" id="id5">
<h3><tt class="docutils literal"><span class="pre">example</span></tt><a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<p>By default, candidate de novo variants are reported for families
in the database.  One can restrict the analysis to variants in
specific familes with the <tt class="docutils literal"><span class="pre">--families</span></tt> option.  Families should be provided
as a comma-separated list</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini de_novo --families 1 my.db
<span class="nv">$ </span>gemini de_novo --families 1,7 my.db
</pre></div>
</div>
</div>
</div>
<div class="section" id="autosomal-recessive-find-variants-meeting-an-autosomal-recessive-model">
<h2><tt class="docutils literal"><span class="pre">autosomal_recessive</span></tt>: Find variants meeting an autosomal recessive model.<a class="headerlink" href="#autosomal-recessive-find-variants-meeting-an-autosomal-recessive-model" title="Permalink to this headline">¶</a></h2>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p>By default, this tool requires that you identify familial relationships
via a PED file when loading your VCF into GEMINI.  For example:</p>
<p><tt class="docutils literal"><span class="pre">gemini</span> <span class="pre">load</span> <span class="pre">-v</span> <span class="pre">my.vcf</span> <span class="pre">-p</span> <span class="pre">my.ped</span> <span class="pre">my.db</span></tt></p>
<p class="last">However, in the absence of established parent/child relationships in the PED
file, GEMINI will issue a WARNING, yet will attempt to identify autosomal
recessive candidates for all samples marked as &#8220;affected&#8221;.</p>
</div>
<div class="section" id="id6">
<h3>Genotype Requirements<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>all affecteds must be hom_alt</li>
<li>[affected] no unaffected can be hom_alt (can be unknown)</li>
<li>[strict] if parents exist they must be unaffected and het for all affected kids</li>
<li>[strict] if there are no affecteds that have a parent, a warning is issued.</li>
</ul>
<p>if <cite>&#8211;lenient</cite> is specified, the 2 points prefixed with &#8220;[strict]&#8221; are not required.</p>
<p>if <cite>&#8211;allow-unaffected</cite> is specified, the point prefix with &#8220;[affected]&#8221; is not required.</p>
</div>
<div class="section" id="default-behavior">
<h3><tt class="docutils literal"><span class="pre">default</span> <span class="pre">behavior</span></tt><a class="headerlink" href="#default-behavior" title="Permalink to this headline">¶</a></h3>
<p>Assuming you have defined the familial relationships between samples when
loading your VCF into GEMINI, one can leverage a built-in tool for
identifying variants that meet an autosomal recessive inheritance pattern.
The reported variants will be restricted to those variants having the
potential to impact the function of affecting protein coding transcripts.</p>
<p>For the following examples, let&#8217;s assume we have a PED file for 3 different
families as follows (the kids are affected in each family, but the parents
are not):</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>cat families.ped
1   1_dad   0       0       -1      1
1   1_mom   0       0       -1      1
1   1_kid   1_dad   1_mom   -1      2
2   2_dad   0       0       -1      1
2   2_mom   0       0       -1      1
2   2_kid   2_dad   2_mom   -1      2
3   3_dad   0       0       -1      1
3   3_mom   0       0       -1      1
3   3_kid   3_dad   3_mom   -1      2
</pre></div>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini autosomal_recessive test.auto_rec.db --columns <span class="s2">&quot;chrom,start,end,gene&quot;</span>
chrom       start   end     gene    variant_id      family_id       family_members  family_genotypes        samples family_count
chr10       48003991        48003992        ASAH2C  2       2       1_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,1_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,1_kid<span class="o">(</span>child;affected<span class="o">)</span>       C/T,C/T,T/T     1_kid   1
chr10       48004991        48004992        ASAH2C  3       3       2_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,2_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,2_kid<span class="o">(</span>child;affected<span class="o">)</span>       C/T,C/T,T/T     2_kid   1
chr10       135369531       135369532       SYCE1   5       5       3_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,3_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,3_kid<span class="o">(</span>child;affected<span class="o">)</span>       T/C,T/C,C/C     3_kid   1
chr10       1142207 1142208 WDR37   1       1       1_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,1_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,1_kid<span class="o">(</span>child;affected<span class="o">)</span>       T/C,T/C,C/C     1_kid   2
chr10       1142207 1142208 WDR37   1       1       2_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,2_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,2_kid<span class="o">(</span>child;affected<span class="o">)</span>       T/C,T/C,C/C     2_kid   2
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The output will always start with the requested columns and end with the 5 extra columns
enumerated at the start of this document.</p>
</div>
<p>To restrict the report to genes with variants (doesn&#8217;t have
to be the _same_ variant) observed in at least two kindreds, use the following:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini autosomal_recessive <span class="se">\</span>
    --columns <span class="s2">&quot;gene, chrom, start, end, ref, alt, impact, impact_severity&quot;</span> <span class="se">\</span>
    --min-kindreds 2 <span class="se">\</span>
    test.auto_rec.db
gene        chrom   start   end     ref     alt     impact  impact_severity variant_id      family_id       family_members  family_genotypes        samples family_count
ASAH2C      chr10   48003991        48003992        C       T       non_syn_coding  MED     2       2       1_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,1_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,1_kid<span class="o">(</span>child;affected<span class="o">)</span>       C/T,C/T,T/T     1_kid   1
ASAH2C      chr10   48004991        48004992        C       T       non_syn_coding  MED     3       3       2_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,2_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,2_kid<span class="o">(</span>child;affected<span class="o">)</span>       C/T,C/T,T/T     2_kid   1
WDR37       chr10   1142207 1142208 T       C       stop_loss       HIGH    1       1       1_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,1_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,1_kid<span class="o">(</span>child;affected<span class="o">)</span>       T/C,T/C,C/C     1_kid   2
WDR37       chr10   1142207 1142208 T       C       stop_loss       HIGH    1       1       2_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,2_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,2_kid<span class="o">(</span>child;affected<span class="o">)</span>       T/C,T/C,C/C     2_kid   2
</pre></div>
</div>
<p>to report only those with a HIGH predicted functional consequence, we could use the
following:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini autosomal_recessive <span class="se">\</span>
    --columns <span class="s2">&quot;gene, chrom, start, end, ref, alt, impact, impact_severity&quot;</span> <span class="se">\</span>
    --min-kindreds 2 <span class="se">\</span>
    --filter <span class="s2">&quot;impact_severity = &#39;HIGH&#39;&quot;</span> <span class="se">\</span>
    test.auto_rec.db
gene        chrom   start   end     ref     alt     impact  impact_severity variant_id      family_id       family_members  family_genotypes        samples family_count
WDR37       chr10   1142207 1142208 T       C       stop_loss       HIGH    1       1       1_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,1_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,1_kid<span class="o">(</span>child;affected<span class="o">)</span>       T/C,T/C,C/C     1_kid   2
WDR37       chr10   1142207 1142208 T       C       stop_loss       HIGH    1       1       2_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,2_mom<span class="o">(</span>mom;unaffected<span class="o">)</span>,2_kid<span class="o">(</span>child;affected<span class="o">)</span>       T/C,T/C,C/C     2_kid   2
</pre></div>
</div>
<p>To limit to confidently called genotypes:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini autosomal_dominant <span class="se">\</span>
    --columns <span class="s2">&quot;gene, chrom, start, end, ref, alt, impact, impact_severity&quot;</span> <span class="se">\</span>
    --filter <span class="s2">&quot;impact_severity = &#39;HIGH&#39;&quot;</span> <span class="se">\</span>
    --min-kindreds 1 <span class="se">\</span>
    --gt-pl-max 10 <span class="se">\</span>
    my.db
</pre></div>
</div>
</div>
</div>
<div class="section" id="autosomal-dominant-find-variants-meeting-an-autosomal-dominant-model">
<h2><tt class="docutils literal"><span class="pre">autosomal_dominant</span></tt>: Find variants meeting an autosomal dominant model.<a class="headerlink" href="#autosomal-dominant-find-variants-meeting-an-autosomal-dominant-model" title="Permalink to this headline">¶</a></h2>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p>0. version 0.16.0 changes the behavior of this tool to be more strict.
To regain more lenient behavior, specify &#8211;lenient and &#8211;allow-unaffected.</p>
<p>By default, this tool requires that you identify familial relationships
via a PED file when loading your VCF into GEMINI.  For example:</p>
<p class="last"><tt class="docutils literal"><span class="pre">gemini</span> <span class="pre">load</span> <span class="pre">-v</span> <span class="pre">my.vcf</span> <span class="pre">-p</span> <span class="pre">my.ped</span> <span class="pre">my.db</span></tt></p>
</div>
<div class="section" id="id7">
<h3>Genotype Requirements<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>All affecteds must be het</li>
<li>[affected] No unaffected can be het or homalt (can be unknown)</li>
<li>de_novo mutations are not auto_dom (at least not in the first generation)</li>
<li>At least 1 affected must have 1 affected parent.</li>
<li>[strict] All affecteds must have parents with known phenotype.</li>
<li>[strict] All affected kids must have at least 1 affected parent</li>
<li>[strict] If no affected has a parent, a warning is issued.</li>
</ul>
<p>If <cite>&#8211;lenient</cite> is specified, the items prefixed with &#8220;[strict]&#8221; are not required.</p>
<p>If <cite>&#8211;allow-unaffected</cite> is specified, the item prefix with &#8220;[affected]&#8221; is not required.</p>
</div>
<div class="section" id="id8">
<h3><tt class="docutils literal"><span class="pre">default</span> <span class="pre">behavior</span></tt><a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<p>For the following examples, let&#8217;s assume we have a PED file for 3 different
families as follows (the kids are affected in each family, but the parents
are not):</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>cat families.ped
1   1_dad   0       0       -1      1
1   1_mom   0       0       -1      1
1   1_kid   1_dad   1_mom   -1      2
2   2_dad   0       0       -1      1
2   2_mom   0       0       -1      2
2   2_kid   2_dad   2_mom   -1      2
3   3_dad   0       0       -1      2
3   3_mom   0       0       -1      -9
3   3_kid   3_dad   3_mom   -1      2
</pre></div>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini autosomal_dominant test.auto_dom.db --columns <span class="s2">&quot;chrom,start,end,gene&quot;</span>
chrom       start   end     gene    variant_id      family_id       family_members  family_genotypes        samples family_count
chr10       48003991        48003992        ASAH2C  3       3       2_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,2_mom<span class="o">(</span>mom;affected<span class="o">)</span>,2_kid<span class="o">(</span>child;affected<span class="o">)</span> C/C,C/T,C/T     2_mom,2_kid     2
chr10       48004991        48004992        ASAH2C  4       4       2_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,2_mom<span class="o">(</span>mom;affected<span class="o">)</span>,2_kid<span class="o">(</span>child;affected<span class="o">)</span> C/C,C/T,C/T     2_mom,2_kid     2
chr10       48003991        48003992        ASAH2C  3       3       3_dad<span class="o">(</span>dad;affected<span class="o">)</span>,3_mom<span class="o">(</span>mom;unknown<span class="o">)</span>,3_kid<span class="o">(</span>child;affected<span class="o">)</span>    C/T,C/C,C/T     3_dad,3_kid     2
chr10       48004991        48004992        ASAH2C  4       4       3_dad<span class="o">(</span>dad;affected<span class="o">)</span>,3_mom<span class="o">(</span>mom;unknown<span class="o">)</span>,3_kid<span class="o">(</span>child;affected<span class="o">)</span>    C/T,C/C,C/T     3_dad,3_kid     2
chr10       135336655       135336656       SPRN    5       5       3_dad<span class="o">(</span>dad;affected<span class="o">)</span>,3_mom<span class="o">(</span>mom;unknown<span class="o">)</span>,3_kid<span class="o">(</span>child;affected<span class="o">)</span>    G/A,G/G,G/A     3_dad,3_kid     1
chr10       1142207 1142208 WDR37   1       1       2_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,2_mom<span class="o">(</span>mom;affected<span class="o">)</span>,2_kid<span class="o">(</span>child;affected<span class="o">)</span> T/T,T/C,T/C     2_mom,2_kid     2
chr10       1142207 1142208 WDR37   1       1       3_dad<span class="o">(</span>dad;affected<span class="o">)</span>,3_mom<span class="o">(</span>mom;unknown<span class="o">)</span>,3_kid<span class="o">(</span>child;affected<span class="o">)</span>    T/C,T/T,T/C     3_dad,3_kid     2
</pre></div>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini autosomal_dominant <span class="se">\</span>
    --columns <span class="s2">&quot;gene, chrom, start, end, ref, alt, impact, impact_severity&quot;</span> <span class="se">\</span>
    --min-kindreds 2 <span class="se">\</span>
    test.auto_dom.db
gene        chrom   start   end     ref     alt     impact  impact_severity variant_id      family_id       family_members  family_genotypes        samples family_count
ASAH2C      chr10   48003991        48003992        C       T       non_syn_coding  MED     3       3       2_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,2_mom<span class="o">(</span>mom;affected<span class="o">)</span>,2_kid<span class="o">(</span>child;affected<span class="o">)</span> C/C,C/T,C/T     2_mom,2_kid     2
ASAH2C      chr10   48004991        48004992        C       T       non_syn_coding  MED     4       4       2_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,2_mom<span class="o">(</span>mom;affected<span class="o">)</span>,2_kid<span class="o">(</span>child;affected<span class="o">)</span> C/C,C/T,C/T     2_mom,2_kid     2
ASAH2C      chr10   48003991        48003992        C       T       non_syn_coding  MED     3       3       3_dad<span class="o">(</span>dad;affected<span class="o">)</span>,3_mom<span class="o">(</span>mom;unknown<span class="o">)</span>,3_kid<span class="o">(</span>child;affected<span class="o">)</span>    C/T,C/C,C/T     3_dad,3_kid     2
ASAH2C      chr10   48004991        48004992        C       T       non_syn_coding  MED     4       4       3_dad<span class="o">(</span>dad;affected<span class="o">)</span>,3_mom<span class="o">(</span>mom;unknown<span class="o">)</span>,3_kid<span class="o">(</span>child;affected<span class="o">)</span>    C/T,C/C,C/T     3_dad,3_kid     2
WDR37       chr10   1142207 1142208 T       C       stop_loss       HIGH    1       1       2_dad<span class="o">(</span>dad;unaffected<span class="o">)</span>,2_mom<span class="o">(</span>mom;affected<span class="o">)</span>,2_kid<span class="o">(</span>child;affected<span class="o">)</span> T/T,T/C,T/C     2_mom,2_kid     2
WDR37       chr10   1142207 1142208 T       C       stop_loss       HIGH    1       1       3_dad<span class="o">(</span>dad;affected<span class="o">)</span>,3_mom<span class="o">(</span>mom;unknown<span class="o">)</span>,3_kid<span class="o">(</span>child;affected<span class="o">)</span>    T/C,T/T,T/C     3_dad,3_kid     2
</pre></div>
</div>
</div>
</div>
<div class="section" id="pathways-map-genes-and-variants-to-kegg-pathways">
<h2><tt class="docutils literal"><span class="pre">pathways</span></tt>: Map genes and variants to KEGG pathways.<a class="headerlink" href="#pathways-map-genes-and-variants-to-kegg-pathways" title="Permalink to this headline">¶</a></h2>
<p>Mapping genes to biological pathways is useful in understanding the
function/role played by a gene. Likewise, genes involved in common pathways
is helpful in understanding heterogeneous diseases. We have integrated
the KEGG pathway mapping for gene variants, to explain/annotate variation.
This requires your VCF be annotated with either snpEff/VEP.</p>
<p>Examples:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini pathways -v 68 example.db
chrom   start   end     ref     alt     impact  sample  genotype        gene    transcript      pathway
chr10   52004314        52004315        T       C       intron  M128215 C/C     ASAH2   ENST00000395526 hsa00600:Sphingolipid_metabolism,hsa01100:Metabolic_pathways
chr10   126678091       126678092       G       A       stop_gain       M128215 G/A     CTBP2   ENST00000531469 hsa05220:Chronic_myeloid_leukemia,hsa04310:Wnt_signaling_pathway,hsa04330:Notch_signaling_pathway,hsa05200:Pathways_in_cancer
chr16   72057434        72057435        C       T       non_syn_coding  M10475  C/T     DHODH   ENST00000219240 hsa01100:Metabolic_pathways,hsa00240:Pyrimidine_metabolism
</pre></div>
</div>
<p>Here, -v specifies the version of the Ensembl genes used to build the KEGG
pathway map. Hence, use versions that match the VEP/snpEff versions of the
annotated vcf for correctness. For e.g VEP v2.6 and snpEff v3.1 use Ensembl
68 version of the genomes.</p>
<p>We currently support versions 66 through 71 of the Ensembl genes</p>
<div class="section" id="lof">
<h3><tt class="docutils literal"><span class="pre">--lof</span></tt><a class="headerlink" href="#lof" title="Permalink to this headline">¶</a></h3>
<p>By default, all gene variants that map to pathways are reported.  However,
one may want to restrict the analysis to LoF variants using the <tt class="docutils literal"><span class="pre">--lof</span></tt> option.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini pathways --lof -v 68 example.db
chrom   start   end     ref     alt     impact  sample  genotype        gene    transcript      pathway
chr10   126678091       126678092       G       A       stop_gain       M128215 G/A     CTBP2   ENST00000531469 hsa05220:Chronic_myeloid_leukemia,hsa04310:Wnt_signaling_pathway,hsa04330:Notch_signaling_pathway,hsa05200:Pathways_in_cancer
</pre></div>
</div>
</div>
</div>
<div class="section" id="interactions-find-genes-among-variants-that-are-interacting-partners">
<h2><tt class="docutils literal"><span class="pre">interactions</span></tt>: Find genes among variants that are interacting partners.<a class="headerlink" href="#interactions-find-genes-among-variants-that-are-interacting-partners" title="Permalink to this headline">¶</a></h2>
<p>Integrating the knowledge of the known protein-protein interactions would be
useful in explaining variation data. Meaning to say that a damaging variant
in an interacting partner of a  potential protein may be equally interesting
as the protein itself. We have used the HPRD binary interaction data to build
a p-p network graph which can be explored by GEMINI.</p>
<p>Examples:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini interactions -g CTBP2 -r 3 example.db
sample  gene    order_of_interaction    interacting_gene
M128215 CTBP2   0_order:        CTBP2
M128215 CTBP2   1_order:        RAI2
M128215 CTBP2   2_order:        RB1
M128215 CTBP2   3_order:        TGM2,NOTCH2NL
</pre></div>
</div>
<p>Return CTBP2 (-g) interacting gene variants till the third order (-r)</p>
<div class="section" id="lof-interactions">
<h3><tt class="docutils literal"><span class="pre">lof_interactions</span></tt><a class="headerlink" href="#lof-interactions" title="Permalink to this headline">¶</a></h3>
<p>Use this option to restrict your analysis to only LoF variants.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini lof_interactions -r 3 example.db
sample  lof_gene        order_of_interaction    interacting_gene
M128215 TGM2    1_order:        RB1
M128215 TGM2    2_order:        none
M128215 TGM2    3_order:        NOTCH2NL,CTBP2
</pre></div>
</div>
<p>Meaning to say return all LoF gene TGM2 (in sample M128215) interacting
partners to a 3rd order of interaction.</p>
</div>
<div class="section" id="var">
<h3><tt class="docutils literal"><span class="pre">--var</span></tt><a class="headerlink" href="#var" title="Permalink to this headline">¶</a></h3>
<p>An extended variant information (chrom, start, end etc.) for the interacting gene
may be achieved with the &#8211;var option for both the <tt class="docutils literal"><span class="pre">interactions</span></tt> and the
<tt class="docutils literal"><span class="pre">lof_interactions</span></tt></p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini interactions -g CTBP2 -r 3 --var example.db
sample  gene    order_of_interaction    interacting_gene        var_id  chrom   start   end     impact  biotype in_dbsnp        clinvar_sig     clinvar_disease_name    aaf_1kg_all     aaf_esp_all
M128215 CTBP2   0       CTBP2   5       chr10   126678091       126678092       stop_gain       protein_coding  1       None    None    None    None
M128215 CTBP2   1       RAI2    9       chrX    17819376        17819377        non_syn_coding  protein_coding  1       None    None    1       0.000473
M128215 CTBP2   2       RB1     7       chr13   48873834        48873835        upstream        protein_coding  1       None    None    0.94    None
M128215 CTBP2   3       NOTCH2NL        1       chr1    145273344       145273345       non_syn_coding  protein_coding  1       None    None    None    None
M128215 CTBP2   3       TGM2    8       chr20   36779423        36779424        stop_gain       protein_coding  0       None    None    None    None
</pre></div>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini lof_interactions -r 3 --var example.db
sample  lof_gene        order_of_interaction    interacting_gene        var_id  chrom   start   end     impact  biotype in_dbsnp        clinvar_sig     clinvar_disease_name    aaf_1kg_all     aaf_esp_all
M128215 TGM2    1       RB1     7       chr13   48873834        48873835        upstream        protein_coding  1       None    None    0.94    None
M128215 TGM2    3       NOTCH2NL        1       chr1    145273344       145273345       non_syn_coding  protein_coding  1       None    None    None    None
M128215 TGM2    3       CTBP2   5       chr10   126678091       126678092       stop_gain       protein_coding  1       None    None    None    None
</pre></div>
</div>
</div>
</div>
<div class="section" id="lof-sieve-filter-lof-variants-by-transcript-position-and-type">
<h2><tt class="docutils literal"><span class="pre">lof_sieve</span></tt>: Filter LoF variants by transcript position and type<a class="headerlink" href="#lof-sieve-filter-lof-variants-by-transcript-position-and-type" title="Permalink to this headline">¶</a></h2>
<p>Not all candidate LoF variants are created equal. For e.g, a nonsense (stop gain)
variant impacting the first 5% of a polypeptide is far more likely to be deleterious
than one affecting the last 5%. Assuming you&#8217;ve annotated your VCF with snpEff v3.0+,
the lof_sieve tool reports the fractional position (e.g. 0.05 for the first 5%) of
the mutation in the amino acid sequence. In addition, it also reports the predicted
function of the transcript so that one can segregate candidate LoF variants that
affect protein_coding transcripts from processed RNA, etc.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini lof_sieve chr22.low.exome.snpeff.100samples.vcf.db
chrom   start   end ref alt highest_impact  aa_change   var_trans_pos   trans_aa_length var_trans_pct   sample  genotype    gene    transcript  trans_type
chr22   17072346    17072347    C   T   stop_gain   W365*   365 557 0.655296229803  NA19327 C|T CCT8L2  ENST00000359963 protein_coding
chr22   17072346    17072347    C   T   stop_gain   W365*   365 557 0.655296229803  NA19375 T|C CCT8L2  ENST00000359963 protein_coding
chr22   17129539    17129540    C   T   splice_donor    None    None    None    None    NA18964 T|C TPTEP1  ENST00000383140 lincRNA
chr22   17129539    17129540    C   T   splice_donor    None    None    None    None    NA19675 T|C TPTEP1  ENST00000383140 lincRNA
</pre></div>
</div>
</div>
<div class="section" id="annotate-adding-your-own-custom-annotations">
<h2><tt class="docutils literal"><span class="pre">annotate</span></tt>: adding your own custom annotations<a class="headerlink" href="#annotate-adding-your-own-custom-annotations" title="Permalink to this headline">¶</a></h2>
<p>It is inevitable that researchers will want to enhance the gemini framework with
their own, custom annotations. <tt class="docutils literal"><span class="pre">gemini</span></tt> provides a sub-command called
<tt class="docutils literal"><span class="pre">annotate</span></tt> for exactly this purpose. As long as you provide a <tt class="docutils literal"><span class="pre">tabix</span></tt>&#8216;ed
annotation file in BED or VCF format, the <tt class="docutils literal"><span class="pre">annotate</span></tt> tool will, for each
variant in the variants table, screen for overlaps in your annotation file and
update a one or more new column in the variants table that you may specify on the command
line. This is best illustrated by example.</p>
<p>Let&#8217;s assume you have already created a gemini database of a VCF file using
the <tt class="docutils literal"><span class="pre">load</span></tt> module.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini load -v my.vcf -t snpEff my.db
</pre></div>
</div>
<p>Now, let&#8217;s imagine you have an annotated file in BED format (<tt class="docutils literal"><span class="pre">important.bed</span></tt>)
that describes regions of the genome that are particularly relevant to your
lab&#8217;s research. You would like to annotate in the gemini database which variants
overlap these crucial regions. We want to store this knowledge in a new column
in the <tt class="docutils literal"><span class="pre">variants</span></tt> table called <tt class="docutils literal"><span class="pre">important_variant</span></tt> that tracks whether a given
variant overlapped (1) or did not overlap (0) intervals in your annotation file.</p>
<p>To do this, you must first TABIX your BED file:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>bgzip important.bed
<span class="nv">$ </span>tabix -p bed important.bed.gz
</pre></div>
</div>
<div class="section" id="a-boolean-did-a-variant-overlap-a-region-or-not">
<h3><tt class="docutils literal"><span class="pre">-a</span> <span class="pre">boolean</span></tt> Did a variant overlap a region or not?<a class="headerlink" href="#a-boolean-did-a-variant-overlap-a-region-or-not" title="Permalink to this headline">¶</a></h3>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Formerly, the <tt class="docutils literal"><span class="pre">-a</span></tt> option was the <tt class="docutils literal"><span class="pre">-t</span></tt> option.</p>
</div>
<p>Now, you can use this TABIX&#8217;ed file to annotate which variants overlap your
important regions.  In the example below, the results will be stored in a new
column called &#8220;important&#8221;.  The <tt class="docutils literal"><span class="pre">-t</span> <span class="pre">boolean</span></tt> option says that you just want to
track whether (1) or not (0) the variant overlapped one or more of your regions.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini annotate -f important.bed.gz -c important -a boolean my.db
</pre></div>
</div>
<p>Since a new columns has been created in the database, we can now directly query
the new column.  In the example results below, the first and third variants
overlapped a crucial region while the second did not.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query <span class="se">\</span>
    -q <span class="s2">&quot;select chrom, start, end, variant_id, important from variants&quot;</span> <span class="se">\</span>
    my.db <span class="se">\</span>
    | head -3
chr22   100    101    1   1
chr22   200    201    2   0
chr22   300    500    3   1
</pre></div>
</div>
</div>
<div class="section" id="a-count-how-many-regions-did-a-variant-overlap">
<h3><tt class="docutils literal"><span class="pre">-a</span> <span class="pre">count</span></tt> How many regions did a variant overlap?<a class="headerlink" href="#a-count-how-many-regions-did-a-variant-overlap" title="Permalink to this headline">¶</a></h3>
<p>Instead of a simple yes or no, we can use the <tt class="docutils literal"><span class="pre">-t</span> <span class="pre">count</span></tt> option to <em>count</em>
how many important regions a variant overlapped.  It turns out that the 3rd
variant actually overlapped two important regions.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini annotate -f important.bed.gz -c important -a count my.db

<span class="nv">$ </span>gemini query <span class="se">\</span>
    -q <span class="s2">&quot;select chrom, start, end, variant_id, crucial from variants&quot;</span> <span class="se">\</span>
    my.db <span class="se">\</span>
    | head -3
chr22   100    101    1   1
chr22   200    201    2   0
chr22   300    500    3   2
</pre></div>
</div>
</div>
<div class="section" id="a-extract-extract-specific-values-from-a-bed-file">
<h3><tt class="docutils literal"><span class="pre">-a</span> <span class="pre">extract</span></tt> Extract specific values from a BED file<a class="headerlink" href="#a-extract-extract-specific-values-from-a-bed-file" title="Permalink to this headline">¶</a></h3>
<p>Lastly, we may also extract values from specific fields in a BED
file (or from the INFO field in a VCF) and populate one or more new columns
in the database based on
overlaps with the annotation file and the values of the fields therein.
To do this, we use the <tt class="docutils literal"><span class="pre">-a</span> <span class="pre">extract</span></tt> option.</p>
<p>This is best described with an example.  To set this up, let&#8217;s imagine
that we have a VCF file from a different experiment and we want to annotate
the variants in our GEMINI database with the allele frequency and depth
tags from the INFO fields for the same variants in this other VCF file.</p>
<blockquote>
<div># bgzip and tabix the vcf for use with the annotate tool.
$ bgzip other.vcf
$ tabix other.vcf.gz</div></blockquote>
<p>Now that we have a proper TABIX&#8217;ed VCF file, we can use the <tt class="docutils literal"><span class="pre">-a</span> <span class="pre">extract</span></tt> option to populate new
columns in the GEMINI database.  In order to do so, we must specify:</p>
<blockquote>
<div><ol class="arabic simple">
<li>its type (e.g., text, int, float,)  (<tt class="docutils literal"><span class="pre">-t</span></tt>)</li>
<li>the field in the INFO column of the VCF file that we should use to extract data with which to populate the new column (<tt class="docutils literal"><span class="pre">-e</span></tt>)</li>
<li>what operation should be used to summarize the data in the event of multiple overlaps in the annotation file  (<tt class="docutils literal"><span class="pre">-o</span></tt>)</li>
<li>(optionally) the name of the column we want to add (<tt class="docutils literal"><span class="pre">-c</span></tt>), if this is not specified, it will use the value from <tt class="docutils literal"><span class="pre">-e</span></tt>.</li>
</ol>
</div></blockquote>
<p>For example, let&#8217;s imagine we want to create a new column called &#8220;other_allele_freq&#8221; using the
AF field in our VCF file to populate it.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini annotate -f other.vcf.gz <span class="se">\</span>
                  -a extract <span class="se">\</span>
                  -c other_allele_freq <span class="se">\</span>
                  -t float <span class="se">\</span>
                  -e AF <span class="se">\</span>
                  -o mean <span class="se">\</span>
                  my.db
</pre></div>
</div>
<p>This create a new column in <tt class="docutils literal"><span class="pre">my.db</span></tt> called <tt class="docutils literal"><span class="pre">other_allele_freq</span></tt> and this
new column will be a FLOAT.  In the event of multiple records in the VCF
file overlapping a variant in the database, the average (mean) of the allele
frequencies values from the VCF file will be used.</p>
<p>At this point, one can query the database based on the values of the
new <tt class="docutils literal"><span class="pre">other_allele_freq</span></tt> column:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select * from variants where other_allele_freq &lt; 0.01&quot;</span> my.db
</pre></div>
</div>
</div>
<div class="section" id="t-type-specifying-the-column-type-s-when-using-a-extract">
<h3><tt class="docutils literal"><span class="pre">-t</span> <span class="pre">TYPE</span></tt> Specifying the column type(s) when using <tt class="docutils literal"><span class="pre">-a</span> <span class="pre">extract</span></tt><a class="headerlink" href="#t-type-specifying-the-column-type-s-when-using-a-extract" title="Permalink to this headline">¶</a></h3>
<p>The <tt class="docutils literal"><span class="pre">annotate</span></tt> tool will create three different types of columns via the <tt class="docutils literal"><span class="pre">-t</span></tt> option:</p>
<blockquote>
<div><ol class="arabic simple">
<li>Floating point columns for annotations with decimal precision as above (<tt class="docutils literal"><span class="pre">-t</span> <span class="pre">float</span></tt>)</li>
<li>Integer columns for integral annotations (<tt class="docutils literal"><span class="pre">-t</span> <span class="pre">integer</span></tt>)</li>
<li>Text columns for string columns such as &#8220;valid&#8221;, &#8220;yes&#8221;, etc. (<tt class="docutils literal"><span class="pre">-t</span> <span class="pre">text</span></tt>)</li>
</ol>
</div></blockquote>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <tt class="docutils literal"><span class="pre">-t</span></tt> option is only valid when using the <tt class="docutils literal"><span class="pre">-a</span> <span class="pre">extract</span></tt> option.</p>
</div>
</div>
<div class="section" id="o-operation-specifying-the-summary-operations-when-using-a-extract">
<h3><tt class="docutils literal"><span class="pre">-o</span> <span class="pre">OPERATION</span></tt> Specifying the summary operations when using <tt class="docutils literal"><span class="pre">-a</span> <span class="pre">extract</span></tt><a class="headerlink" href="#o-operation-specifying-the-summary-operations-when-using-a-extract" title="Permalink to this headline">¶</a></h3>
<p>In the event of multiple overlaps between a variant and records in the annotation
file, the <tt class="docutils literal"><span class="pre">annotate</span></tt> tool can summarize the values observed with multiple options:</p>
<blockquote>
<div><ol class="arabic simple">
<li><tt class="docutils literal"><span class="pre">-o</span> <span class="pre">mean</span></tt>.  Compute the average of the values.  <strong>They must be numeric</strong>.</li>
<li><tt class="docutils literal"><span class="pre">-o</span> <span class="pre">median</span></tt>. Compute the median of the values.  <strong>They must be numeric</strong>.</li>
<li><tt class="docutils literal"><span class="pre">-o</span> <span class="pre">mix</span></tt>. Compute the minimum of the values.  <strong>They must be numeric</strong>.</li>
<li><tt class="docutils literal"><span class="pre">-o</span> <span class="pre">max</span></tt>. Compute the maximum of the values.  <strong>They must be numeric</strong>.</li>
<li><tt class="docutils literal"><span class="pre">-o</span> <span class="pre">mode</span></tt>. Compute the maximum of the values.  <strong>They must be numeric</strong>.</li>
<li><tt class="docutils literal"><span class="pre">-o</span> <span class="pre">first</span></tt>. Use the value from the <strong>first</strong> record in the annotation file.</li>
<li><tt class="docutils literal"><span class="pre">-o</span> <span class="pre">last</span></tt>. Use the value from the <strong>last</strong> record in the annotation file.</li>
<li><tt class="docutils literal"><span class="pre">-o</span> <span class="pre">list</span></tt>. Create a comma-separated list of the observed values.  <strong>-t must be text</strong></li>
<li><tt class="docutils literal"><span class="pre">-o</span> <span class="pre">uniq_list</span></tt>. Create a comma-separated list of the <strong>distinct</strong> (i.e., non-redundant) observed values.  <strong>-t must be text</strong></li>
<li><tt class="docutils literal"><span class="pre">-o</span> <span class="pre">sum</span></tt>. Compute the sum of the values. <strong>They must be numeric</strong>.</li>
</ol>
</div></blockquote>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The <tt class="docutils literal"><span class="pre">-o</span></tt> option is only valid when using the <tt class="docutils literal"><span class="pre">-a</span> <span class="pre">extract</span></tt> option.</p>
</div>
</div>
<div class="section" id="annotating-with-vcf">
<h3>Annotating with VCF<a class="headerlink" href="#annotating-with-vcf" title="Permalink to this headline">¶</a></h3>
<p>Most of the examples to this point have pulled a column from a <cite>tabix</cite> indexed bed file.
It is likewise possible to pull from the INFO field  of a <cite>tabix</cite> index VCF. The syntax
is identical but the <tt class="docutils literal"><span class="pre">-e</span></tt> operation will specify the names of fields in the INFO column
to pull. By default, those names will be used, but that can still be specified with the
<cite>-c</cite> column.
Here are some example uses</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c"># put a DP column in the db:</span>
gemini annotate -f anno.vcf.gz -o list -e DP -t integer my.db

<span class="c"># ... and name it &#39;depth&#39;</span>
gemini annotate -f anno.vcf.gz -o list -e DP -c depth -t integer my.db

<span class="c"># use multiple columns</span>
gemini annotate -f anno.vcf.gz -o list,mean -e DP,Qmeter -c depth,qmeter -t integer my.db
</pre></div>
</div>
<p>Missing values are allowed since we expect that in some cases an annotation VCF will not
have all INFO fields specified for all variants.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">We recommend decomposing and normalizing variants before annotating.
See <a class="reference internal" href="preprocessing.html#preprocess"><em>Step 1. split, left-align, and trim variants</em></a> for a detailed explanation of how to do this.</p>
</div>
</div>
<div class="section" id="extracting-and-populating-multiple-columns-at-once">
<h3>Extracting and populating multiple columns at once.<a class="headerlink" href="#extracting-and-populating-multiple-columns-at-once" title="Permalink to this headline">¶</a></h3>
<p>One can also extract and populate multiple columns at once by providing
comma-separated lists (no spaces) of column names (<tt class="docutils literal"><span class="pre">-c</span></tt>), types (<tt class="docutils literal"><span class="pre">-t</span></tt>), numbers (<tt class="docutils literal"><span class="pre">-e</span></tt>),
and summary operations (<tt class="docutils literal"><span class="pre">-o</span></tt>).  For example, recall that in the VCF example above,
we created a TABIX&#8217;ed BED file containg the allele frequency and depth values from
the INFO field as the 4th and 5th columns in the BED, respectively.</p>
<p>Instead of running the <tt class="docutils literal"><span class="pre">annotate</span></tt> tool twice (once for eaxh column), we can
run the tool once and load both columns in the same run.  For example:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini annotate -f other.bed.gz <span class="se">\</span>
                  -a extract <span class="se">\</span>
                  -c other_allele_freq,other_depth <span class="se">\</span>
                  -t float,integer <span class="se">\</span>
                  -e 4,5 <span class="se">\</span>
                  -o mean,max <span class="se">\</span>
                  my.db
</pre></div>
</div>
<p>We can then use each of the new columns to filter variants with a GEMINI query:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini query -q <span class="s2">&quot;select * from variants \</span>
<span class="s2">                   where other_allele_freq &lt; 0.01 \</span>
<span class="s2">                   and other_depth &gt; 100&quot;</span> my.db
</pre></div>
</div>
</div>
</div>
<div class="section" id="region-extracting-variants-from-specific-regions-or-genes">
<h2><tt class="docutils literal"><span class="pre">region</span></tt>: Extracting variants from specific regions or genes<a class="headerlink" href="#region-extracting-variants-from-specific-regions-or-genes" title="Permalink to this headline">¶</a></h2>
<p>One often is concerned with variants found solely in a particular gene or
genomic region. <tt class="docutils literal"><span class="pre">gemini</span></tt> allows one to extract variants that fall within
specific genomic coordinates as follows:</p>
<div class="section" id="reg">
<h3><tt class="docutils literal"><span class="pre">--reg</span></tt><a class="headerlink" href="#reg" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini region --reg chr1:100-200 my.db
</pre></div>
</div>
</div>
<div class="section" id="gene">
<h3><tt class="docutils literal"><span class="pre">--gene</span></tt><a class="headerlink" href="#gene" title="Permalink to this headline">¶</a></h3>
<p>Or, one can extract variants based on a specific gene name.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini region --gene PTPN22 my.db
</pre></div>
</div>
</div>
<div class="section" id="id9">
<h3><tt class="docutils literal"><span class="pre">--columns</span></tt><a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<p>By default, this tool reports all columns in the <tt class="docutils literal"><span class="pre">variants</span></tt> table. One may
choose to report only a subset of the columns using the <tt class="docutils literal"><span class="pre">--columns</span></tt> option.  For
example, to report just the <tt class="docutils literal"><span class="pre">gene,</span> <span class="pre">chrom,</span> <span class="pre">start,</span> <span class="pre">end,</span> <span class="pre">ref,</span> <span class="pre">alt,</span> <span class="pre">impact</span></tt>, and <tt class="docutils literal"><span class="pre">impact_severity</span></tt> columns, one
would use the following:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini region --gene DHODH <span class="se">\</span>
                --columns <span class="s2">&quot;chrom, start, end, ref, alt, gene, impact&quot;</span> <span class="se">\</span>
                my.db

chr16   72057281    72057282    A   G   DHODH   intron
chr16   72057434    72057435    C   T   DHODH   non_syn_coding
chr16   72059268    72059269    T   C   DHODH   downstream
</pre></div>
</div>
</div>
<div class="section" id="id10">
<h3><tt class="docutils literal"><span class="pre">--filter</span></tt><a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h3>
<p>By default, this tool will report all variants regardless of their putative
functional impact.  In order to apply additional constraints on the variants
returned, one can use the <tt class="docutils literal"><span class="pre">--filter</span></tt> option.  Using SQL syntax, conditions
applied with the <a href="#id11"><span class="problematic" id="id12">``</span></a>&#8211;filter option become WHERE clauses in the query issued to
the GEMINI database.  For example, if we wanted to restrict candidate variants
to solely those with a HIGH predicted functional consequence, we could use the
following:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini region --gene DHODH <span class="se">\</span>
                --columns <span class="s2">&quot;chrom, start, end, ref, alt, gene, impact&quot;</span> <span class="se">\</span>
                --filter <span class="s2">&quot;alt=&#39;G&#39;&quot;</span>
                my.db

chr16   72057281    72057282    A   G   DHODH   intron
</pre></div>
</div>
</div>
<div class="section" id="json">
<h3><tt class="docutils literal"><span class="pre">--json</span></tt><a class="headerlink" href="#json" title="Permalink to this headline">¶</a></h3>
<p>Reporting query output in JSON format may enable
HTML/Javascript apps to query GEMINI and retrieve
the output in a format that is amenable to web development protocols.</p>
<p>To report in JSON format, use the <tt class="docutils literal"><span class="pre">--json</span></tt> option. For example:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini region --gene DHODH <span class="se">\</span>
                --columns <span class="s2">&quot;chrom, start, end, ref, alt, gene, impact&quot;</span> <span class="se">\</span>
                --filter <span class="s2">&quot;alt=&#39;G&#39;&quot;</span>
                --json
                my.db

<span class="o">{</span><span class="s2">&quot;chrom&quot;</span>: <span class="s2">&quot;chr16&quot;</span>, <span class="s2">&quot;start&quot;</span>: 72057281, <span class="s2">&quot;end&quot;</span>: 72057282, <span class="s2">&quot;ref&quot;</span>: <span class="s2">&quot;A&quot;</span>, <span class="s2">&quot;alt&quot;</span>: <span class="s2">&quot;G&quot;</span>, <span class="s2">&quot;gene&quot;</span>: <span class="s2">&quot;DHODH&quot;</span><span class="o">}</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="windower-conducting-analyses-on-genome-windows">
<h2><tt class="docutils literal"><span class="pre">windower</span></tt>: Conducting analyses on genome &#8220;windows&#8221;.<a class="headerlink" href="#windower-conducting-analyses-on-genome-windows" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal"><span class="pre">gemini</span></tt> includes a convenient tool for computing variation metrics across
genomic windows (both fixed and sliding). Here are a few examples to whet your
appetite.  If you&#8217;re still hungry, contact us.</p>
<p>Compute the average nucleotide diversity for all variants found in
non-overlapping, 50Kb windows.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini windower -w 50000 -s 0 -t nucl_div -o mean my.db
</pre></div>
</div>
<p>Compute the average nucleotide diversity for all variants found in 50Kb windows
that overlap by 10kb.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini windower -w 50000 -s 10000 -t nucl_div -o mean my.db
</pre></div>
</div>
<p>Compute the max value for HWE statistic for all variants in a window of size
10kb</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini windower  -w 10000 -t hwe -o max my.db
</pre></div>
</div>
</div>
<div class="section" id="stats-compute-useful-variant-statistics">
<h2><tt class="docutils literal"><span class="pre">stats</span></tt>: Compute useful variant statistics.<a class="headerlink" href="#stats-compute-useful-variant-statistics" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">stats</span></tt> tool computes some useful variant statistics like</p>
<p>Compute the transition and transversion ratios for the snps</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini stats --tstv my.db
ts      tv      ts/tv
4       5       0.8
</pre></div>
</div>
<div class="section" id="tstv-coding">
<h3><tt class="docutils literal"><span class="pre">--tstv-coding</span></tt><a class="headerlink" href="#tstv-coding" title="Permalink to this headline">¶</a></h3>
<p>Compute the transition/transversion ratios for the snps in the coding
regions.</p>
</div>
<div class="section" id="tstv-noncoding">
<h3><tt class="docutils literal"><span class="pre">--tstv-noncoding</span></tt><a class="headerlink" href="#tstv-noncoding" title="Permalink to this headline">¶</a></h3>
<p>Compute the transition/transversion ratios for the snps in the non-coding
regions.</p>
<p>Compute the type and count of the snps.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini stats --snp-counts my.db
<span class="nb">type    </span>count
A-&gt;G    2
C-&gt;T    1
G-&gt;A    1
</pre></div>
</div>
<p>Calculate the site frequency spectrum of the variants.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini stats --sfs my.db
aaf     count
0.125   2
0.375   1
</pre></div>
</div>
<p>Compute the pair-wise genetic distance between each sample</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini stats --mds my.db
sample1 sample2 distance
M10500  M10500  0.0
M10475  M10478  1.25
M10500  M10475  2.0
M10500  M10478  0.5714
</pre></div>
</div>
<p>Return a count of the types of genotypes per sample</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini stats --gts-by-sample my.db
sample  num_hom_ref     num_het num_hom_alt     num_unknown     total
M10475  4       1       3       1       9
M10478  2       2       4       1       9
</pre></div>
</div>
<p>Return the total variants per sample (sum of homozygous
and heterozygous variants)</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini stats --vars-by-sample my.db
sample  total
M10475  4
M10478  6
</pre></div>
</div>
</div>
<div class="section" id="summarize">
<h3><tt class="docutils literal"><span class="pre">--summarize</span></tt><a class="headerlink" href="#summarize" title="Permalink to this headline">¶</a></h3>
<p>If none of these tools are exactly what you want, you can summarize the variants
per sample of an arbitrary query using the &#8211;summarize flag. For example, if you
wanted to know, for each sample, how many variants are on chromosome 1 that are also
in dbSNP:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini stats --summarize <span class="s2">&quot;select * from variants where in_dbsnp=1 and chrom=&#39;chr1&#39;&quot;</span> my.db
sample  total   num_het num_hom_alt
M10475  1       1       0
M128215 1       1       0
M10478  2       2       0
M10500  2       1       1
</pre></div>
</div>
</div>
</div>
<div class="section" id="burden-perform-sample-wise-gene-level-burden-calculations">
<h2><tt class="docutils literal"><span class="pre">burden</span></tt>: perform sample-wise gene-level burden calculations<a class="headerlink" href="#burden-perform-sample-wise-gene-level-burden-calculations" title="Permalink to this headline">¶</a></h2>
<p>The <tt class="docutils literal"><span class="pre">burden</span></tt> tool provides a set of utilities to perform burden
summaries on a per-gene, per sample basis. By default, it outputs
a table of gene-wise counts of all high impact variants in coding regions for
each sample:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini burden test.burden.db
gene    M10475  M10478  M10500  M128215
WDR37   2       2       2       2
CTBP2   0       0       0       1
DHODH   1       0       0       0
</pre></div>
</div>
<div class="section" id="nonsynonymous">
<h3><tt class="docutils literal"><span class="pre">--nonsynonymous</span></tt><a class="headerlink" href="#nonsynonymous" title="Permalink to this headline">¶</a></h3>
<p>If you want to be a little bit less restrictive, you can include all
non-synonymous variants instead:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini burden --nonsynonymous test.burden.db
gene    M10475  M10478  M10500  M128215
SYCE1   0       1       1       0
WDR37   2       2       2       2
CTBP2   0       0       0       1
ASAH2C  2       1       1       0
DHODH   1       0       0       0
</pre></div>
</div>
</div>
<div class="section" id="calpha">
<h3><tt class="docutils literal"><span class="pre">--calpha</span></tt><a class="headerlink" href="#calpha" title="Permalink to this headline">¶</a></h3>
<p>If your database has been loaded with a PED file describing case and
control samples, you can calculate the
<a class="reference external" href="http://www.plosgenetics.org/article/info%3Adoi%2F10.1371%2Fjournal.pgen.1001322">c-alpha</a>
statistic for cases vs. control:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini burden --calpha test.burden.db
gene    T       c       Z       p_value
SYCE1   -0.5    0.25    -1.0    0.841344746069
WDR37   -1.0    1.5     -0.816496580928 0.792891910879
CTBP2   0.0     0.0     nan     nan
ASAH2C  -0.5    0.75    -0.57735026919  0.718148569175
DHODH   0.0     0.0     nan     nan
</pre></div>
</div>
<p>To calculate the P-value using a permutation test, use the <tt class="docutils literal"><span class="pre">--permutations</span></tt> option,
specifying the number of permutations of the case/control labels you want to use.</p>
</div>
<div class="section" id="min-aaf-and-max-aaf-for-calpha">
<h3><tt class="docutils literal"><span class="pre">--min-aaf</span></tt> and <tt class="docutils literal"><span class="pre">--max-aaf</span></tt> for <tt class="docutils literal"><span class="pre">--calpha</span></tt><a class="headerlink" href="#min-aaf-and-max-aaf-for-calpha" title="Permalink to this headline">¶</a></h3>
<p>By default, all variants affecting a given gene will be included in the
C-alpha computation.  However, one may establish alternate allele frequency
boundaries for the variants included using the <tt class="docutils literal"><span class="pre">--min-aaf</span></tt> and
<tt class="docutils literal"><span class="pre">--max-aaf</span></tt> options.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini burden --calpha test.burden.db --min-aaf 0.0 --max-aaf 0.01
</pre></div>
</div>
</div>
<div class="section" id="cases-and-controls-for-calpha">
<h3><tt class="docutils literal"><span class="pre">--cases</span></tt> and <tt class="docutils literal"><span class="pre">--controls</span> <span class="pre">for</span> <span class="pre">``--calpha</span></tt><a class="headerlink" href="#cases-and-controls-for-calpha" title="Permalink to this headline">¶</a></h3>
<p>If you do not have a PED file loaded, or your PED file does not follow the
standard <a class="reference external" href="http://pngu.mgh.harvard.edu/~purcell/plink/data.shtml">PED phenotype encoding format</a>
you can still perform the c-alpha test, but you have to specify which samples
are the control samples and which are the case samples:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini burden --controls M10475 M10478 --cases M10500 M128215 --calpha test.burden.db
gene    T       c       Z       p_value
SYCE1   -0.5    0.25    -1.0    0.841344746069
WDR37   -1.0    1.5     -0.816496580928 0.792891910879
CTBP2   0.0     0.0     nan     nan
ASAH2C  -0.5    0.75    -0.57735026919  0.718148569175
DHODH   0.0     0.0     nan     nan
</pre></div>
</div>
</div>
<div class="section" id="nonsynonymous-calpha">
<h3><tt class="docutils literal"><span class="pre">--nonsynonymous</span></tt> <tt class="docutils literal"><span class="pre">--calpha</span></tt><a class="headerlink" href="#nonsynonymous-calpha" title="Permalink to this headline">¶</a></h3>
<p>If you would rather consider all nonsynonymous variants for the C-alpha test rather
than just the medium and high impact variants, add the <tt class="docutils literal"><span class="pre">--nonsynonymous</span></tt> flag.</p>
</div>
</div>
<div class="section" id="roh-identifying-runs-of-homozygosity">
<h2><tt class="docutils literal"><span class="pre">ROH</span></tt>: Identifying runs of homozygosity<a class="headerlink" href="#roh-identifying-runs-of-homozygosity" title="Permalink to this headline">¶</a></h2>
<p>Runs of homozygosity are long stretches of homozygous genotypes that reflect
segments shared identically by descent and are a result of consanguinity or
natural selection. Consanguinity elevates the occurrence of rare recessive
diseases (e.g. cystic fibrosis) that represent homozygotes for strongly deleterious
mutations. Hence, the identification of these runs holds medical value.</p>
<p>The &#8216;roh&#8217; tool in GEMINI returns runs of homozygosity identified in whole genome data.
The tool basically looks at every homozygous position on the chromosome as a possible
start site for the run and looks for those that could give rise to a potentially long
stretch of homozygous genotypes.</p>
<p>For e.g. for the given example allowing <tt class="docutils literal"><span class="pre">1</span> <span class="pre">HET</span></tt> genotype (h) and <tt class="docutils literal"><span class="pre">2</span> <span class="pre">UKW</span></tt> genotypes (u)
the possible roh runs (H) would be:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">genotype_run</span> <span class="o">=</span> H H H H h H H H H u H H H H H u H H H H H H H h H H H H H h H H H H H
<span class="nv">roh_run1</span>     <span class="o">=</span> H H H H h H H H H u H H H H H u H H H H H H H
<span class="nv">roh_run2</span>     <span class="o">=</span>           H H H H u H H H H H u H H H H H H H h H H H H H
<span class="nv">roh_run3</span>     <span class="o">=</span>                     H H H H H u H H H H H H H h H H H H H
<span class="nv">roh_run4</span>     <span class="o">=</span>                                 H H H H H H H h H H H H H
</pre></div>
</div>
<p>roh returned for &#8211;min-snps = 20 would be:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">roh_run1</span>     <span class="o">=</span> H H H H h H H H H u H H H H H u H H H H H H H
<span class="nv">roh_run2</span>     <span class="o">=</span>           H H H H u H H H H H u H H H H H H H h H H H H H
</pre></div>
</div>
<p>As you can see, the immediate homozygous position right of a break (h or u) would be the possible
start of a new roh run and genotypes to the left of a break are pruned since they cannot
be part of a longer run than we have seen before.</p>
<p>Return <tt class="docutils literal"><span class="pre">roh</span></tt> with minimum of 50 snps, a minimum run length of 1 mb and a minimum sample depth of 20
for sample S138 (with default values for allowed number of HETS, UNKS and total depth).</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini roh --min-snps 50 <span class="se">\</span>
           --min-gt-depth 20 <span class="se">\</span>
                   --min-size 1000000 <span class="se">\</span>
                   -s S138 <span class="se">\</span>
                   roh_run.db
chrom   start   end     sample  num_of_snps     density_per_kb  run_length_in_bp
chr2 233336080 234631638 S138 2583 1.9953 1295558
chr2    238341281       239522281       S138    2899    2.4555  1181000
</pre></div>
</div>
</div>
<div class="section" id="set-somatic-flag-somatic-variants">
<h2><tt class="docutils literal"><span class="pre">set_somatic</span></tt>: Flag somatic variants<a class="headerlink" href="#set-somatic-flag-somatic-variants" title="Permalink to this headline">¶</a></h2>
<p>Somatic mutations in a tumor-normal pair are variants that are present in
the tumor but not in the normal sample.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>1. This tool requires that you specify the sample layout via a PED file
when loading your VCF into GEMINI via:</p>
<p class="last"><tt class="docutils literal"><span class="pre">gemini</span> <span class="pre">load</span> <span class="pre">-v</span> <span class="pre">my.vcf</span> <span class="pre">-p</span> <span class="pre">my.ped</span> <span class="pre">my.db</span></tt></p>
</div>
<p><cite>Example PED file format for GEMINI</cite></p>
<div class="highlight-bash"><div class="highlight"><pre><span class="c">#Family_ID      Individual_ID   Paternal_ID     Maternal_ID     Sex     Phenotype       Ethnicity</span>
1       Normal  -9      -9      0       1       -9
1       Tumor   -9      -9      0       2       -9
</pre></div>
</div>
<div class="section" id="id13">
<h3><tt class="docutils literal"><span class="pre">default</span> <span class="pre">behavior</span></tt><a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h3>
<p>By default, <tt class="docutils literal"><span class="pre">set_somatic</span></tt> simply marks variants that are genotyped as
homozygous reference in the normal sample and non-reference in the tumor.
More stringent somatic filtering criteria are available through tunable
command line parameters.</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini set_somatic <span class="se">\</span>
    --min-depth 30 <span class="se">\</span>
    --min-qual 20 <span class="se">\</span>
    --min-somatic-score 18 <span class="se">\</span>
    --min-tumor-depth 10 <span class="se">\</span>
    --min-norm-depth 10 <span class="se">\</span>
    tumor_normal.db
tum_name        tum_gt  tum_alt_freq    tum_alt_depth   tum_depth       nrm_name        nrm_gt  nrm_alt_freq    nrm_alt_depth   nrm_depth       chrom   start   end     ref     alt     gene
tumor   GAAAAAAAAAAAAAGGTGAAAATT/GAAAAAAAAAAAAGGTGAAAATT        0.217391304348  5       23      normal  GAAAAAAAAAAAAAGGTGAAAATT/GAAAAAAAAAAAAAGGTGAAAATT       0.0     0       25      chrX    132838304       132838328       GAAAAAAAAAAAAAGGTGAAAATT        GAAAAAAAAAAAAGGTGAAAATT GPC3
tumor   CTGCTATTTTG/CG  0.22    11      50      normal  CTGCTATTTTG/CTGCTATTTTG 0.0     0       70      chr17   59861630        59861641        CTGCTATTTTG     CG      BRIP1
tumor   C/A     0.555555555556  10      18      normal  C/C     0.0     0       17      chr17   7578460 7578461 C       A       TP53
tumor   C/T     0.1875  12      64      normal  C/C     0.0     0       30      chr2    128046288       128046289       C       T       ERCC3
Identified and <span class="nb">set </span>4 somatic mutations
</pre></div>
</div>
</div>
<div class="section" id="min-depth-none">
<h3><tt class="docutils literal"><span class="pre">--min-depth</span> <span class="pre">[None]</span></tt><a class="headerlink" href="#min-depth-none" title="Permalink to this headline">¶</a></h3>
<p>The minimum required combined depth for tumor and normal samples.</p>
</div>
<div class="section" id="min-qual-none">
<h3><tt class="docutils literal"><span class="pre">--min-qual</span> <span class="pre">[None]</span></tt><a class="headerlink" href="#min-qual-none" title="Permalink to this headline">¶</a></h3>
<p>The minimum required variant quality score.</p>
</div>
<div class="section" id="min-somatic-score-none">
<h3><tt class="docutils literal"><span class="pre">--min-somatic-score</span> <span class="pre">[None]</span></tt><a class="headerlink" href="#min-somatic-score-none" title="Permalink to this headline">¶</a></h3>
<p>The minimum required somatic score (SSC). This score is produced by various
somatic variant detection algorithms including SpeedSeq, SomaticSniper,
and VarScan 2.</p>
</div>
<div class="section" id="max-norm-alt-freq-none">
<h3><tt class="docutils literal"><span class="pre">--max-norm-alt-freq</span> <span class="pre">[None]</span></tt><a class="headerlink" href="#max-norm-alt-freq-none" title="Permalink to this headline">¶</a></h3>
<p>The maximum frequency of the alternate allele allowed in the normal sample.</p>
</div>
<div class="section" id="max-norm-alt-count-none">
<h3><tt class="docutils literal"><span class="pre">--max-norm-alt-count</span> <span class="pre">[None]</span></tt><a class="headerlink" href="#max-norm-alt-count-none" title="Permalink to this headline">¶</a></h3>
<p>The maximum count of the alternate allele allowed in the normal sample.</p>
</div>
<div class="section" id="min-norm-depth-none">
<h3><tt class="docutils literal"><span class="pre">--min-norm-depth</span> <span class="pre">[None]</span></tt><a class="headerlink" href="#min-norm-depth-none" title="Permalink to this headline">¶</a></h3>
<p>The minimum depth required in the normal sample.</p>
</div>
<div class="section" id="min-tumor-alt-freq-none">
<h3><tt class="docutils literal"><span class="pre">--min-tumor-alt-freq</span> <span class="pre">[None]</span></tt><a class="headerlink" href="#min-tumor-alt-freq-none" title="Permalink to this headline">¶</a></h3>
<p>The minimum frequency of the alternate allele required in the tumor sample.</p>
</div>
<div class="section" id="min-tumor-alt-count-none">
<h3><tt class="docutils literal"><span class="pre">--min-tumor-alt-count</span> <span class="pre">[None]</span></tt><a class="headerlink" href="#min-tumor-alt-count-none" title="Permalink to this headline">¶</a></h3>
<p>The minimum count of the alternate allele required in the tumor sample.</p>
</div>
<div class="section" id="min-tumor-depth-none">
<h3><tt class="docutils literal"><span class="pre">--min-tumor-depth</span> <span class="pre">[None]</span></tt><a class="headerlink" href="#min-tumor-depth-none" title="Permalink to this headline">¶</a></h3>
<p>The minimum depth required in the tumor sample.</p>
</div>
<div class="section" id="chrom-none">
<h3><tt class="docutils literal"><span class="pre">--chrom</span> <span class="pre">[None]</span></tt><a class="headerlink" href="#chrom-none" title="Permalink to this headline">¶</a></h3>
<p>A specific chromosome on which to flag somatic mutations.</p>
</div>
<div class="section" id="dry-run">
<h3><tt class="docutils literal"><span class="pre">--dry-run</span></tt><a class="headerlink" href="#dry-run" title="Permalink to this headline">¶</a></h3>
<p>Don&#8217;t set the is_somatic flag, just report what _would_ be set. For testing
purposes.</p>
</div>
</div>
<div class="section" id="actionable-mutations-report-actionable-somatic-mutations-and-drug-gene-interactions">
<h2><tt class="docutils literal"><span class="pre">actionable_mutations</span></tt>: Report actionable somatic mutations and drug-gene interactions<a class="headerlink" href="#actionable-mutations-report-actionable-somatic-mutations-and-drug-gene-interactions" title="Permalink to this headline">¶</a></h2>
<p>Actionable mutations are somatic variants in COSMIC cancer census genes with
medium or high impact severity predictions. This tool reports actionable
mutations as well as their known drug interactions (if any) from DGIdb.
Current functionality is only for SNVs and indels.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ol class="last arabic">
<li><dl class="first docutils">
<dt>This tool requires somatic variants to have been flagged using</dt>
<dd><p class="first last"><tt class="docutils literal"><span class="pre">set_somatic</span></tt></p>
</dd>
</dl>
</li>
</ol>
</div>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini actionable_mutations tumor_normal.db
tum_name        chrom   start   end     ref     alt     gene    impact  is_somatic      in_cosmic_census        dgidb_info
tumor   chr2    128046288       128046289       C       T       ERCC3   non_syn_coding  1       1       None
tumor   chr17   7578460 7578461 C       A       TP53    non_syn_coding  1       1       <span class="o">{</span><span class="s1">&#39;searchTerm&#39;</span>: <span class="s1">&#39;TP53&#39;</span>, <span class="s1">&#39;geneCategories&#39;</span>: <span class="o">[</span><span class="s1">&#39;CLINICALLY ACTIONABLE&#39;</span>, <span class="s1">&#39;DRUGGABLE GENOME&#39;</span>, <span class="s1">&#39;TUMOR SUPPRESSOR&#39;</span>, <span class="s1">&#39;TRANSCRIPTION FACTOR COMPLEX&#39;</span>, <span class="s1">&#39;DRUG RESISTANCE&#39;</span>, <span class="s1">&#39;HISTONE MODIFICATION&#39;</span>, <span class="s1">&#39;DNA REPAIR&#39;</span>, <span class="s1">&#39;TRANSCRIPTION FACTOR BINDING&#39;</span><span class="o">]</span>, <span class="s1">&#39;geneName&#39;</span>: <span class="s1">&#39;TP53&#39;</span>, <span class="s1">&#39;geneLongName&#39;</span>: <span class="s1">&#39;tumor protein p53&#39;</span>, <span class="s1">&#39;interactions&#39;</span>: <span class="o">[{</span><span class="s1">&#39;source&#39;</span>: <span class="s1">&#39;DrugBank&#39;</span>, <span class="s1">&#39;interactionId&#39;</span>: <span class="s1">&#39;711cbe42-4930-4b46-963e-79ab35bbbd0f&#39;</span>, <span class="s1">&#39;interactionType&#39;</span>: <span class="s1">&#39;n/a&#39;</span>, <span class="s1">&#39;drugName&#39;</span>: <span class="s1">&#39;1-(9-ETHYL-9H-CARBAZOL-3-YL)-N-METHYLMETHANAMINE&#39;</span><span class="o">}</span>, <span class="o">{</span><span class="s1">&#39;source&#39;</span>: <span class="s1">&#39;PharmGKB&#39;</span>, <span class="s1">&#39;interactionId&#39;</span>: <span class="s1">&#39;8234d9b9-085d-49b1-aac2-cf5375d91477&#39;</span>, <span class="s1">&#39;interactionType&#39;</span>: <span class="s1">&#39;n/a&#39;</span>, <span class="s1">&#39;drugName&#39;</span>: <span class="s1">&#39;FLUOROURACIL&#39;</span><span class="o">}</span>, <span class="o">{</span><span class="s1">&#39;source&#39;</span>: <span class="s1">&#39;PharmGKB&#39;</span>, <span class="s1">&#39;interactionId&#39;</span>: <span class="s1">&#39;605d7bca-7ed9-428e-aa7c-f76aafd66b54&#39;</span>, <span class="s1">&#39;interactionType&#39;</span>: <span class="s1">&#39;n/a&#39;</span>, <span class="s1">&#39;drugName&#39;</span>: <span class="s1">&#39;PACLITAXEL&#39;</span><span class="o">}</span>, <span class="o">{</span><span class="s1">&#39;source&#39;</span>: <span class="s1">&#39;TTD&#39;</span>, <span class="s1">&#39;interactionId&#39;</span>: <span class="s1">&#39;1fe9db63-3581-435b-b22a-12d45c8c9864&#39;</span>, <span class="s1">&#39;interactionType&#39;</span>: <span class="s1">&#39;activator&#39;</span>, <span class="s1">&#39;drugName&#39;</span>: <span class="s1">&#39;CURAXIN CBLC102&#39;</span><span class="o">}</span>, <span class="o">{</span><span class="s1">&#39;source&#39;</span>: <span class="s1">&#39;TALC&#39;</span>, <span class="s1">&#39;interactionId&#39;</span>: <span class="s1">&#39;8f8f6822-cb9e-40aa-8360-5532e059f1e7&#39;</span>, <span class="s1">&#39;interactionType&#39;</span>: <span class="s1">&#39;vaccine&#39;</span>, <span class="s1">&#39;drugName&#39;</span>: <span class="s1">&#39;EP-2101&#39;</span><span class="o">}</span>, <span class="o">{</span><span class="s1">&#39;source&#39;</span>: <span class="s1">&#39;TALC&#39;</span>, <span class="s1">&#39;interactionId&#39;</span>: <span class="s1">&#39;d59e14bc-b9a5-4c9f-a5aa-7ba322f0fa0e&#39;</span>, <span class="s1">&#39;interactionType&#39;</span>: <span class="s1">&#39;vaccine&#39;</span>, <span class="s1">&#39;drugName&#39;</span>: <span class="s1">&#39;MUTANT P53 PEPTIDE PULSED DENDRITIC CELL&#39;</span><span class="o">}</span>, <span class="o">{</span><span class="s1">&#39;source&#39;</span>: <span class="s1">&#39;TALC&#39;</span>, <span class="s1">&#39;interactionId&#39;</span>: <span class="s1">&#39;79256b6e-9a16-4fbe-a237-28dbca28bc2a&#39;</span>, <span class="s1">&#39;interactionType&#39;</span>: <span class="s1">&#39;vaccine&#39;</span>, <span class="s1">&#39;drugName&#39;</span>: <span class="s1">&#39;AD.P53-DC&#39;</span><span class="o">}]}</span>
tumor   chr17   59861630        59861641        CTGCTATTTTG     CG      BRIP1   inframe_codon_loss      1       1       None
tumor   chrX    132838304       132838328       GAAAAAAAAAAAAAGGTGAAAATT        GAAAAAAAAAAAAGGTGAAAATT GPC3    splice_region   1       1       None
</pre></div>
</div>
</div>
<div class="section" id="fusions-report-putative-gene-fusions">
<h2><tt class="docutils literal"><span class="pre">fusions</span></tt>: Report putative gene fusions<a class="headerlink" href="#fusions-report-putative-gene-fusions" title="Permalink to this headline">¶</a></h2>
<p>Report putative somatic gene fusions from structural variants in a tumor-normal
pair. Putative fusions join two genes and preserve transcript strand
orientation.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<ol class="last arabic">
<li><dl class="first docutils">
<dt>This tool requires somatic variants to have been flagged using</dt>
<dd><p class="first last"><tt class="docutils literal"><span class="pre">set_somatic</span></tt></p>
</dd>
</dl>
</li>
</ol>
</div>
<div class="section" id="id14">
<h3><tt class="docutils literal"><span class="pre">default</span> <span class="pre">behavior</span></tt><a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h3>
<p>By default, <tt class="docutils literal"><span class="pre">fusions</span></tt> reports structural variants that are flagged as
somatic, join two different genes, and preserve transcript strand orientation.
These may be further filtered using tunable command line parameters.</p>
<div class="highlight-bash"><div class="highlight"><pre>    <span class="nv">$ </span>gemini fusions <span class="se">\</span>
        --min_qual 5 <span class="se">\</span>
        --in_cosmic_census <span class="se">\</span>
        tumor_normal.db
    chromA   breakpointA_start  breakpointA_end     chromB  breakpointB_start   breakpointB_end var_id  qual    strandA strandB sv_type geneA   geneB   tool    evidence_type   is_precise  sample
chr3        176909953       176909982       chr3    178906001       178906030       1233    9.58    -       +       complex TBL1XR1 PIK3CA  LUMPY   PE      0       tumor
</pre></div>
</div>
</div>
<div class="section" id="id15">
<h3><tt class="docutils literal"><span class="pre">--min_qual</span> <span class="pre">[None]</span></tt><a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h3>
<p>The minimum required variant quality score.</p>
</div>
<div class="section" id="evidence-type-string">
<h3><tt class="docutils literal"><span class="pre">--evidence_type</span> <span class="pre">STRING</span></tt><a class="headerlink" href="#evidence-type-string" title="Permalink to this headline">¶</a></h3>
<p>The required supporting evidence types for the variant from
LUMPY (&#8220;PE&#8221;, &#8220;SR&#8221;, or &#8220;PE,SR&#8221;).</p>
</div>
<div class="section" id="in-cosmic-census">
<h3><tt class="docutils literal"><span class="pre">--in_cosmic_census</span></tt><a class="headerlink" href="#in-cosmic-census" title="Permalink to this headline">¶</a></h3>
<p>Require at least one of the affected genes to be in the
COSMIC cancer gene census.</p>
</div>
</div>
<div class="section" id="db-info-list-the-gemini-database-tables-and-columns">
<h2><tt class="docutils literal"><span class="pre">db_info</span></tt>: List the gemini database tables and columns<a class="headerlink" href="#db-info-list-the-gemini-database-tables-and-columns" title="Permalink to this headline">¶</a></h2>
<p>Because of the sheer number of annotations that are stored in gemini, there are
admittedly too many columns to remember by rote.  If you can&#8217;t recall the name of
particular column, just use the <tt class="docutils literal"><span class="pre">db_info</span></tt> tool.  It will report all of the
tables and all of the columns / types in each table:</p>
<div class="highlight-bash"><div class="highlight"><pre><span class="nv">$ </span>gemini db_info test.db
table_name          column_name                   <span class="nb">type</span>
variants            chrom                         text
variants            start                         integer
variants            end                           integer
variants            variant_id                    integer
variants            anno_id                       integer
variants            ref                           text
variants            alt                           text
variants            qual                          float
variants            filter                        text
variants            <span class="nb">type                          </span>text
variants            sub_type                      text
variants            gts                           blob
variants            gt_types                      blob
variants            gt_phases                     blob
variants            gt_depths                     blob
variants            call_rate                     float
variants            in_dbsnp                      bool
variants            rs_ids                        text
variants            in_omim                       bool
variants            clin_sigs                     text
variants            cyto_band                     text
variants            rmsk                          text
variants            in_cpg_island                 bool
variants            in_segdup                     bool
variants            is_conserved                  bool
variants            num_hom_ref                   integer
variants            num_het                       integer
variants            num_hom_alt                   integer
variants            num_unknown                   integer
variants            aaf                           float
variants            hwe                           float
variants            inbreeding_coeff              float
variants            pi                            float
variants            recomb_rate                   float
variants            gene                          text
variants            transcript                    text
variants            is_exonic                     bool
variants            is_coding                     bool
variants            is_lof                        bool
variants            exon                          text
variants            codon_change                  text
variants            aa_change                     text
variants            aa_length                     text
variants            biotype                       text
variants            impact                        text
variants            impact_severity               text
variants            polyphen_pred                 text
variants            polyphen_score                float
variants            sift_pred                     text
variants            sift_score                    float
variants            anc_allele                    text
variants            rms_bq                        float
variants            cigar                         text
variants            depth                         integer
variants            strand_bias                   float
variants            rms_map_qual                  float
variants            in_hom_run                    integer
variants            num_mapq_zero                 integer
variants            num_alleles                   integer
variants            num_reads_w_dels              float
variants            haplotype_score               float
variants            qual_depth                    float
variants            allele_count                  integer
variants            allele_bal                    float
variants            in_hm2                        bool
variants            in_hm3                        bool
variants            is_somatic
variants            in_esp                        bool
variants            aaf_esp_ea                    float
variants            aaf_esp_aa                    float
variants            aaf_esp_all                   float
variants            exome_chip                    bool
variants            in_1kg                        bool
variants            aaf_1kg_amr                   float
variants            aaf_1kg_asn                   float
variants            aaf_1kg_afr                   float
variants            aaf_1kg_eur                   float
variants            aaf_1kg_all                   float
variants            grc                           text
variants            gms_illumina                  float
variants            gms_solid                     float
variants            gms_iontorrent                float
variants            encode_tfbs
variants            encode_consensus_gm12878      text
variants            encode_consensus_h1hesc       text
variants            encode_consensus_helas3       text
variants            encode_consensus_hepg2        text
variants            encode_consensus_huvec        text
variants            encode_consensus_k562         text
variants            encode_segway_gm12878         text
variants            encode_segway_h1hesc          text
variants            encode_segway_helas3          text
variants            encode_segway_hepg2           text
variants            encode_segway_huvec           text
variants            encode_segway_k562            text
variants            encode_chromhmm_gm12878       text
variants            encode_chromhmm_h1hesc        text
variants            encode_chromhmm_helas3        text
variants            encode_chromhmm_hepg2         text
variants            encode_chromhmm_huvec         text
variants            encode_chromhmm_k562          text
variant_impacts     variant_id                    integer
variant_impacts     anno_id                       integer
variant_impacts     gene                          text
variant_impacts     transcript                    text
variant_impacts     is_exonic                     bool
variant_impacts     is_coding                     bool
variant_impacts     is_lof                        bool
variant_impacts     exon                          text
variant_impacts     codon_change                  text
variant_impacts     aa_change                     text
variant_impacts     aa_length                     text
variant_impacts     biotype                       text
variant_impacts     impact                        text
variant_impacts     impact_severity               text
variant_impacts     polyphen_pred                 text
variant_impacts     polyphen_score                float
variant_impacts     sift_pred                     text
variant_impacts     sift_score                    float
samples             sample_id                     integer
samples             name                          text
samples             family_id                     integer
samples             paternal_id                   integer
samples             maternal_id                   integer
samples             sex                           text
samples             phenotype                     text
samples             ethnicity                     text
</pre></div>
</div>
</div>
</div>




  

<div id="disqus_thread"></div>
 <script type="text/javascript">
     /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
     var disqus_shortname = 'geminidocs'; // required: replace example with your forum shortname

     /* * * DON'T EDIT BELOW THIS LINE * * */
     (function() {
         var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
         dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
         (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
     })();
 </script>
 <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
 <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>






          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/gemini.png" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Built-in analysis tools</a><ul>
<li><a class="reference internal" href="#common-args-common-arguments"><tt class="docutils literal"><span class="pre">common_args</span></tt>: common arguments</a><ul>
<li><a class="reference internal" href="#columns"><tt class="docutils literal"><span class="pre">--columns</span></tt></a></li>
<li><a class="reference internal" href="#min-kindreds-1"><tt class="docutils literal"><span class="pre">--min-kindreds</span> <span class="pre">1</span></tt></a></li>
<li><a class="reference internal" href="#families"><tt class="docutils literal"><span class="pre">--families</span></tt></a></li>
<li><a class="reference internal" href="#filter"><tt class="docutils literal"><span class="pre">--filter</span></tt></a></li>
<li><a class="reference internal" href="#d-0-depth"><tt class="docutils literal"><span class="pre">-d</span> <span class="pre">[0]</span></tt> (depth)</a></li>
<li><a class="reference internal" href="#allow-unaffected"><tt class="docutils literal"><span class="pre">--allow-unaffected</span></tt></a></li>
<li><a class="reference internal" href="#lenient"><tt class="docutils literal"><span class="pre">--lenient</span></tt></a></li>
<li><a class="reference internal" href="#gt-pl-max"><tt class="docutils literal"><span class="pre">--gt-pl-max</span></tt></a></li>
</ul>
</li>
<li><a class="reference internal" href="#comp-hets-identifying-potential-compound-heterozygotes"><tt class="docutils literal"><span class="pre">comp_hets</span></tt>: Identifying potential compound heterozygotes</a><ul>
<li><a class="reference internal" href="#genotype-requirements">Genotype Requirements</a></li>
<li><a class="reference internal" href="#pattern-only">Pattern Only</a></li>
</ul>
</li>
<li><a class="reference internal" href="#mendelian-error-identify-non-mendelian-transmission"><tt class="docutils literal"><span class="pre">mendelian_error</span></tt>: Identify non-mendelian transmission.</a><ul>
<li><a class="reference internal" href="#id1">Genotype Requirements</a></li>
</ul>
</li>
<li><a class="reference internal" href="#de-novo-identifying-potential-de-novo-mutations"><tt class="docutils literal"><span class="pre">de_novo</span></tt>: Identifying potential de novo mutations.</a><ul>
<li><a class="reference internal" href="#id2">Genotype Requirements</a></li>
<li><a class="reference internal" href="#example">example</a></li>
<li><a class="reference internal" href="#id3"><tt class="docutils literal"><span class="pre">example</span></tt></a></li>
<li><a class="reference internal" href="#id4"><tt class="docutils literal"><span class="pre">example</span></tt></a></li>
<li><a class="reference internal" href="#id5"><tt class="docutils literal"><span class="pre">example</span></tt></a></li>
</ul>
</li>
<li><a class="reference internal" href="#autosomal-recessive-find-variants-meeting-an-autosomal-recessive-model"><tt class="docutils literal"><span class="pre">autosomal_recessive</span></tt>: Find variants meeting an autosomal recessive model.</a><ul>
<li><a class="reference internal" href="#id6">Genotype Requirements</a></li>
<li><a class="reference internal" href="#default-behavior"><tt class="docutils literal"><span class="pre">default</span> <span class="pre">behavior</span></tt></a></li>
</ul>
</li>
<li><a class="reference internal" href="#autosomal-dominant-find-variants-meeting-an-autosomal-dominant-model"><tt class="docutils literal"><span class="pre">autosomal_dominant</span></tt>: Find variants meeting an autosomal dominant model.</a><ul>
<li><a class="reference internal" href="#id7">Genotype Requirements</a></li>
<li><a class="reference internal" href="#id8"><tt class="docutils literal"><span class="pre">default</span> <span class="pre">behavior</span></tt></a></li>
</ul>
</li>
<li><a class="reference internal" href="#pathways-map-genes-and-variants-to-kegg-pathways"><tt class="docutils literal"><span class="pre">pathways</span></tt>: Map genes and variants to KEGG pathways.</a><ul>
<li><a class="reference internal" href="#lof"><tt class="docutils literal"><span class="pre">--lof</span></tt></a></li>
</ul>
</li>
<li><a class="reference internal" href="#interactions-find-genes-among-variants-that-are-interacting-partners"><tt class="docutils literal"><span class="pre">interactions</span></tt>: Find genes among variants that are interacting partners.</a><ul>
<li><a class="reference internal" href="#lof-interactions"><tt class="docutils literal"><span class="pre">lof_interactions</span></tt></a></li>
<li><a class="reference internal" href="#var"><tt class="docutils literal"><span class="pre">--var</span></tt></a></li>
</ul>
</li>
<li><a class="reference internal" href="#lof-sieve-filter-lof-variants-by-transcript-position-and-type"><tt class="docutils literal"><span class="pre">lof_sieve</span></tt>: Filter LoF variants by transcript position and type</a></li>
<li><a class="reference internal" href="#annotate-adding-your-own-custom-annotations"><tt class="docutils literal"><span class="pre">annotate</span></tt>: adding your own custom annotations</a><ul>
<li><a class="reference internal" href="#a-boolean-did-a-variant-overlap-a-region-or-not"><tt class="docutils literal"><span class="pre">-a</span> <span class="pre">boolean</span></tt> Did a variant overlap a region or not?</a></li>
<li><a class="reference internal" href="#a-count-how-many-regions-did-a-variant-overlap"><tt class="docutils literal"><span class="pre">-a</span> <span class="pre">count</span></tt> How many regions did a variant overlap?</a></li>
<li><a class="reference internal" href="#a-extract-extract-specific-values-from-a-bed-file"><tt class="docutils literal"><span class="pre">-a</span> <span class="pre">extract</span></tt> Extract specific values from a BED file</a></li>
<li><a class="reference internal" href="#t-type-specifying-the-column-type-s-when-using-a-extract"><tt class="docutils literal"><span class="pre">-t</span> <span class="pre">TYPE</span></tt> Specifying the column type(s) when using <tt class="docutils literal"><span class="pre">-a</span> <span class="pre">extract</span></tt></a></li>
<li><a class="reference internal" href="#o-operation-specifying-the-summary-operations-when-using-a-extract"><tt class="docutils literal"><span class="pre">-o</span> <span class="pre">OPERATION</span></tt> Specifying the summary operations when using <tt class="docutils literal"><span class="pre">-a</span> <span class="pre">extract</span></tt></a></li>
<li><a class="reference internal" href="#annotating-with-vcf">Annotating with VCF</a></li>
<li><a class="reference internal" href="#extracting-and-populating-multiple-columns-at-once">Extracting and populating multiple columns at once.</a></li>
</ul>
</li>
<li><a class="reference internal" href="#region-extracting-variants-from-specific-regions-or-genes"><tt class="docutils literal"><span class="pre">region</span></tt>: Extracting variants from specific regions or genes</a><ul>
<li><a class="reference internal" href="#reg"><tt class="docutils literal"><span class="pre">--reg</span></tt></a></li>
<li><a class="reference internal" href="#gene"><tt class="docutils literal"><span class="pre">--gene</span></tt></a></li>
<li><a class="reference internal" href="#id9"><tt class="docutils literal"><span class="pre">--columns</span></tt></a></li>
<li><a class="reference internal" href="#id10"><tt class="docutils literal"><span class="pre">--filter</span></tt></a></li>
<li><a class="reference internal" href="#json"><tt class="docutils literal"><span class="pre">--json</span></tt></a></li>
</ul>
</li>
<li><a class="reference internal" href="#windower-conducting-analyses-on-genome-windows"><tt class="docutils literal"><span class="pre">windower</span></tt>: Conducting analyses on genome &#8220;windows&#8221;.</a></li>
<li><a class="reference internal" href="#stats-compute-useful-variant-statistics"><tt class="docutils literal"><span class="pre">stats</span></tt>: Compute useful variant statistics.</a><ul>
<li><a class="reference internal" href="#tstv-coding"><tt class="docutils literal"><span class="pre">--tstv-coding</span></tt></a></li>
<li><a class="reference internal" href="#tstv-noncoding"><tt class="docutils literal"><span class="pre">--tstv-noncoding</span></tt></a></li>
<li><a class="reference internal" href="#summarize"><tt class="docutils literal"><span class="pre">--summarize</span></tt></a></li>
</ul>
</li>
<li><a class="reference internal" href="#burden-perform-sample-wise-gene-level-burden-calculations"><tt class="docutils literal"><span class="pre">burden</span></tt>: perform sample-wise gene-level burden calculations</a><ul>
<li><a class="reference internal" href="#nonsynonymous"><tt class="docutils literal"><span class="pre">--nonsynonymous</span></tt></a></li>
<li><a class="reference internal" href="#calpha"><tt class="docutils literal"><span class="pre">--calpha</span></tt></a></li>
<li><a class="reference internal" href="#min-aaf-and-max-aaf-for-calpha"><tt class="docutils literal"><span class="pre">--min-aaf</span></tt> and <tt class="docutils literal"><span class="pre">--max-aaf</span></tt> for <tt class="docutils literal"><span class="pre">--calpha</span></tt></a></li>
<li><a class="reference internal" href="#cases-and-controls-for-calpha"><tt class="docutils literal"><span class="pre">--cases</span></tt> and <tt class="docutils literal"><span class="pre">--controls</span> <span class="pre">for</span> <span class="pre">``--calpha</span></tt></a></li>
<li><a class="reference internal" href="#nonsynonymous-calpha"><tt class="docutils literal"><span class="pre">--nonsynonymous</span></tt> <tt class="docutils literal"><span class="pre">--calpha</span></tt></a></li>
</ul>
</li>
<li><a class="reference internal" href="#roh-identifying-runs-of-homozygosity"><tt class="docutils literal"><span class="pre">ROH</span></tt>: Identifying runs of homozygosity</a></li>
<li><a class="reference internal" href="#set-somatic-flag-somatic-variants"><tt class="docutils literal"><span class="pre">set_somatic</span></tt>: Flag somatic variants</a><ul>
<li><a class="reference internal" href="#id13"><tt class="docutils literal"><span class="pre">default</span> <span class="pre">behavior</span></tt></a></li>
<li><a class="reference internal" href="#min-depth-none"><tt class="docutils literal"><span class="pre">--min-depth</span> <span class="pre">[None]</span></tt></a></li>
<li><a class="reference internal" href="#min-qual-none"><tt class="docutils literal"><span class="pre">--min-qual</span> <span class="pre">[None]</span></tt></a></li>
<li><a class="reference internal" href="#min-somatic-score-none"><tt class="docutils literal"><span class="pre">--min-somatic-score</span> <span class="pre">[None]</span></tt></a></li>
<li><a class="reference internal" href="#max-norm-alt-freq-none"><tt class="docutils literal"><span class="pre">--max-norm-alt-freq</span> <span class="pre">[None]</span></tt></a></li>
<li><a class="reference internal" href="#max-norm-alt-count-none"><tt class="docutils literal"><span class="pre">--max-norm-alt-count</span> <span class="pre">[None]</span></tt></a></li>
<li><a class="reference internal" href="#min-norm-depth-none"><tt class="docutils literal"><span class="pre">--min-norm-depth</span> <span class="pre">[None]</span></tt></a></li>
<li><a class="reference internal" href="#min-tumor-alt-freq-none"><tt class="docutils literal"><span class="pre">--min-tumor-alt-freq</span> <span class="pre">[None]</span></tt></a></li>
<li><a class="reference internal" href="#min-tumor-alt-count-none"><tt class="docutils literal"><span class="pre">--min-tumor-alt-count</span> <span class="pre">[None]</span></tt></a></li>
<li><a class="reference internal" href="#min-tumor-depth-none"><tt class="docutils literal"><span class="pre">--min-tumor-depth</span> <span class="pre">[None]</span></tt></a></li>
<li><a class="reference internal" href="#chrom-none"><tt class="docutils literal"><span class="pre">--chrom</span> <span class="pre">[None]</span></tt></a></li>
<li><a class="reference internal" href="#dry-run"><tt class="docutils literal"><span class="pre">--dry-run</span></tt></a></li>
</ul>
</li>
<li><a class="reference internal" href="#actionable-mutations-report-actionable-somatic-mutations-and-drug-gene-interactions"><tt class="docutils literal"><span class="pre">actionable_mutations</span></tt>: Report actionable somatic mutations and drug-gene interactions</a></li>
<li><a class="reference internal" href="#fusions-report-putative-gene-fusions"><tt class="docutils literal"><span class="pre">fusions</span></tt>: Report putative gene fusions</a><ul>
<li><a class="reference internal" href="#id14"><tt class="docutils literal"><span class="pre">default</span> <span class="pre">behavior</span></tt></a></li>
<li><a class="reference internal" href="#id15"><tt class="docutils literal"><span class="pre">--min_qual</span> <span class="pre">[None]</span></tt></a></li>
<li><a class="reference internal" href="#evidence-type-string"><tt class="docutils literal"><span class="pre">--evidence_type</span> <span class="pre">STRING</span></tt></a></li>
<li><a class="reference internal" href="#in-cosmic-census"><tt class="docutils literal"><span class="pre">--in_cosmic_census</span></tt></a></li>
</ul>
</li>
<li><a class="reference internal" href="#db-info-list-the-gemini-database-tables-and-columns"><tt class="docutils literal"><span class="pre">db_info</span></tt>: List the gemini database tables and columns</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="querying.html"
                        title="previous chapter">Querying the GEMINI database</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="browser.html"
                        title="next chapter">The GEMINI browser interface</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/content/tools.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="browser.html" title="The GEMINI browser interface"
             >next</a> |</li>
        <li class="right" >
          <a href="querying.html" title="Querying the GEMINI database"
             >previous</a> |</li>
        <li><a href="../index.html">gemini v0.16.3</a> &raquo;</li> 
      </ul>
    </div>

    <div class="footer">
        &copy; Copyright 2012,2013,2014,2015.
      Last updated on Jul 22, 2015.
    <a href="http://www.gnu.org/licenses/gpl-3.0.txt"><p>GPL3 licensed</p></a>
    </div>

<!-- 
<div id="editor-trap">
      <h3>Edit and improve this document!</h3>

      <p>
        This file can be edited directly through the Web. Anyone can
        update and fix errors in this document with few clicks --
        no downloads needed.
      <p>

      <ol>

        <li>
          Go to
        <a href="https://github.com/arq5x/gemini/blob/master/docs/content/tools.rst">
          Built-in analysis tools
        </a> on GitHub.
      </li>

        <li>
        <b>Edit</b> files using GitHub's text editor in your web browser (see the 'Edit' tab on the top right of the file)
      </li>

      <li>
        Fill in the <b>Commit message</b> text box at the bottom of the page describing <i>why</i>
        you made the changes. Press the <b>Propose file change</b> button next to it when done.
      </li>

      <li>
        Then click <b>Send a pull request</b>.
      </li>

      <li>
        Your changes are now queued for review under the project's <a href="https://github.com/arq5x/gemini/pulls">Pull requests</a> tab on GitHub!
      </li>
      </ol>

      <p>
        For an introduction to the documentation format please see <a href="http://docutils.sourceforge.net/docs/user/rst/quickstart.html">the reST primer</a>.
      </p>

  </div>
 -->
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', UA-39150563-1]);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>


  </body>
</html>